// @a5c-ai/babysitter-sdk/dist/runtime/processContext.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createProcessContext = createProcessContext;
exports.withProcessContext = withProcessContext;
exports.getActiveProcessContext = getActiveProcessContext;
exports.requireProcessContext = requireProcessContext;
const async_hooks_1 = require("async_hooks");
const task_1 = require("./intrinsics/task");
const breakpoint_1 = require("./intrinsics/breakpoint");
const sleep_1 = require("./intrinsics/sleep");
const orchestratorTask_1 = require("./intrinsics/orchestratorTask");
const hook_1 = require("./intrinsics/hook");
const dispatcher_1 = require("../hooks/dispatcher");
const parallel_1 = require("./intrinsics/parallel");
const exceptions_1 = require("./exceptions");
const runLogger_1 = require("../logging/runLogger");
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
const contextStorage = new async_hooks_1.AsyncLocalStorage();
function createProcessContext(init) {
    const safeLogger = typeof init.logger === "function" ? init.logger : undefined;
    const internal = {
        ...init,
        logger: safeLogger,
        now: init.now ?? (() => new Date()),
        logSeq: 0,
        recordedLogSeqs: init.recordedLogSeqs ?? new Set(),
        nonInteractive: init.nonInteractive ?? false,
    };
    const parallelHelpers = {
        all: (thunks) => (0, parallel_1.runParallelAll)(thunks),
        map: (items, fn) => (0, parallel_1.runParallelMap)(items, fn),
    };
    const processContext = {
        now: () => internal.now(),
        task: (task, args, options) => (0, task_1.runTaskIntrinsic)({
            task,
            args,
            invokeOptions: options,
            context: internal,
        }),
        breakpoint: (payload, options) => (0, breakpoint_1.runBreakpointIntrinsic)(payload, internal, options),
        sleepUntil: (target, options) => (0, sleep_1.runSleepIntrinsic)(target, internal, options),
        orchestratorTask: (payload, options) => (0, orchestratorTask_1.runOrchestratorTaskIntrinsic)(payload, internal, options),
        hook: (hookType, payload, options) => (0, hook_1.runHookIntrinsic)(hookType, payload, internal, options),
        parallel: parallelHelpers,
        // Always provide a callable logger to processes so `ctx.log(...)` never throws.
        //
        // Replay-aware: each ctx.log call gets a sequential logSeq. If the seq
        // is already in the journal (from a previous iteration), the call is a
        // no-op.  Only NEW log entries write to both the journal and the log
        // file, preventing replay duplicates.
        log: (...args) => {
            // Support ctx.log('label', 'message') and ctx.log('message')
            let label;
            let message;
            if (args.length >= 2 && typeof args[0] === "string" && typeof args[1] === "string") {
                label = args[0];
                message = args[1];
            }
            else if (typeof args[0] === "string") {
                message = args[0];
            }
            else {
                return;
            }
            if (!message)
                return;
            // Assign a sequential log number for replay deduplication
            const seq = ++internal.logSeq;
            // Already recorded in a previous iteration — skip
            if (internal.recordedLogSeqs.has(seq)) {
                return;
            }
            // Mark as recorded so same-iteration duplicates are also skipped
            internal.recordedLogSeqs.add(seq);
            // 1. Record log seq to state file for replay deduplication.
            //    Written to a separate file (not journal) to avoid race conditions
            //    with journal sequence numbering from fire-and-forget writes.
            void node_fs_1.promises.appendFile(path.join(internal.runDir, "state", "logSeqs.txt"), `${seq}\n`).catch(() => {
                // Never let logging break orchestration.
            });
            // 2. Write to structured log file (~/.a5c/logs/<runId>/process.log)
            //    Only written when the journal entry is new (deduped above).
            void (0, runLogger_1.appendRunLog)({
                timestamp: new Date().toISOString(),
                level: "info",
                type: "process",
                label,
                message,
                runId: internal.runId,
                processId: internal.processId,
                source: "ctx.log",
            }).catch(() => {
                // Never let logging break orchestration.
            });
            // 3. Dispatch babysitter-log hook for extensibility
            const hookPayload = label ? `${label} ${message}` : message;
            void (0, dispatcher_1.callHook)({
                hookType: "babysitter-log",
                payload: hookPayload,
                cwd: internal.runDir,
            }).catch(() => {
                // Never let logging break orchestration.
            });
        },
    };
    return {
        context: processContext,
        internalContext: internal,
    };
}
function withProcessContext(internal, fn) {
    return contextStorage.run(internal, () => Promise.resolve().then(fn));
}
function getActiveProcessContext() {
    return contextStorage.getStore();
}
function requireProcessContext() {
    const ctx = getActiveProcessContext();
    if (!ctx) {
        throw new exceptions_1.MissingProcessContextError();
    }
    return ctx;
}
