// @a5c-ai/babysitter-sdk/dist/runtime/intrinsics/breakpoint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runBreakpointIntrinsic = runBreakpointIntrinsic;
const task_1 = require("./task");
const journal_1 = require("../../storage/journal");
const BREAKPOINT_TASK_ID = "__sdk.breakpoint";
const DEFAULT_BREAKPOINT_LABEL = "breakpoint";
const breakpointTask = {
    id: BREAKPOINT_TASK_ID,
    build(args) {
        return {
            kind: "breakpoint",
            title: args.label,
            metadata: {
                payload: args?.payload,
                requestedAt: args.requestedAt,
                label: args.label,
                expert: args.expert,
                tags: args.tags,
                strategy: args.strategy,
            },
        };
    },
};
function runBreakpointIntrinsic(payload, context, options) {
    const label = deriveBreakpointLabel(payload, options?.label);
    // In non-interactive mode, auto-approve breakpoints without dispatching a task.
    const ctx = context;
    if (ctx.nonInteractive) {
        const bpLabel = options?.label ?? "unnamed";
        void (0, journal_1.appendEvent)({
            runDir: context.runDir,
            eventType: "PROCESS_LOG",
            event: { logSeq: -1, label: "breakpoint:skipped", message: `Breakpoint '${bpLabel}' auto-approved (non-interactive mode)` },
        }).catch(() => {
            // Never let logging break orchestration.
        });
        return Promise.resolve({ approved: true, response: "Auto-approved (non-interactive mode)" });
    }
    const invokeOptions = { ...options, label };
    return (0, task_1.runTaskIntrinsic)({
        task: breakpointTask,
        args: {
            payload,
            label,
            requestedAt: context.now().toISOString(),
            expert: options?.expert,
            tags: options?.tags,
            strategy: options?.strategy,
        },
        invokeOptions,
        context,
    });
}
function deriveBreakpointLabel(payload, provided) {
    if (typeof provided === "string" && provided.length) {
        return provided;
    }
    if (payload && typeof payload === "object" && "label" in payload) {
        const inferred = payload.label;
        if (typeof inferred === "string" && inferred.length) {
            return inferred;
        }
    }
    return DEFAULT_BREAKPOINT_LABEL;
}
