// @a5c-ai/babysitter-sdk/dist/runtime/replay/createReplayEngine.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createReplayEngine = createReplayEngine;
const runFiles_1 = require("../../storage/runFiles");
const journal_1 = require("../../storage/journal");
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
const effectIndex_1 = require("./effectIndex");
const replayCursor_1 = require("./replayCursor");
const processContext_1 = require("../processContext");
const constants_1 = require("../constants");
const exceptions_1 = require("../exceptions");
const stateCache_1 = require("./stateCache");
async function createReplayEngine(options) {
    const metadata = await (0, runFiles_1.readRunMetadata)(options.runDir);
    ensureCompatibleLayout(metadata.layoutVersion, options.runDir);
    const inputs = await (0, runFiles_1.readRunInputs)(options.runDir);
    // Load journal once — shared by effect index builder and log seq scanner.
    // Wraps parse errors into RunFailedError for consistency with the old
    // code-path where buildEffectIndex loaded the journal internally.
    let journal;
    try {
        journal = await (0, journal_1.loadJournal)(options.runDir);
    }
    catch (error) {
        const err = error;
        if (err?.code === "JOURNAL_PARSE_FAILED") {
            throw new exceptions_1.RunFailedError("Failed to parse journal event", {
                path: err.path,
                runDir: options.runDir,
                error: err.message,
            });
        }
        throw error;
    }
    const effectIndex = await (0, effectIndex_1.buildEffectIndex)({ runDir: options.runDir, events: journal });
    const { snapshot: stateCacheSnapshot, rebuildMeta: stateRebuild } = await resolveStateCacheSnapshot({
        runDir: options.runDir,
        effectIndex,
    });
    // Build set of already-recorded log seqs for replay deduplication.
    // Read from state/logSeqs.txt (not journal) to avoid race conditions.
    const recordedLogSeqs = await readRecordedLogSeqs(options.runDir);
    const replayCursor = new replayCursor_1.ReplayCursor();
    const processId = metadata.processId ?? metadata.request ?? metadata.runId;
    const { context, internalContext } = (0, processContext_1.createProcessContext)({
        runId: metadata.runId,
        runDir: options.runDir,
        processId,
        effectIndex,
        replayCursor,
        now: options.now,
        logger: options.logger,
        recordedLogSeqs,
        nonInteractive: Boolean(metadata.nonInteractive),
    });
    return {
        runId: metadata.runId,
        runDir: options.runDir,
        metadata,
        inputs,
        effectIndex,
        replayCursor,
        context,
        internalContext,
        stateCache: stateCacheSnapshot,
        stateRebuild,
    };
}
async function resolveStateCacheSnapshot({ runDir, effectIndex, }) {
    let existingSnapshot = null;
    let corrupted = false;
    try {
        existingSnapshot = await (0, stateCache_1.readStateCache)(runDir);
    }
    catch {
        corrupted = true;
    }
    if (corrupted || !existingSnapshot) {
        const reason = corrupted ? "corrupt_cache" : "missing_cache";
        const rebuilt = await (0, stateCache_1.rebuildStateCache)(runDir, { effectIndex, reason });
        return { snapshot: rebuilt, rebuildMeta: { reason, previous: null } };
    }
    const journalHead = effectIndex.getJournalHead() ?? null;
    if (!(0, stateCache_1.journalHeadsEqual)(existingSnapshot.journalHead, journalHead)) {
        const rebuilt = await (0, stateCache_1.rebuildStateCache)(runDir, {
            effectIndex,
            reason: "journal_mismatch",
        });
        return {
            snapshot: rebuilt,
            rebuildMeta: { reason: "journal_mismatch", previous: existingSnapshot.journalHead ?? null },
        };
    }
    return { snapshot: existingSnapshot, rebuildMeta: null };
}
/**
 * Read recorded log sequence numbers from the state/logSeqs.txt file.
 * Each line contains one seq number. Written by ctx.log in processContext.
 * Uses a separate file (not journal) to avoid race conditions between
 * fire-and-forget log writes and awaited effect writes.
 */
async function readRecordedLogSeqs(runDir) {
    const seqs = new Set();
    try {
        const content = await node_fs_1.promises.readFile(path.join(runDir, "state", "logSeqs.txt"), "utf8");
        for (const line of content.split("\n")) {
            const n = Number(line.trim());
            if (Number.isFinite(n) && n > 0) {
                seqs.add(n);
            }
        }
    }
    catch {
        // File doesn't exist yet — no log seqs recorded.
    }
    return seqs;
}
function ensureCompatibleLayout(layoutVersion, runDir) {
    if (!layoutVersion) {
        throw new exceptions_1.RunFailedError("Run metadata is missing layoutVersion", { runDir });
    }
    if (layoutVersion !== constants_1.replaySchemaVersion) {
        throw new exceptions_1.RunFailedError("Run layout version is not supported by this runtime", {
            expected: constants_1.replaySchemaVersion,
            actual: layoutVersion,
            runDir,
        });
    }
}
