// @agentsh/secure-sandbox/dist/chunk-LMN3KM53.js
import {
  envPrefix,
  shellEscape
} from "./chunk-OANLKSOD.js";

// src/adapters/cloudflare.ts
function cloudflare(sandbox) {
  return {
    async exec(cmd, args, opts) {
      let command = `${envPrefix(opts?.env)}${shellEscape(cmd, args)}`;
      if (opts?.detached) {
        sandbox.exec(`nohup ${command} > /dev/null 2>&1 &`, { cwd: opts?.cwd }).catch(() => {
        });
        return { stdout: "", stderr: "", exitCode: 0 };
      }
      const result = await sandbox.exec(command, { cwd: opts?.cwd });
      return {
        stdout: result.stdout ?? "",
        stderr: result.stderr ?? "",
        exitCode: result.exitCode
      };
    },
    async writeFile(path, content) {
      const buf = Buffer.isBuffer(content) ? content : Buffer.from(content);
      const b64 = buf.toString("base64");
      const cmd = shellEscape("sh", ["-c", 'printf "%s" "$1" | base64 -d > "$2"', "_", b64, path]);
      const result = await sandbox.exec(cmd);
      if ((result.exitCode ?? 0) !== 0) {
        throw new Error(`writeFile failed (exit ${result.exitCode}): ${result.stderr ?? ""}`);
      }
    },
    async readFile(path) {
      const cmd = shellEscape("cat", [path]);
      const result = await sandbox.exec(cmd);
      if ((result.exitCode ?? 0) !== 0) {
        throw new Error(`readFile failed (exit ${result.exitCode}): ${result.stderr ?? ""}`);
      }
      return result.stdout ?? "";
    },
    async stop() {
    }
  };
}

export {
  cloudflare
};
