// @agentsh/secure-sandbox/dist/chunk-LNDICGZU.js
// src/core/errors.ts
var AgentSHError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "AgentSHError";
  }
};
var PolicyValidationError = class extends AgentSHError {
  issues;
  constructor({ issues }) {
    const summaries = issues.map((issue) => `${issue.path.join(".")}: ${issue.message}`).join("; ");
    super(`Policy validation failed: ${summaries}`);
    this.name = "PolicyValidationError";
    this.issues = issues;
  }
};
var MissingPeerDependencyError = class extends AgentSHError {
  packageName;
  versionRange;
  constructor({
    packageName,
    versionRange
  }) {
    super(
      `${packageName} is required but not installed. Run: npm install ${packageName}@"${versionRange}"`
    );
    this.name = "MissingPeerDependencyError";
    this.packageName = packageName;
    this.versionRange = versionRange;
  }
};
var IncompatibleProviderVersionError = class extends AgentSHError {
  installed;
  required;
  packageName;
  constructor({
    installed,
    required,
    packageName
  }) {
    super(
      `${packageName} version ${installed} is not supported. @agentsh/secure-sandbox requires ${packageName} ${required}. Please upgrade: npm install ${packageName}@latest`
    );
    this.name = "IncompatibleProviderVersionError";
    this.installed = installed;
    this.required = required;
    this.packageName = packageName;
  }
};
var ProvisioningError = class extends AgentSHError {
  phase;
  command;
  stderr;
  constructor({
    phase,
    command,
    stderr
  }) {
    super(`Provisioning failed at phase: ${phase}`);
    this.name = "ProvisioningError";
    this.phase = phase;
    this.command = command;
    this.stderr = stderr;
  }
};
var IntegrityError = class extends AgentSHError {
  expected;
  actual;
  constructor({
    expected,
    actual,
    message
  }) {
    super(message ?? `Checksum mismatch: expected ${expected}, got ${actual}`);
    this.name = "IntegrityError";
    this.expected = expected;
    this.actual = actual;
  }
};
var RuntimeError = class extends AgentSHError {
  sessionId;
  command;
  stderr;
  constructor({
    sessionId,
    command,
    stderr
  }) {
    super(`agentsh exec failed (session ${sessionId})`);
    this.name = "RuntimeError";
    this.sessionId = sessionId;
    this.command = command;
    this.stderr = stderr;
  }
};

// src/policies/schema.ts
import { z, ZodError } from "zod";
var stringOrArray = z.union([z.string(), z.array(z.string())]);
var FileOpSchema = z.enum(["read", "write", "create", "delete"]);
var FileAllowRule = z.object({ allow: stringOrArray, ops: z.array(FileOpSchema).optional() }).strict();
var FileDenyRule = z.object({ deny: stringOrArray, ops: z.array(FileOpSchema).optional() }).strict();
var FileRedirectRule = z.object({
  redirect: stringOrArray,
  to: z.string(),
  ops: z.array(FileOpSchema).optional()
}).strict();
var FileAuditRule = z.object({ audit: stringOrArray, ops: z.array(FileOpSchema).optional() }).strict();
var FileSoftDeleteRule = z.object({ softDelete: stringOrArray }).strict();
var FileRuleSchema = z.union([
  FileAllowRule,
  FileDenyRule,
  FileRedirectRule,
  FileAuditRule,
  FileSoftDeleteRule
]);
var NetworkAllowRule = z.object({
  allow: stringOrArray,
  ports: z.array(z.number().int().min(1).max(65535)).optional()
}).strict();
var NetworkDenyRule = z.object({ deny: stringOrArray }).strict();
var NetworkRedirectRule = z.object({ redirect: z.string(), to: z.string() }).strict();
var NetworkRuleSchema = z.union([
  NetworkAllowRule,
  NetworkDenyRule,
  NetworkRedirectRule
]);
var CommandRedirectTarget = z.union([
  z.string(),
  z.object({ cmd: z.string(), args: z.array(z.string()) }).strict()
]);
var CommandAllowRule = z.object({ allow: stringOrArray }).strict();
var CommandDenyRule = z.object({ deny: stringOrArray }).strict();
var CommandRedirectRule = z.object({ redirect: stringOrArray, to: CommandRedirectTarget }).strict();
var CommandRuleSchema = z.union([
  CommandAllowRule,
  CommandDenyRule,
  CommandRedirectRule
]);
var EnvRuleSchema = z.object({
  commands: z.array(z.string()),
  allow: z.array(z.string()).optional(),
  deny: z.array(z.string()).optional()
}).strict();
var DnsRedirectSchema = z.object({
  match: z.string(),
  resolveTo: z.string()
}).strict();
var ConnectRedirectSchema = z.object({
  match: z.string(),
  redirectTo: z.string()
}).strict();
var LicenseSpdxMatchSchema = z.object({
  allow: z.array(z.string()).optional(),
  deny: z.array(z.string()).optional()
}).strict();
var PackageMatchSchema = z.object({
  packages: z.array(z.string()).optional(),
  namePatterns: z.array(z.string()).optional(),
  findingType: z.string().optional(),
  severity: stringOrArray.optional(),
  reasons: z.array(z.string()).optional(),
  licenseSpdx: LicenseSpdxMatchSchema.optional(),
  ecosystem: z.string().optional(),
  options: z.record(z.unknown()).optional()
}).strict();
var PackageRuleSchema = z.object({
  match: PackageMatchSchema,
  action: z.enum(["allow", "warn", "approve", "block"]),
  reason: z.string().optional()
}).strict();
var PolicyDefinitionSchema = z.object({
  file: z.array(FileRuleSchema).optional(),
  network: z.array(NetworkRuleSchema).optional(),
  commands: z.array(CommandRuleSchema).optional(),
  env: z.array(EnvRuleSchema).optional(),
  dns: z.array(DnsRedirectSchema).optional(),
  connect: z.array(ConnectRedirectSchema).optional(),
  packageRules: z.array(PackageRuleSchema).optional()
}).strict();
function validatePolicy(policy) {
  try {
    return PolicyDefinitionSchema.parse(policy);
  } catch (err) {
    if (err instanceof ZodError) {
      throw new PolicyValidationError({ issues: err.issues });
    }
    throw err;
  }
}

// src/policies/merge.ts
var CATEGORIES = ["file", "network", "commands", "env", "dns", "connect", "packageRules"];
function merge(base, ...overrides) {
  return validatePolicy(mergeInternal(base, overrides, "append"));
}
function mergePrepend(base, ...overrides) {
  return validatePolicy(mergeInternal(base, overrides, "prepend"));
}
function mergeInternal(base, overrides, mode) {
  const result = { ...base };
  for (const override of overrides) {
    for (const key of CATEGORIES) {
      if (override[key] != null) {
        const baseRules = result[key] ?? [];
        result[key] = mode === "append" ? [...baseRules, ...override[key]] : [...override[key], ...baseRules];
      }
    }
  }
  return result;
}

// src/policies/presets.ts
function agentDefault(extensions) {
  const base = {
    file: [
      { allow: "/workspace/**", ops: ["read", "write", "create"] },
      // Git/version-control credentials
      { deny: ["/workspace/.git/config", "/workspace/.netrc"] },
      // Secrets and credentials
      { deny: ["**/.env", "**/.env.*", "**/credentials*", "**/*.pem", "**/*.key"] },
      { deny: ["~/.ssh/**", "/proc/*/environ"] },
      // Cloud provider credentials
      { deny: ["~/.aws/**", "~/.gcp/**", "~/.azure/**", "~/.config/gcloud/**"] },
      // Shell config injection (persistence)
      { deny: ["~/.bashrc", "~/.zshrc", "~/.profile", "~/.bash_profile"] },
      // Credential stores
      { deny: ["~/.gitconfig", "~/.netrc", "~/.curlrc", "~/.wgetrc"] },
      // PATH hijacking
      { deny: "~/.local/bin/**" },
      // Agent config files — allow reads (project context), deny writes (prompt injection persistence)
      { deny: ["**/.cursorrules", "**/CLAUDE.md", "**/copilot-instructions.md"], ops: ["write", "create", "delete"] }
    ],
    network: [
      {
        allow: [
          "registry.npmjs.org",
          "registry.yarnpkg.com",
          "pypi.org",
          "files.pythonhosted.org",
          "crates.io",
          "static.crates.io",
          "index.crates.io",
          "proxy.golang.org",
          "sum.golang.org",
          "github.com",
          "raw.githubusercontent.com"
        ],
        ports: [443]
      },
      { deny: "*" }
    ],
    commands: [
      // Allow safe commands (order matters — first match wins)
      {
        allow: [
          "bash",
          "sh",
          "echo",
          "cat",
          "head",
          "tail",
          "grep",
          "find",
          "ls",
          "wc",
          "sort",
          "uniq",
          "diff",
          "pwd",
          "date",
          "which",
          "whoami",
          "id",
          "uname",
          "printf",
          "test",
          "true",
          "false",
          "mkdir",
          "cp",
          "mv",
          "rm",
          "touch",
          "chmod",
          "tr",
          "cut",
          "sed",
          "awk",
          "tee",
          "xargs",
          "basename",
          "dirname",
          "realpath",
          "base64",
          "md5sum",
          "sha256sum",
          "tar",
          "gzip",
          "gunzip"
        ]
      },
      // Allow dev tools
      {
        allow: [
          "git",
          "node",
          "npm",
          "npx",
          "yarn",
          "pnpm",
          "bun",
          "python",
          "python3",
          "pip",
          "pip3",
          "cargo",
          "rustc",
          "go",
          "make",
          "cmake"
        ]
      },
      // Deny dangerous commands
      { deny: ["env", "printenv", "sudo", "su", "doas"] },
      { deny: ["shutdown", "reboot", "halt", "poweroff"] },
      { deny: ["nc", "ncat", "netcat", "socat", "telnet"] },
      { deny: ["git push --force", "git reset --hard"] },
      {
        redirect: ["curl", "wget"],
        to: { cmd: "agentsh-fetch", args: ["--audit"] }
      }
    ],
    packageRules: [
      // Critical vulnerability = block
      {
        match: { findingType: "vulnerability", severity: "critical" },
        action: "block",
        reason: "Critical vulnerability \u2014 review before installing"
      },
      // Known malware = block
      {
        match: { findingType: "malware" },
        action: "block",
        reason: "Known malware detected"
      },
      // Typosquat = block
      {
        match: { findingType: "reputation", reasons: ["typosquat"] },
        action: "block",
        reason: "Package flagged as potential typosquat"
      },
      // Medium vulnerability = warn
      {
        match: { findingType: "vulnerability", severity: "medium" },
        action: "warn",
        reason: "Medium vulnerability \u2014 review before using"
      },
      // Copyleft licenses = block
      {
        match: {
          findingType: "license",
          licenseSpdx: { deny: ["AGPL-3.0-only", "SSPL-1.0"] }
        },
        action: "block",
        reason: "Copyleft license incompatible with proprietary code"
      },
      // Package too new = approve (requires human confirmation)
      {
        match: {
          findingType: "reputation",
          reasons: ["package_too_new"]
        },
        action: "approve",
        reason: "Package published recently \u2014 requires approval"
      }
    ]
  };
  return extensions ? merge(base, extensions) : base;
}
function devSafe(extensions) {
  const base = {
    file: [
      { allow: "/workspace/**", ops: ["read", "write", "create"] },
      { deny: ["**/.env", "**/.env.*", "**/credentials*", "**/*.pem", "**/*.key"] },
      { deny: ["~/.ssh/**", "/proc/*/environ"] },
      { deny: ["~/.aws/**", "~/.gcp/**", "~/.azure/**", "~/.config/gcloud/**"] },
      { deny: ["~/.bashrc", "~/.zshrc", "~/.profile", "~/.bash_profile"] },
      { deny: ["~/.gitconfig", "~/.netrc", "~/.curlrc", "~/.wgetrc"] }
    ],
    network: [
      {
        allow: ["registry.npmjs.org", "registry.yarnpkg.com"],
        ports: [443]
      }
    ],
    commands: [{ deny: ["env", "printenv", "shutdown", "reboot"] }]
  };
  return extensions ? merge(base, extensions) : base;
}
function ciStrict(extensions) {
  const base = {
    file: [
      { allow: "/workspace/**" },
      { deny: ["**/.env", "**/.env.*", "**/credentials*", "**/*.pem", "**/*.key"] },
      { deny: ["~/.aws/**", "~/.gcp/**", "~/.azure/**", "~/.config/gcloud/**"] },
      { deny: "/**" }
    ],
    network: [
      {
        allow: [
          "registry.npmjs.org",
          "registry.yarnpkg.com",
          "pypi.org",
          "files.pythonhosted.org",
          "crates.io",
          "static.crates.io",
          "index.crates.io",
          "proxy.golang.org",
          "sum.golang.org"
        ],
        ports: [443]
      },
      { deny: "*" }
    ],
    commands: [
      { deny: ["env", "printenv", "shutdown", "reboot", "sudo"] }
    ]
  };
  return extensions ? merge(base, extensions) : base;
}
function agentSandbox(extensions) {
  const base = {
    file: [
      { allow: "/workspace/**", ops: ["read"] },
      { deny: "/**" }
    ],
    network: [{ deny: "*" }],
    commands: [
      { deny: ["env", "printenv", "sudo", "su", "shutdown", "reboot"] }
    ]
  };
  return extensions ? merge(base, extensions) : base;
}

export {
  AgentSHError,
  PolicyValidationError,
  MissingPeerDependencyError,
  IncompatibleProviderVersionError,
  ProvisioningError,
  IntegrityError,
  RuntimeError,
  PolicyDefinitionSchema,
  validatePolicy,
  merge,
  mergePrepend,
  agentDefault,
  devSafe,
  ciStrict,
  agentSandbox
};
