// @agentsh/secure-sandbox/dist/chunk-5IG6ABIZ.js
import {
  blaxel
} from "./chunk-UYEAO27E.js";
import {
  cloudflare
} from "./chunk-LMN3KM53.js";
import {
  daytona
} from "./chunk-45FKFVMC.js";
import {
  e2b
} from "./chunk-2P37YGN7.js";
import {
  envPrefix,
  shellEscape
} from "./chunk-OANLKSOD.js";
import {
  vercel
} from "./chunk-JY5ERJTX.js";
import {
  agentDefault,
  mergePrepend
} from "./chunk-LNDICGZU.js";
import {
  __export
} from "./chunk-PZ5AY32C.js";

// src/adapters/index.ts
var adapters_exports = {};
__export(adapters_exports, {
  blaxel: () => blaxel,
  cloudflare: () => cloudflare,
  daytona: () => daytona,
  e2b: () => e2b,
  modal: () => modal,
  modalDefaults: () => modalDefaults,
  sprites: () => sprites,
  spritesDefaults: () => spritesDefaults,
  vercel: () => vercel
});

// src/adapters/sprites.ts
function sprites(sprite) {
  function sh(cmd, opts) {
    if (opts) return sprite.execFile("sh", ["-c", cmd], opts);
    return sprite.execFile("sh", ["-c", cmd]);
  }
  return {
    async exec(cmd, args, opts) {
      const command = `${envPrefix(opts?.env)}${shellEscape(cmd, args)}`;
      const fullCmd = opts?.sudo ? `sudo ${command}` : command;
      try {
        if (opts?.detached) {
          sh(`nohup ${fullCmd} > /dev/null 2>&1 &`).catch(() => {
          });
          return { stdout: "", stderr: "", exitCode: 0 };
        }
        const result = await sh(fullCmd, { cwd: opts?.cwd });
        return {
          stdout: result.stdout ?? "",
          stderr: result.stderr ?? "",
          exitCode: 0
        };
      } catch (err) {
        return {
          stdout: err.stdout ?? "",
          stderr: err.stderr ?? err.message ?? "",
          exitCode: err.exitCode ?? err.code ?? 1
        };
      }
    },
    async writeFile(path, content) {
      const buf = Buffer.isBuffer(content) ? content : Buffer.from(content);
      const b64 = buf.toString("base64");
      try {
        await sh(`printf '%s' '${b64}' | base64 -d > '${path.replace(/'/g, "'\\''")}'`);
      } catch (err) {
        throw new Error(`writeFile failed (exit ${err.exitCode ?? err.code ?? 1}): ${err.stderr ?? err.message ?? ""}`);
      }
    },
    async readFile(path) {
      try {
        const result = await sh(`cat '${path.replace(/'/g, "'\\''")}'`);
        return result.stdout ?? "";
      } catch (err) {
        throw new Error(`readFile failed (exit ${err.exitCode ?? err.code ?? 1}): ${err.stderr ?? err.message ?? ""}`);
      }
    },
    async stop() {
      await sprite.delete();
    }
  };
}
function spritesDefaults() {
  const serverConfig = {
    grpc: { addr: "0.0.0.0:50051" },
    logging: { level: "info", format: "json", output: "stdout" },
    sessions: {
      defaultTimeout: "30m",
      idleTimeout: "10m",
      cleanupInterval: "5m"
    },
    audit: { enabled: true, sqlitePath: "/var/lib/agentsh/audit.db" },
    sandboxLimits: { maxMemoryMb: 512, maxCpuPercent: 90, maxProcesses: 100 },
    fuse: { deferred: true },
    networkIntercept: { interceptMode: "tproxy", proxyListenAddr: "127.0.0.1:8888" },
    seccompDetails: {
      execve: true,
      fileMonitor: { enabled: true, enforceWithoutFuse: true }
    },
    cgroups: { enabled: true },
    unixSockets: { enabled: true },
    proxy: { mode: "mitm", port: 8080 },
    dlp: {
      mode: "redact",
      patterns: { credit_card: true, ssn: true, api_key: true }
    },
    approvals: { enabled: false },
    metrics: { enabled: true, path: "/metrics" },
    health: { path: "/healthz", readinessPath: "/readyz" },
    development: { disableAuth: false, verboseErrors: false }
  };
  return {
    installStrategy: "preinstalled",
    realPaths: true,
    serverConfig
  };
}

// src/adapters/modal.ts
function modal(sandbox) {
  async function run(cmd) {
    try {
      const proc = await sandbox.exec("sh", "-c", cmd);
      await proc.wait();
      const stdout = typeof proc.stdout?.read === "function" ? await proc.stdout.read() : proc.stdout ?? "";
      const stderr = typeof proc.stderr?.read === "function" ? await proc.stderr.read() : proc.stderr ?? "";
      const exitCode = proc.returncode ?? proc.exitCode ?? 0;
      return { stdout, stderr, exitCode };
    } catch (err) {
      return {
        stdout: err.stdout ?? "",
        stderr: err.stderr ?? err.message ?? "",
        exitCode: err.returncode ?? err.exitCode ?? err.code ?? 1
      };
    }
  }
  return {
    async exec(cmd, args, opts) {
      const command = `${envPrefix(opts?.env)}${shellEscape(cmd, args)}`;
      if (opts?.detached) {
        run(`nohup ${command} > /dev/null 2>&1 &`).catch(() => {
        });
        return { stdout: "", stderr: "", exitCode: 0 };
      }
      return run(command);
    },
    async writeFile(path, content) {
      const buf = Buffer.isBuffer(content) ? content : Buffer.from(content);
      const b64 = buf.toString("base64");
      const result = await run(
        `printf '%s' '${b64}' | base64 -d > '${path.replace(/'/g, "'\\''")}'`
      );
      if (result.exitCode !== 0) {
        throw new Error(
          `writeFile failed (exit ${result.exitCode}): ${result.stderr}`
        );
      }
    },
    async readFile(path) {
      const result = await run(`cat '${path.replace(/'/g, "'\\''")}'`);
      if (result.exitCode !== 0) {
        throw new Error(
          `readFile failed (exit ${result.exitCode}): ${result.stderr}`
        );
      }
      return result.stdout;
    },
    async stop() {
      await sandbox.terminate?.();
    }
  };
}
function modalDefaults() {
  const serverConfig = {
    grpc: { addr: "127.0.0.1:9090" },
    logging: { level: "info", format: "text", output: "stderr" },
    sessions: {
      defaultTimeout: "1h",
      idleTimeout: "15m",
      cleanupInterval: "5m"
    },
    audit: { enabled: true, sqlitePath: "/var/lib/agentsh/events.db" },
    sandboxLimits: { maxMemoryMb: 4096, maxCpuPercent: 100, maxProcesses: 256 },
    allowDegraded: true,
    fuse: {
      deferred: true,
      deferredMarkerFile: "/tmp/.agentsh-fuse-enabled",
      deferredEnableCommand: ["/bin/chmod", "666", "/dev/fuse"]
    },
    networkIntercept: { interceptMode: "all", proxyListenAddr: "127.0.0.1:0" },
    cgroups: { enabled: false },
    unixSockets: { enabled: false },
    ptrace: {
      enabled: true,
      attachMode: "children",
      maskTracerPid: "off",
      trace: {
        execve: true,
        file: false,
        network: true,
        signal: true
      },
      performance: {
        seccompPrefilter: false,
        maxTracees: 500,
        maxHoldMs: 5e3
      },
      onAttachFailure: "fail_open"
    },
    envInject: {
      BASH_ENV: "/usr/lib/agentsh/bash_startup.sh"
    },
    dlp: {
      mode: "redact",
      patterns: { api_keys: true, credit_card: true, ssn: true }
    },
    approvals: { enabled: false },
    metrics: { enabled: true, path: "/metrics" },
    health: { path: "/health", readinessPath: "/ready" }
  };
  const policy = mergePrepend(agentDefault(), {
    file: [
      // System shared libraries (required for any dynamically linked binary)
      { allow: ["/lib/**", "/lib64/**", "/usr/lib/**", "/usr/lib64/**"], ops: ["read"] },
      // System binaries
      { allow: ["/bin/**", "/sbin/**", "/usr/bin/**", "/usr/sbin/**", "/usr/local/bin/**"], ops: ["read"] },
      // Linker config
      { allow: ["/etc/ld.so.cache", "/etc/ld.so.conf", "/etc/ld.so.conf.d/**"], ops: ["read"] },
      // Shared data (locales, terminfo, mime types)
      { allow: "/usr/share/**", ops: ["read"] },
      // Device files
      { allow: ["/dev/null", "/dev/zero", "/dev/urandom", "/dev/random", "/dev/fd/**"], ops: ["read", "write"] },
      // /proc/self for process introspection
      { allow: "/proc/self/**", ops: ["read"] },
      // Temp directories
      { allow: ["/tmp/**", "/var/tmp/**"], ops: ["read", "write", "create"] },
      // System config needed by tools (SSL certs, resolv.conf, hosts)
      { allow: ["/etc/ssl/**", "/etc/ca-certificates/**", "/etc/resolv.conf", "/etc/hosts", "/etc/hostname", "/etc/nsswitch.conf", "/etc/passwd", "/etc/group"], ops: ["read"] }
    ]
  });
  return {
    policy,
    installStrategy: "download",
    realPaths: true,
    skipShim: true,
    serverConfig
  };
}

export {
  sprites,
  spritesDefaults,
  modal,
  modalDefaults,
  adapters_exports
};
