// @agentsh/secure-sandbox/dist/chunk-4FJHYLAB.js
import {
  PolicyDefinitionSchema,
  agentDefault,
  agentSandbox,
  ciStrict,
  devSafe,
  merge,
  mergePrepend,
  validatePolicy
} from "./chunk-LNDICGZU.js";
import {
  __export
} from "./chunk-PZ5AY32C.js";

// src/policies/index.ts
var policies_exports = {};
__export(policies_exports, {
  PolicyDefinitionSchema: () => PolicyDefinitionSchema,
  agentDefault: () => agentDefault,
  agentSandbox: () => agentSandbox,
  ciStrict: () => ciStrict,
  devSafe: () => devSafe,
  merge: () => merge,
  mergePrepend: () => mergePrepend,
  serializePolicy: () => serializePolicy,
  systemPolicyYaml: () => systemPolicyYaml,
  validatePolicy: () => validatePolicy
});

// src/policies/serialize.ts
import yaml from "js-yaml";
function toArray(value) {
  return Array.isArray(value) ? value : [value];
}
var FILE_DECISION_KEYS = [
  "allow",
  "deny",
  "redirect",
  "audit",
  "softDelete"
];
var SIMPLE_DECISION_KEYS = ["allow", "deny", "redirect"];
function findDecision(rule, keys) {
  for (const k of keys) {
    if (k in rule) {
      return { key: k, value: rule[k] };
    }
  }
  throw new Error(`No decision key found in rule: ${JSON.stringify(rule)}`);
}
function yamlDecision(key) {
  return key === "softDelete" ? "soft_delete" : key;
}
function serializeFileRules(rules) {
  return rules.map((rule, i) => {
    const r = rule;
    const { key, value } = findDecision(r, FILE_DECISION_KEYS);
    const paths = toArray(value);
    const out = {
      name: `file-rule-${i}`,
      paths
    };
    if ("ops" in r && r.ops) {
      out.operations = r.ops;
    }
    out.decision = yamlDecision(key);
    if (key === "redirect" && "to" in r) {
      out.redirect_to = r.to;
    }
    return out;
  });
}
function serializeNetworkRules(rules) {
  return rules.map((rule, i) => {
    const r = rule;
    const { key, value } = findDecision(r, SIMPLE_DECISION_KEYS);
    const domains = toArray(value);
    const out = {
      name: `network-rule-${i}`,
      domains,
      decision: key
    };
    if ("ports" in r && r.ports) {
      out.ports = r.ports;
    }
    if (key === "redirect" && "to" in r) {
      out.redirect_to = r.to;
    }
    return out;
  });
}
function serializeCommandRules(rules) {
  return rules.map((rule, i) => {
    const r = rule;
    const { key, value } = findDecision(r, SIMPLE_DECISION_KEYS);
    const commands = toArray(value);
    const out = {
      name: `command-rule-${i}`,
      commands,
      decision: key
    };
    if (key === "redirect" && "to" in r) {
      const to = r.to;
      if (typeof to === "string") {
        out.redirect_to = to;
      } else if (typeof to === "object" && to !== null) {
        const target = to;
        out.redirect_to = { command: target.cmd, args: target.args };
      }
    }
    return out;
  });
}
function serializeEnvRules(rules) {
  return rules.map((rule, i) => {
    const out = {
      name: `env-rule-${i}`,
      commands: rule.commands
    };
    if (rule.allow) {
      out.allow = rule.allow;
    }
    if (rule.deny) {
      out.deny = rule.deny;
    }
    return out;
  });
}
function serializeDnsRedirects(redirects) {
  return redirects.map((r) => ({
    match: r.match,
    resolve_to: r.resolveTo
  }));
}
function serializeConnectRedirects(redirects) {
  return redirects.map((r) => ({
    match: r.match,
    redirect_to: r.redirectTo
  }));
}
function serializePackageRules(rules) {
  return rules.map((rule) => {
    const match = {};
    if (rule.match.packages) {
      match.packages = rule.match.packages;
    }
    if (rule.match.namePatterns) {
      match.name_patterns = rule.match.namePatterns;
    }
    if (rule.match.findingType) {
      match.finding_type = rule.match.findingType;
    }
    if (rule.match.severity !== void 0) {
      match.severity = rule.match.severity;
    }
    if (rule.match.reasons) {
      match.reasons = rule.match.reasons;
    }
    if (rule.match.licenseSpdx) {
      match.license_spdx = rule.match.licenseSpdx;
    }
    if (rule.match.ecosystem) {
      match.ecosystem = rule.match.ecosystem;
    }
    if (rule.match.options) {
      match.options = rule.match.options;
    }
    const out = {
      match,
      action: rule.action
    };
    if (rule.reason) {
      out.reason = rule.reason;
    }
    return out;
  });
}
function serializePolicy(policy) {
  const doc = {
    version: 1,
    name: "secure-sandbox-policy"
  };
  if (policy.file && policy.file.length > 0) {
    doc.file_rules = serializeFileRules(policy.file);
  }
  if (policy.network && policy.network.length > 0) {
    doc.network_rules = serializeNetworkRules(policy.network);
  }
  if (policy.commands && policy.commands.length > 0) {
    doc.command_rules = serializeCommandRules(policy.commands);
  }
  if (policy.env && policy.env.length > 0) {
    doc.env_rules = serializeEnvRules(policy.env);
  }
  if (policy.dns && policy.dns.length > 0) {
    doc.dns_redirects = serializeDnsRedirects(policy.dns);
  }
  if (policy.connect && policy.connect.length > 0) {
    doc.connect_redirects = serializeConnectRedirects(policy.connect);
  }
  if (policy.packageRules && policy.packageRules.length > 0) {
    doc.package_rules = serializePackageRules(policy.packageRules);
  }
  return yaml.dump(doc, { lineWidth: -1 });
}
function systemPolicyYaml() {
  const doc = {
    version: 1,
    name: "_system-protection",
    file_rules: [
      {
        name: "_system-protect-config",
        paths: ["/etc/agentsh/**"],
        operations: ["write", "create", "delete"],
        decision: "deny",
        message: "Policy files are immutable during agent execution"
      },
      {
        name: "_system-protect-binary",
        paths: ["/usr/local/bin/agentsh*", "/usr/bin/agentsh*"],
        operations: ["write", "create", "delete"],
        decision: "deny",
        message: "agentsh binary is immutable during agent execution"
      },
      {
        name: "_system-protect-shim-files",
        paths: ["/usr/bin/agentsh-shell-shim", "/bin/bash", "/bin/sh"],
        operations: ["write", "create", "delete"],
        decision: "deny",
        message: "Shell and shim binaries are immutable during agent execution"
      }
    ],
    command_rules: [
      {
        name: "_system-protect-process",
        commands: ["kill", "killall", "pkill"],
        args_match: ["agentsh"],
        decision: "deny",
        message: "Cannot terminate agentsh processes"
      }
    ]
  };
  return yaml.dump(doc, { lineWidth: -1 });
}

export {
  serializePolicy,
  systemPolicyYaml,
  policies_exports
};
