// @a5c-ai/babysitter-sdk/dist/compression/config.js
"use strict";
/**
 * CompressionConfig — type definitions and defaults for the feature-toggle system.
 *
 * Priority (highest wins): env vars > project config > user config > defaults.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_COMPRESSION_CONFIG = void 0;
exports.DEFAULT_COMPRESSION_CONFIG = {
    enabled: true,
    layers: {
        userPromptHook: {
            enabled: true,
            engine: 'density-filter',
            threshold: 500,
            keepRatio: 0.78,
        },
        commandOutputHook: {
            enabled: true,
            engine: 'command-compressor',
            excludeCommands: ['node', 'python', 'ruby'],
        },
        sdkContextHook: {
            enabled: true,
            engine: 'sentence-extractor',
            targetReduction: 0.15,
            minCompressionTokens: 150,
            perTaskKind: {
                agent: 0.15,
                skill: 0.2,
                breakpoint: 0.1,
            },
        },
        processLibraryCache: {
            enabled: true,
            engine: 'sentence-extractor',
            targetReduction: 0.35,
            ttlHours: 24,
        },
    },
};
