// @a5c-ai/babysitter-sdk/dist/compression/config-loader.js
"use strict";
/**
 * config-loader — loads and merges CompressionConfig from multiple sources.
 *
 * Priority (highest wins):
 *   env vars  >  project config (.a5c/compression.config.json)
 *             >  user config (~/.a5c/compression.config.json)
 *             >  built-in defaults
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMPRESSION_ENV_VARS = void 0;
exports.loadCompressionConfig = loadCompressionConfig;
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
const os = __importStar(require("node:os"));
const config_1 = require("./config");
exports.COMPRESSION_ENV_VARS = {
    ENABLED: 'BABYSITTER_COMPRESSION_ENABLED',
    USER_PROMPT: 'BABYSITTER_COMPRESSION_USER_PROMPT',
    COMMANDS: 'BABYSITTER_COMPRESSION_COMMANDS',
    SDK_CONTEXT: 'BABYSITTER_COMPRESSION_SDK_CONTEXT',
    LIBRARY_CACHE: 'BABYSITTER_COMPRESSION_LIBRARY_CACHE',
};
function parseBool(value) {
    if (value === undefined)
        return undefined;
    const v = value.trim().toLowerCase();
    if (v === '1' || v === 'true' || v === 'yes')
        return true;
    if (v === '0' || v === 'false' || v === 'no')
        return false;
    return undefined;
}
function readJsonFile(filePath) {
    try {
        if (!(0, node_fs_1.existsSync)(filePath))
            return undefined;
        const raw = (0, node_fs_1.readFileSync)(filePath, 'utf8');
        return JSON.parse(raw);
    }
    catch {
        return undefined;
    }
}
function deepMerge(base, override) {
    if (typeof base !== 'object' ||
        base === null ||
        typeof override !== 'object' ||
        override === null) {
        return override ?? base;
    }
    const result = { ...base };
    for (const key of Object.keys(override)) {
        const overrideVal = override[key];
        const baseVal = base[key];
        if (overrideVal !== null &&
            typeof overrideVal === 'object' &&
            !Array.isArray(overrideVal) &&
            baseVal !== null &&
            typeof baseVal === 'object' &&
            !Array.isArray(baseVal)) {
            result[key] = deepMerge(baseVal, overrideVal);
        }
        else if (overrideVal !== undefined) {
            result[key] = overrideVal;
        }
    }
    return result;
}
function loadFileConfig(dir) {
    const filePath = path.join(dir, '.a5c', 'compression.config.json');
    const parsed = readJsonFile(filePath);
    if (parsed === undefined || typeof parsed !== 'object' || parsed === null) {
        return {};
    }
    return parsed;
}
function applyEnvVars(config) {
    let result = { ...config };
    const masterEnabled = parseBool(process.env[exports.COMPRESSION_ENV_VARS.ENABLED]);
    if (masterEnabled !== undefined) {
        result = { ...result, enabled: masterEnabled };
        if (!masterEnabled) {
            result = {
                ...result,
                layers: {
                    userPromptHook: { ...result.layers.userPromptHook, enabled: false },
                    commandOutputHook: { ...result.layers.commandOutputHook, enabled: false },
                    sdkContextHook: { ...result.layers.sdkContextHook, enabled: false },
                    processLibraryCache: { ...result.layers.processLibraryCache, enabled: false },
                },
            };
        }
    }
    if (result.enabled) {
        const userPrompt = parseBool(process.env[exports.COMPRESSION_ENV_VARS.USER_PROMPT]);
        if (userPrompt !== undefined) {
            result = {
                ...result,
                layers: { ...result.layers, userPromptHook: { ...result.layers.userPromptHook, enabled: userPrompt } },
            };
        }
        const commands = parseBool(process.env[exports.COMPRESSION_ENV_VARS.COMMANDS]);
        if (commands !== undefined) {
            result = {
                ...result,
                layers: { ...result.layers, commandOutputHook: { ...result.layers.commandOutputHook, enabled: commands } },
            };
        }
        const sdkContext = parseBool(process.env[exports.COMPRESSION_ENV_VARS.SDK_CONTEXT]);
        if (sdkContext !== undefined) {
            result = {
                ...result,
                layers: { ...result.layers, sdkContextHook: { ...result.layers.sdkContextHook, enabled: sdkContext } },
            };
        }
        const libraryCache = parseBool(process.env[exports.COMPRESSION_ENV_VARS.LIBRARY_CACHE]);
        if (libraryCache !== undefined) {
            result = {
                ...result,
                layers: { ...result.layers, processLibraryCache: { ...result.layers.processLibraryCache, enabled: libraryCache } },
            };
        }
    }
    return result;
}
/**
 * Load the effective CompressionConfig by merging all sources.
 *
 * @param projectDir  Root of the project (directory that contains .a5c/).
 *                    Defaults to `process.cwd()`.
 */
function loadCompressionConfig(projectDir) {
    const cwd = projectDir ?? process.cwd();
    let config = JSON.parse(JSON.stringify(config_1.DEFAULT_COMPRESSION_CONFIG));
    const userConfig = loadFileConfig(os.homedir());
    if (Object.keys(userConfig).length > 0) {
        config = deepMerge(config, userConfig);
    }
    const projectConfig = loadFileConfig(cwd);
    if (Object.keys(projectConfig).length > 0) {
        config = deepMerge(config, projectConfig);
    }
    config = applyEnvVars(config);
    return config;
}
