// @a5c-ai/babysitter-sdk/dist/compression/library-cache.js
"use strict";
/**
 * library-cache — compress-and-cache process library files (processLibraryCache layer).
 *
 * On first access (or after TTL expiry) a file is read, compressed with
 * densityFilterText, and written to .a5c/cache/compression/<hash>.json.
 * Subsequent reads within TTL are served from cache with zero re-compression cost.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOrCompressFile = getOrCompressFile;
exports.findLibraryFiles = findLibraryFiles;
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
const crypto = __importStar(require("node:crypto"));
const density_filter_1 = require("./density-filter");
function cacheEntryPath(cacheDir, filePath) {
    const hash = crypto.createHash('sha256').update(path.resolve(filePath)).digest('hex').slice(0, 16);
    return path.join(cacheDir, `${hash}.json`);
}
function isCacheExpired(compressedAt, ttlHours) {
    return Date.now() - new Date(compressedAt).getTime() > ttlHours * 3_600_000;
}
function readCache(cacheDir, filePath, ttlHours) {
    const entryPath = cacheEntryPath(cacheDir, filePath);
    if (!(0, node_fs_1.existsSync)(entryPath))
        return null;
    try {
        const entry = JSON.parse((0, node_fs_1.readFileSync)(entryPath, 'utf8'));
        if (isCacheExpired(entry.compressedAt, ttlHours))
            return null;
        return entry.compressedContent;
    }
    catch {
        return null;
    }
}
function writeCache(cacheDir, filePath, entry) {
    try {
        (0, node_fs_1.mkdirSync)(cacheDir, { recursive: true });
        (0, node_fs_1.writeFileSync)(cacheEntryPath(cacheDir, filePath), JSON.stringify(entry, null, 2), 'utf8');
    }
    catch {
        // Best-effort
    }
}
/**
 * Return compressed content for `filePath`, reading from cache when fresh.
 * On cache miss: reads, compresses, writes to cache, returns compressed content.
 * Returns null if the file doesn't exist or can't be read.
 */
function getOrCompressFile(filePath, targetReduction, ttlHours, cacheDir) {
    const resolved = path.resolve(filePath);
    if (!(0, node_fs_1.existsSync)(resolved))
        return null;
    const cached = readCache(cacheDir, resolved, ttlHours);
    if (cached !== null)
        return cached;
    let content;
    try {
        content = (0, node_fs_1.readFileSync)(resolved, 'utf8');
    }
    catch {
        return null;
    }
    const originalTokens = (0, density_filter_1.estimateTokens)(content);
    const compressedContent = originalTokens > 50 ? (0, density_filter_1.densityFilterText)(content, targetReduction) : content;
    writeCache(cacheDir, resolved, {
        filePath: resolved,
        compressedContent,
        originalTokens,
        compressedTokens: (0, density_filter_1.estimateTokens)(compressedContent),
        compressedAt: new Date().toISOString(),
    });
    return compressedContent;
}
/**
 * Find all SKILL.md and AGENT.md files under `rootDir` (max depth 6).
 * Used by the session-start hook to pre-warm the cache.
 */
function findLibraryFiles(rootDir) {
    const results = [];
    const LIBRARY_FILENAMES = new Set(['SKILL.md', 'AGENT.md']);
    function scan(dir, depth) {
        if (depth > 6)
            return;
        let names;
        try {
            names = (0, node_fs_1.readdirSync)(dir);
        }
        catch {
            return;
        }
        for (const name of names) {
            if (name.startsWith('.') && name !== '.claude')
                continue;
            if (name === 'node_modules')
                continue;
            const fullPath = path.join(dir, name);
            let isDir = false;
            try {
                isDir = (0, node_fs_1.statSync)(fullPath).isDirectory();
            }
            catch {
                continue;
            }
            if (isDir) {
                scan(fullPath, depth + 1);
            }
            else if (LIBRARY_FILENAMES.has(name)) {
                results.push(fullPath);
            }
        }
    }
    scan(rootDir, 0);
    return results;
}
