// @a5c-ai/babysitter-sdk/dist/processLibrary/active.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultProcessLibrarySpec = getDefaultProcessLibrarySpec;
exports.getActiveProcessLibraryStatePath = getActiveProcessLibraryStatePath;
exports.cloneProcessLibrary = cloneProcessLibrary;
exports.updateProcessLibrary = updateProcessLibrary;
exports.bindActiveProcessLibrary = bindActiveProcessLibrary;
exports.resolveActiveProcessLibrary = resolveActiveProcessLibrary;
exports.ensureActiveProcessLibrary = ensureActiveProcessLibrary;
const node_fs_1 = require("node:fs");
const node_os_1 = __importDefault(require("node:os"));
const node_path_1 = __importDefault(require("node:path"));
const node_child_process_1 = require("node:child_process");
const node_util_1 = require("node:util");
const execFile = (0, node_util_1.promisify)(node_child_process_1.execFile);
const ACTIVE_PROCESS_LIBRARY_FILENAME = "process-library.json";
const DEFAULT_PROCESS_LIBRARY_REPO = "https://github.com/a5c-ai/babysitter.git";
const DEFAULT_PROCESS_LIBRARY_SUBPATH = "library";
const DEFAULT_PROCESS_LIBRARY_REFERENCE_SUBPATH = "library/reference";
function resolveStateDir(stateDir) {
    if (stateDir && stateDir.trim()) {
        return node_path_1.default.resolve(stateDir);
    }
    if (process.env.BABYSITTER_GLOBAL_STATE_DIR?.trim()) {
        return node_path_1.default.resolve(process.env.BABYSITTER_GLOBAL_STATE_DIR);
    }
    return node_path_1.default.join(node_os_1.default.homedir(), ".a5c");
}
function splitProcessLibrarySubpath(value) {
    return value
        .split(/[\\/]+/)
        .map((part) => part.trim())
        .filter(Boolean);
}
async function pathExists(targetPath) {
    try {
        await node_fs_1.promises.access(targetPath);
        return true;
    }
    catch {
        return false;
    }
}
function getDefaultProcessLibrarySpec(options = {}) {
    const stateDir = resolveStateDir(options.stateDir);
    const repo = options.repo?.trim() ||
        process.env.BABYSITTER_PROCESS_LIBRARY_REPO?.trim() ||
        DEFAULT_PROCESS_LIBRARY_REPO;
    const ref = options.ref?.trim() || process.env.BABYSITTER_PROCESS_LIBRARY_REF?.trim() || undefined;
    const cloneDir = node_path_1.default.resolve(options.cloneDir?.trim() || node_path_1.default.join(stateDir, "process-library", "babysitter-repo"));
    const processSubpath = process.env.BABYSITTER_PROCESS_LIBRARY_SUBPATH?.trim() ||
        DEFAULT_PROCESS_LIBRARY_SUBPATH;
    const referenceSubpath = process.env.BABYSITTER_PROCESS_LIBRARY_REFERENCE_SUBPATH?.trim() ||
        DEFAULT_PROCESS_LIBRARY_REFERENCE_SUBPATH;
    const processRoot = node_path_1.default.resolve(options.processDir?.trim() ||
        node_path_1.default.join(cloneDir, ...splitProcessLibrarySubpath(processSubpath)));
    const referenceRoot = options.processDir?.trim()
        ? node_path_1.default.resolve(options.processDir, "reference")
        : node_path_1.default.resolve(cloneDir, ...splitProcessLibrarySubpath(referenceSubpath));
    return {
        stateDir,
        repo,
        ...(ref ? { ref } : {}),
        cloneDir,
        processRoot,
        referenceRoot,
    };
}
function getActiveProcessLibraryStatePath(stateDir) {
    return node_path_1.default.join(resolveStateDir(stateDir), "active", ACTIVE_PROCESS_LIBRARY_FILENAME);
}
async function readState(stateDir) {
    const stateFile = getActiveProcessLibraryStatePath(stateDir);
    try {
        const raw = await node_fs_1.promises.readFile(stateFile, "utf8");
        return JSON.parse(raw);
    }
    catch (error) {
        const code = error?.code;
        if (code === "ENOENT") {
            return {
                version: 1,
                updatedAt: new Date().toISOString(),
                runBindings: {},
                sessionBindings: {},
            };
        }
        throw error;
    }
}
async function writeState(stateDir, state) {
    const stateFile = getActiveProcessLibraryStatePath(stateDir);
    await node_fs_1.promises.mkdir(node_path_1.default.dirname(stateFile), { recursive: true });
    await node_fs_1.promises.writeFile(stateFile, JSON.stringify(state, null, 2), "utf8");
    return stateFile;
}
function normalizeBinding(dir, repoUrl, ref, revision) {
    return {
        dir: node_path_1.default.resolve(dir),
        ...(repoUrl ? { repoUrl } : {}),
        ...(ref ? { ref } : {}),
        ...(revision ? { revision } : {}),
        boundAt: new Date().toISOString(),
    };
}
async function ensureDirectoryExists(dir) {
    const resolved = node_path_1.default.resolve(dir);
    try {
        const stat = await node_fs_1.promises.stat(resolved);
        if (!stat.isDirectory()) {
            throw new Error(`Process-library path is not a directory: ${resolved}`);
        }
    }
    catch (error) {
        const code = error?.code;
        if (code === "ENOENT") {
            throw new Error(`Process-library directory not found: ${resolved}`);
        }
        throw error;
    }
    return resolved;
}
async function runGit(args, cwd) {
    const gitArgs = process.platform === "win32"
        ? ["-c", "core.longpaths=true", ...args]
        : args;
    return execFile("git", gitArgs, {
        ...(cwd ? { cwd } : {}),
        encoding: "utf8",
    });
}
async function getGitOutput(args, cwd) {
    try {
        const { stdout } = await runGit(args, cwd);
        const value = stdout.trim();
        return value || undefined;
    }
    catch {
        return undefined;
    }
}
async function readGitMetadata(dir) {
    const repoUrl = await getGitOutput(["config", "--get", "remote.origin.url"], dir);
    const revision = await getGitOutput(["rev-parse", "HEAD"], dir);
    return { repoUrl, revision };
}
async function cloneProcessLibrary(options) {
    const dir = node_path_1.default.resolve(options.dir);
    await node_fs_1.promises.mkdir(node_path_1.default.dirname(dir), { recursive: true });
    if (await pathExists(dir)) {
        throw new Error(`Process-library directory already exists at ${dir}. Choose a new --dir or update it instead.`);
    }
    const cloneArgs = ["clone", "--depth", "1"];
    if (options.ref) {
        cloneArgs.push("--branch", options.ref);
    }
    cloneArgs.push(options.repo, dir);
    try {
        await runGit(cloneArgs);
    }
    catch (error) {
        await node_fs_1.promises.rm(dir, { recursive: true, force: true });
        throw error;
    }
    const { revision } = await readGitMetadata(dir);
    if (!revision) {
        throw new Error(`Unable to determine cloned revision for ${dir}`);
    }
    return {
        dir,
        repo: options.repo,
        ...(options.ref ? { ref: options.ref } : {}),
        revision,
    };
}
async function updateProcessLibrary(options) {
    const dir = await ensureDirectoryExists(options.dir);
    if (options.ref) {
        await runGit(["fetch", "--depth", "1", "origin", options.ref], dir);
        await runGit(["checkout", "--force", "FETCH_HEAD"], dir);
    }
    else {
        await runGit(["pull", "--ff-only"], dir);
    }
    const { repoUrl, revision } = await readGitMetadata(dir);
    if (!revision) {
        throw new Error(`Unable to determine updated revision for ${dir}`);
    }
    return {
        dir,
        ...(repoUrl ? { repo: repoUrl } : {}),
        ...(options.ref ? { ref: options.ref } : {}),
        revision,
    };
}
async function bindActiveProcessLibrary(options) {
    const dir = await ensureDirectoryExists(options.dir);
    const state = await readState(options.stateDir);
    const { repoUrl, revision } = await readGitMetadata(dir);
    const binding = normalizeBinding(dir, repoUrl, options.ref, revision);
    let bindingScope = "default";
    let bindingKey;
    if (options.runId) {
        state.runBindings = state.runBindings ?? {};
        state.runBindings[options.runId] = binding;
        bindingScope = "run";
        bindingKey = options.runId;
    }
    if (options.sessionId) {
        state.sessionBindings = state.sessionBindings ?? {};
        state.sessionBindings[options.sessionId] = binding;
        bindingScope = options.runId ? "run+session" : "session";
        bindingKey = options.runId
            ? `${options.runId}:${options.sessionId}`
            : options.sessionId;
    }
    if (!options.runId && !options.sessionId) {
        state.defaultBinding = binding;
    }
    state.updatedAt = new Date().toISOString();
    const stateFile = await writeState(options.stateDir, state);
    return { stateFile, bindingScope, ...(bindingKey ? { bindingKey } : {}), binding };
}
async function resolveActiveProcessLibrary(options = {}) {
    const stateFile = getActiveProcessLibraryStatePath(options.stateDir);
    const state = await readState(options.stateDir);
    if (options.runId && state.runBindings?.[options.runId]) {
        return {
            stateFile,
            bindingScope: "run",
            bindingKey: options.runId,
            binding: state.runBindings[options.runId],
        };
    }
    if (options.sessionId && state.sessionBindings?.[options.sessionId]) {
        return {
            stateFile,
            bindingScope: "session",
            bindingKey: options.sessionId,
            binding: state.sessionBindings[options.sessionId],
        };
    }
    return {
        stateFile,
        bindingScope: state.defaultBinding ? "default" : null,
        binding: state.defaultBinding ?? null,
    };
}
async function ensureActiveProcessLibrary(options = {}) {
    const defaultSpec = getDefaultProcessLibrarySpec(options);
    const resolved = await resolveActiveProcessLibrary(options);
    if (resolved.binding && (await pathExists(resolved.binding.dir))) {
        return {
            ...resolved,
            bootstrapped: false,
            defaultSpec,
        };
    }
    const cloneExists = await pathExists(node_path_1.default.join(defaultSpec.cloneDir, ".git"));
    if (cloneExists) {
        await updateProcessLibrary({
            dir: defaultSpec.cloneDir,
            ...(defaultSpec.ref ? { ref: defaultSpec.ref } : {}),
        });
    }
    else {
        await cloneProcessLibrary({
            repo: defaultSpec.repo,
            dir: defaultSpec.cloneDir,
            ...(defaultSpec.ref ? { ref: defaultSpec.ref } : {}),
        });
    }
    const bound = await bindActiveProcessLibrary({
        stateDir: defaultSpec.stateDir,
        dir: defaultSpec.processRoot,
        runId: options.runId,
        sessionId: options.sessionId,
        ref: defaultSpec.ref,
    });
    return {
        stateFile: bound.stateFile,
        bindingScope: bound.bindingScope === "run+session" ? "run" : bound.bindingScope,
        ...(bound.bindingKey ? { bindingKey: bound.bindingKey } : {}),
        binding: bound.binding,
        bootstrapped: true,
        defaultSpec,
    };
}
