// @a5c-ai/babysitter-sdk/dist/logging/runLogger.js
"use strict";
/**
 * Structured run logger.
 *
 * Provides a common logging facility that writes structured log lines to
 * contextual log files under `~/.a5c/logs/`.  Each log type maps to a
 * specific file name within the run directory or the global log root:
 *
 *   - `process` → `<runId>/process.log`   (ctx.log in process definitions)
 *   - `hook`    → `<runId>/hooks.log`     (hook execution, per-run)
 *   - `hook`    → `hooks.log`             (hook execution, no runId)
 *   - `cli`     → `cli.log`              (CLI command operations)
 *
 * Log lines are append-only, one JSON object per line (JSONL), making them
 * easy to parse, grep, and stream.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultLogDir = getDefaultLogDir;
exports.resolveLogPath = resolveLogPath;
exports.getRunLogPath = getRunLogPath;
exports.getGlobalLogPath = getGlobalLogPath;
exports.formatLogLine = formatLogLine;
exports.appendRunLog = appendRunLog;
exports.createRunLogger = createRunLogger;
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
const os = __importStar(require("node:os"));
// ---------------------------------------------------------------------------
// Paths
// ---------------------------------------------------------------------------
const LOG_DIR_ENV = "BABYSITTER_LOG_DIR";
function getDefaultLogDir() {
    return process.env[LOG_DIR_ENV] || path.join(os.homedir(), ".a5c", "logs");
}
/** Log file names by type. */
const LOG_FILE_NAMES = {
    process: "process.log",
    hook: "hooks.log",
    cli: "cli.log",
};
/**
 * Resolve the log file path for a given entry.
 *
 * - With `runId`: `<logDir>/<runId>/<typeName>.log`
 * - Without `runId`: `<logDir>/<typeName>.log`
 */
function resolveLogPath(logDir, logType, runId) {
    const fileName = LOG_FILE_NAMES[logType];
    if (runId) {
        return path.join(logDir, runId, fileName);
    }
    return path.join(logDir, fileName);
}
/** @deprecated Use resolveLogPath instead. Kept for backwards compat. */
function getRunLogPath(logDir, runId) {
    return resolveLogPath(logDir, "process", runId);
}
/** @deprecated Use resolveLogPath instead. Kept for backwards compat. */
function getGlobalLogPath(logDir) {
    return path.join(logDir, "hooks.log");
}
// ---------------------------------------------------------------------------
// Formatting
// ---------------------------------------------------------------------------
function formatLogLine(entry) {
    const record = {
        ts: entry.timestamp,
        level: entry.level,
    };
    if (entry.type)
        record.type = entry.type;
    if (entry.label)
        record.label = entry.label;
    record.msg = entry.message;
    if (entry.runId)
        record.runId = entry.runId;
    if (entry.processId)
        record.processId = entry.processId;
    if (entry.source)
        record.source = entry.source;
    if (entry.context && Object.keys(entry.context).length > 0) {
        record.ctx = entry.context;
    }
    return JSON.stringify(record) + "\n";
}
// ---------------------------------------------------------------------------
// Core append
// ---------------------------------------------------------------------------
/**
 * Append a single structured log entry to the appropriate log file.
 *
 * File selection uses the entry's `type` field (defaults to "process"):
 * - With `runId`: `<logDir>/<runId>/<type>.log`
 * - Without `runId`: `<logDir>/<type>.log`
 *
 * Creates parent directories on first write.  Returns the resolved log
 * file path.
 */
async function appendRunLog(entry, options) {
    const logDir = options?.logDir || getDefaultLogDir();
    const logType = entry.type || "process";
    const logPath = resolveLogPath(logDir, logType, entry.runId);
    await node_fs_1.promises.mkdir(path.dirname(logPath), { recursive: true });
    const line = formatLogLine(entry);
    await node_fs_1.promises.appendFile(logPath, line, "utf8");
    return logPath;
}
// ---------------------------------------------------------------------------
// Logger factory
// ---------------------------------------------------------------------------
/**
 * Create a bound logger for a specific run.
 *
 * The returned object provides level-specific methods (`info`, `warn`, etc.)
 * that automatically populate `runId`, `processId`, `source`, and `type` on
 * every entry.  Each method is fire-and-forget: errors are swallowed so
 * logging never breaks orchestration.
 */
function createRunLogger(options) {
    const { runId, processId, logDir, source, type: defaultType } = options;
    let pending = Promise.resolve();
    function write(level, label, message, context) {
        pending = pending.then(() => appendRunLog({
            timestamp: new Date().toISOString(),
            level,
            type: defaultType,
            label: label || undefined,
            message,
            runId,
            processId,
            source,
            context,
        }, { logDir }).then(() => undefined)).catch(() => {
            // Never let logging break orchestration.
        });
    }
    return {
        debug: (label, message, context) => write("debug", label, message, context),
        info: (label, message, context) => write("info", label, message, context),
        warn: (label, message, context) => write("warn", label, message, context),
        error: (label, message, context) => write("error", label, message, context),
        /** Raw write — lets callers specify the level explicitly. */
        log: write,
        /** Returns a promise that resolves when all queued writes have completed. */
        flush: () => pending,
    };
}
