// @a5c-ai/babysitter-sdk/dist/harness/codex.js
"use strict";
/**
 * Codex harness adapter.
 *
 * Codex supports lifecycle hook callbacks (SessionStart/Stop/UserPromptSubmit)
 * on hook-capable installs. This adapter keeps the implementation honest:
 * use Codex hook payload identity and project state as the continuation source
 * of truth, with explicit fallback binding only when needed.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCodexAdapter = createCodexAdapter;
const path = __importStar(require("node:path"));
const node_stream_1 = require("node:stream");
const config_1 = require("../config");
const claudeCode_1 = require("./claudeCode");
const session_1 = require("../session");
const context_1 = require("../prompts/context");
const installSupport_1 = require("./installSupport");
function resolveCodexPluginRoot(args = {}) {
    const root = args.pluginRoot || process.env.CODEX_PLUGIN_ROOT;
    return root ? path.resolve(root) : undefined;
}
function resolveCodexStateDir(args) {
    if (args.stateDir)
        return path.resolve(args.stateDir);
    return (0, config_1.getGlobalStateDir)();
}
function resolveCodexSessionId(parsed) {
    if (parsed.sessionId)
        return parsed.sessionId;
    // Cross-harness standard
    if (process.env.BABYSITTER_SESSION_ID)
        return process.env.BABYSITTER_SESSION_ID;
    // Codex-native env vars (auto-injected by Codex CLI into all hooks)
    if (process.env.CODEX_THREAD_ID)
        return process.env.CODEX_THREAD_ID;
    if (process.env.CODEX_SESSION_ID)
        return process.env.CODEX_SESSION_ID;
    return undefined;
}
function readStdin() {
    return new Promise((resolve, reject) => {
        let data = "";
        process.stdin.setEncoding("utf8");
        process.stdin.on("data", (chunk) => {
            data += chunk;
        });
        process.stdin.on("end", () => resolve(data));
        process.stdin.on("error", reject);
    });
}
function parseHookInput(raw) {
    const trimmed = raw.trim();
    if (!trimmed)
        return {};
    try {
        const parsed = JSON.parse(trimmed);
        if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
            return parsed;
        }
    }
    catch {
        // Treat malformed JSON as empty input.
    }
    return {};
}
function firstString(obj, keys) {
    for (const key of keys) {
        const value = obj[key];
        if (typeof value === "string" && value.trim()) {
            return value;
        }
    }
    return undefined;
}
function normalizeCodexHookInput(raw) {
    const parsed = parseHookInput(raw);
    const sessionId = firstString(parsed, [
        "session_id",
        "sessionId",
        "thread_id",
        "threadId",
        "conversation_id",
        "conversationId",
    ]) || resolveCodexSessionId({});
    const transcriptPath = firstString(parsed, [
        "transcript_path",
        "transcriptPath",
    ]);
    const lastAssistantMessage = firstString(parsed, [
        "last_assistant_message",
        "lastAssistantMessage",
        "assistant_message",
        "assistantMessage",
    ]);
    return {
        ...parsed,
        ...(sessionId ? { session_id: sessionId } : {}),
        ...(transcriptPath ? { transcript_path: transcriptPath } : {}),
        ...(lastAssistantMessage
            ? { last_assistant_message: lastAssistantMessage }
            : {}),
    };
}
async function withSyntheticStdin(payload, fn) {
    const originalStdin = process.stdin;
    const fakeStdin = node_stream_1.Readable.from([payload], { encoding: "utf8" });
    fakeStdin.unref = () => { };
    Object.defineProperty(process, "stdin", {
        value: fakeStdin,
        writable: true,
        configurable: true,
    });
    try {
        return await fn();
    }
    finally {
        Object.defineProperty(process, "stdin", {
            value: originalStdin,
            writable: true,
            configurable: true,
        });
    }
}
async function handleCodexSessionStartHookImpl(args) {
    const { verbose } = args;
    let rawInput = "";
    try {
        rawInput = await readStdin();
    }
    catch {
        process.stdout.write("{}\n");
        return 0;
    }
    finally {
        if (typeof process.stdin.unref === "function") {
            process.stdin.unref();
        }
    }
    const normalized = normalizeCodexHookInput(rawInput);
    const sessionId = firstString(normalized, ["session_id"]);
    if (!sessionId) {
        process.stdout.write("{}\n");
        return 0;
    }
    // Codex auto-injects CODEX_THREAD_ID into all hooks — no env file write needed.
    const stateDir = resolveCodexStateDir({
        stateDir: args.stateDir,
        pluginRoot: args.pluginRoot,
    });
    const filePath = (0, session_1.getSessionFilePath)(stateDir, sessionId);
    try {
        if (!(await (0, session_1.sessionFileExists)(filePath))) {
            const nowTs = (0, session_1.getCurrentTimestamp)();
            const state = {
                active: true,
                iteration: 1,
                maxIterations: 256,
                runId: "",
                startedAt: nowTs,
                lastIterationAt: nowTs,
                iterationTimes: [],
            };
            await (0, session_1.writeSessionFile)(filePath, state, "");
            if (verbose) {
                process.stderr.write(`[hook:run session-start] Created Codex session state: ${filePath}\n`);
            }
        }
    }
    catch {
        if (verbose) {
            process.stderr.write(`[hook:run session-start] Failed to create session state in ${stateDir}\n`);
        }
    }
    process.stdout.write("{}\n");
    return 0;
}
async function runCodexWorkspaceOnboarding(options) {
    const workspace = path.resolve(options.workspace ?? process.cwd());
    return (0, installSupport_1.runPackageBinaryViaNpx)({
        harness: "codex",
        packageName: "@a5c-ai/babysitter-codex",
        packageArgs: ["install", "--workspace", workspace],
        summary: "Run the published Babysitter Codex workspace installer for the target repo.",
        options,
        cwd: workspace,
        env: process.env,
        location: workspace,
    });
}
async function installCodexHarness(options) {
    return (0, installSupport_1.installCliViaNpm)({
        harness: "codex",
        cliCommand: "codex",
        packageName: "@openai/codex@latest",
        summary: "Install the Codex CLI globally via npm.",
        options,
    });
}
async function installCodexPlugin(options) {
    const installedSkillDir = (0, installSupport_1.getInstalledCodexSkillDir)();
    if ((0, installSupport_1.isCodexPluginInstalled)()) {
        if (options.workspace) {
            return runCodexWorkspaceOnboarding(options);
        }
        return {
            harness: "codex",
            warning: "babysit is already installed in CODEX_HOME; skipping reinstall.",
            location: installedSkillDir,
        };
    }
    const globalInstall = await (0, installSupport_1.runPackageBinaryViaNpx)({
        harness: "codex",
        packageName: "@a5c-ai/babysitter-codex",
        packageArgs: ["install", "--global"],
        summary: "Install the published Babysitter Codex package and materialize the global Codex skill/hooks/config.",
        options,
        env: process.env,
        location: installedSkillDir,
    });
    if (options.workspace) {
        const onboarding = await runCodexWorkspaceOnboarding({
            ...options,
            dryRun: false,
        });
        return {
            harness: "codex",
            summary: "Ran the published Babysitter Codex installer for global Codex setup and then the published workspace installer for the target repo.",
            location: onboarding.location ?? installedSkillDir,
            output: [
                globalInstall.output?.trim() ?? "",
                onboarding.output?.trim() ?? "",
            ].filter(Boolean).join("\n"),
        };
    }
    return {
        harness: "codex",
        summary: "Ran the published Babysitter Codex installer for global Codex skill/hooks/config and the global process-library binding.",
        location: globalInstall.location ?? installedSkillDir,
        output: globalInstall.output,
    };
}
function createCodexAdapter() {
    const claude = (0, claudeCode_1.createClaudeCodeAdapter)();
    return {
        name: "codex",
        isActive() {
            return !!(process.env.BABYSITTER_SESSION_ID ||
                process.env.CODEX_THREAD_ID ||
                process.env.CODEX_SESSION_ID ||
                process.env.CODEX_PLUGIN_ROOT);
        },
        autoResolvesSessionId() {
            return true;
        },
        resolveSessionId(parsed) {
            return resolveCodexSessionId(parsed);
        },
        resolveStateDir(args) {
            return resolveCodexStateDir(args);
        },
        resolvePluginRoot(args) {
            return resolveCodexPluginRoot(args);
        },
        getMissingSessionIdHint() {
            return ("Use --session-id explicitly, or launch through a Codex hook callback " +
                "that provides a stable session/thread ID.");
        },
        supportsHookType(hookType) {
            void hookType;
            return true;
        },
        async bindSession(opts) {
            const stateDir = resolveCodexStateDir({
                stateDir: opts.stateDir,
                pluginRoot: opts.pluginRoot,
            });
            const result = await claude.bindSession({
                ...opts,
                stateDir,
            });
            return {
                ...result,
                harness: "codex",
            };
        },
        async handleStopHook(args) {
            let rawInput = "";
            try {
                rawInput = await readStdin();
            }
            catch {
                rawInput = "";
            }
            const normalized = normalizeCodexHookInput(rawInput);
            return withSyntheticStdin(JSON.stringify(normalized), () => claude.handleStopHook({
                ...args,
                pluginRoot: resolveCodexPluginRoot({ pluginRoot: args.pluginRoot }),
                stateDir: resolveCodexStateDir({
                    stateDir: args.stateDir,
                    pluginRoot: args.pluginRoot,
                }),
            }));
        },
        handleSessionStartHook(args) {
            return handleCodexSessionStartHookImpl(args);
        },
        findHookDispatcherPath(_startCwd) {
            return null;
        },
        installHarness(options) {
            return installCodexHarness(options);
        },
        installPlugin(options) {
            return installCodexPlugin(options);
        },
        getPromptContext(opts) {
            return (0, context_1.createCodexContext)(opts);
        },
    };
}
