// @a5c-ai/babysitter-sdk/dist/harness/cursor.js
"use strict";
/**
 * Cursor IDE/CLI harness adapter.
 *
 * Centralizes all Cursor-specific behaviors:
 *   - Session ID resolution (conversation_id from hook stdin JSON, persisted
 *     by sessionStart hook to a state file)
 *   - State directory conventions (.a5c/state/ by default)
 *   - Plugin root resolution (CURSOR_PLUGIN_ROOT env var)
 *   - Session binding (run:create → state file with run association)
 *   - Stop hook handler (approve/block via followup_message auto-continue)
 *   - Session-start hook handler (baseline state file creation)
 *
 * Cursor Hook Protocol:
 *   - Config: .cursor/hooks.json (project) or ~/.cursor/hooks.json (user)
 *   - Format: { version: 1, hooks: { <hookType>: [{ command: "...", type: "command" }] } }
 *   - Input:  JSON via stdin (fields include conversation_id, project_dir)
 *   - Output: JSON via stdout
 *   - Stop hook: return { followup_message: "..." } to auto-continue (loop_limit
 *     default 5, null for unlimited). Return {} to allow exit.
 *   - Hook types (in headless CLI): sessionStart, stop, postToolUse,
 *     afterFileEdit, afterShellExecution (work in CLI)
 *   - NOT in headless: afterAgentResponse, afterAgentThought (IDE only)
 *
 * Session ID:
 *   Cursor provides conversation_id in the hook stdin JSON, NOT as an env var.
 *   The sessionStart hook persists it to a state file so the stop hook can
 *   find it on subsequent invocations.
 *
 * CLI invocation:
 *   cursor agent -p -f --trust --approve-mcps --workspace <dir> --model <model> 'prompt'
 *   The prompt is passed positionally (NOT via --prompt flag).
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCursorAdapter = createCursorAdapter;
const path = __importStar(require("node:path"));
const node_fs_1 = require("node:fs");
const journal_1 = require("../storage/journal");
const runFiles_1 = require("../storage/runFiles");
const effectIndex_1 = require("../runtime/replay/effectIndex");
const completionProof_1 = require("../cli/completionProof");
const session_1 = require("../session");
const session_2 = require("../cli/commands/session");
const types_1 = require("./types");
const context_1 = require("../prompts/context");
const config_1 = require("../config");
// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------
const HARNESS_NAME = "cursor";
function createHookLogger(hookName) {
    const logDir = process.env.BABYSITTER_LOG_DIR || ".a5c/logs";
    const logFile = logDir ? path.join(logDir, `${hookName}.log`) : null;
    const context = {};
    if (logFile) {
        try {
            (0, node_fs_1.mkdirSync)(logDir, { recursive: true });
        }
        catch {
            // Best-effort
        }
    }
    function write(level, message) {
        if (!logFile)
            return;
        const ts = new Date().toISOString();
        const ctxParts = Object.entries(context).map(([k, v]) => `${k}=${v}`);
        const ctxStr = ctxParts.length > 0 ? ` [${ctxParts.join(" ")}]` : "";
        const line = `[${level}] ${ts}${ctxStr} ${message}\n`;
        try {
            (0, node_fs_1.appendFileSync)(logFile, line);
        }
        catch {
            // Best-effort
        }
    }
    return {
        info: (msg) => write("INFO", msg),
        warn: (msg) => write("WARN", msg),
        error: (msg) => write("ERROR", msg),
        setContext: (key, value) => {
            context[key] = value;
        },
    };
}
// ---------------------------------------------------------------------------
// Journal event helper
// ---------------------------------------------------------------------------
async function appendStopHookEvent(runDir, data) {
    try {
        await (0, journal_1.appendEvent)({
            runDir,
            eventType: "STOP_HOOK_INVOKED",
            event: {
                ...data,
                harness: HARNESS_NAME,
                timestamp: new Date().toISOString(),
            },
        });
    }
    catch {
        // Best-effort: don't fail the hook if journal write fails
    }
}
// ---------------------------------------------------------------------------
// Stdin reader
// ---------------------------------------------------------------------------
function readStdin() {
    return new Promise((resolve, reject) => {
        let data = "";
        process.stdin.setEncoding("utf8");
        process.stdin.on("data", (chunk) => {
            data += chunk;
        });
        process.stdin.on("end", () => resolve(data));
        process.stdin.on("error", reject);
    });
}
function parseHookInput(raw) {
    const trimmed = raw.trim();
    if (!trimmed)
        return {};
    try {
        const parsed = JSON.parse(trimmed);
        if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
            return parsed;
        }
    }
    catch {
        // Malformed JSON — treat as empty
    }
    return {};
}
function safeStr(obj, key) {
    const val = obj[key];
    return typeof val === "string" ? val : "";
}
// ---------------------------------------------------------------------------
// Pending-by-kind helper
// ---------------------------------------------------------------------------
function countPendingByKind(records) {
    const counts = new Map();
    for (const record of records) {
        const key = record.kind ?? "unknown";
        counts.set(key, (counts.get(key) ?? 0) + 1);
    }
    return Object.fromEntries(Array.from(counts.entries()).sort(([a], [b]) => a.localeCompare(b)));
}
/**
 * Returns true when every pending effect is a breakpoint (human-approval gate).
 * Breakpoints require external human action, so the stop hook should allow exit
 * rather than spinning the orchestration loop uselessly.
 */
function isOnlyBreakpoints(pendingByKind) {
    const keys = Object.keys(pendingByKind);
    return keys.length === 1 && keys[0] === "breakpoint";
}
// ---------------------------------------------------------------------------
// Cleanup helper
// ---------------------------------------------------------------------------
async function cleanupSession(filePath) {
    try {
        await (0, session_1.deleteSessionFile)(filePath);
    }
    catch {
        // Best-effort cleanup
    }
}
// ---------------------------------------------------------------------------
// State directory resolution
// ---------------------------------------------------------------------------
function resolveStateDirInternal(args) {
    if (args.stateDir)
        return path.resolve(args.stateDir);
    return (0, config_1.getGlobalStateDir)();
}
// ---------------------------------------------------------------------------
// Stop hook handler
// ---------------------------------------------------------------------------
/**
 * Cursor stop hook handler.
 *
 * Cursor's stop hook protocol uses `followup_message` to auto-continue:
 *   - Return `{ followup_message: "..." }` → Cursor auto-continues with this
 *     message as the next prompt. Controlled by loop_limit in hooks.json.
 *   - Return `{}` → allow session to exit normally.
 *
 * This is distinct from Claude Code's `{ decision: "block", reason: "..." }`
 * protocol, but serves the same purpose in the orchestration loop.
 */
async function handleStopHookImpl(args) {
    const { verbose } = args;
    const log = createHookLogger("babysitter-stop-hook");
    log.info("handleStopHook started (cursor)");
    // 1. Read hook input JSON from stdin
    let rawInput;
    try {
        rawInput = await readStdin();
    }
    catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        log.warn(`stdin read error: ${msg}`);
        process.stdout.write("{}\n");
        return 0;
    }
    finally {
        if (typeof process.stdin.unref === "function") {
            process.stdin.unref();
        }
    }
    const hookInput = parseHookInput(rawInput);
    log.info("Hook input received");
    // 2. Resolve session ID from hook input or cross-harness env var
    //    Cursor provides conversation_id via stdin JSON only; BABYSITTER_SESSION_ID
    //    may be set externally as a cross-harness fallback.
    const sessionId = safeStr(hookInput, "conversation_id") ||
        process.env.BABYSITTER_SESSION_ID ||
        "";
    if (!sessionId) {
        log.info("No conversation_id in hook input — allowing exit");
        process.stdout.write("{}\n");
        return 0;
    }
    log.setContext("session", sessionId);
    log.info(`Session ID: ${sessionId}`);
    // 3. Resolve state directory
    const stateDir = resolveStateDirInternal(args);
    const runsDir = args.runsDir || ".a5c/runs";
    log.info(`Resolved stateDir: ${stateDir}`);
    // 4. Read session state file
    const filePath = (0, session_1.getSessionFilePath)(stateDir, sessionId);
    log.info(`Checking session file at: ${filePath}`);
    let sessionFile;
    try {
        if (!(await (0, session_1.sessionFileExists)(filePath))) {
            log.info(`No active babysitter loop for session ${sessionId} — allowing exit`);
            process.stdout.write("{}\n");
            return 0;
        }
        sessionFile = await (0, session_1.readSessionFile)(filePath);
    }
    catch {
        log.warn(`Session file read error at ${filePath} — allowing exit`);
        process.stdout.write("{}\n");
        return 0;
    }
    const { state } = sessionFile;
    const prompt = sessionFile.prompt ?? "";
    // 5. Check max iterations
    if (state.maxIterations > 0 && state.iteration >= state.maxIterations) {
        if (verbose) {
            process.stderr.write(`[hook:run stop] Max iterations (${state.maxIterations}) reached\n`);
        }
        if (state.runId) {
            await appendStopHookEvent(path.join(runsDir, state.runId), {
                sessionId,
                iteration: state.iteration,
                decision: "approve",
                reason: "max_iterations_reached",
                runState: "",
                pendingKinds: "",
                hasPromise: false,
            });
        }
        await cleanupSession(filePath);
        process.stdout.write("{}\n");
        return 0;
    }
    // 6. Check iteration timing (runaway loop detection)
    const now = (0, session_1.getCurrentTimestamp)();
    const updatedTimes = state.iteration >= 5
        ? (0, session_1.updateIterationTimes)(state.iterationTimes, state.lastIterationAt, now)
        : state.iterationTimes;
    if ((0, session_1.isIterationTooFast)(updatedTimes)) {
        if (verbose) {
            process.stderr.write(`[hook:run stop] Iteration too fast\n`);
        }
        if (state.runId) {
            await appendStopHookEvent(path.join(runsDir, state.runId), {
                sessionId,
                iteration: state.iteration,
                decision: "approve",
                reason: "iteration_too_fast",
                runState: "",
                pendingKinds: "",
                hasPromise: false,
            });
        }
        await cleanupSession(filePath);
        process.stdout.write("{}\n");
        return 0;
    }
    const iteration = state.iteration;
    const maxIterations = state.maxIterations;
    const runId = state.runId ?? "";
    if (runId)
        log.setContext("run", runId);
    // 7. Extract last response text for completion proof check
    // Cursor provides last_response in the stop hook input
    const lastResponse = safeStr(hookInput, "last_response");
    let hasPromise = false;
    let promiseValue = null;
    if (lastResponse) {
        promiseValue = (0, session_2.extractPromiseTag)(lastResponse);
        hasPromise = promiseValue !== null;
        log.info(`last_response extracted (${lastResponse.length} chars)`);
    }
    // 8. If no run is bound, allow exit
    if (!runId) {
        log.info("No run associated with session — allowing exit");
        await cleanupSession(filePath);
        process.stdout.write("{}\n");
        return 0;
    }
    // 9. Get run state and completion proof
    let runState = "";
    let completionProof = "";
    let pendingKinds = "";
    let onlyBreakpointsPending = false;
    try {
        let runDir = path.isAbsolute(runId)
            ? runId
            : path.join(runsDir, runId);
        // Fallback: search alternative locations
        if (!(0, node_fs_1.existsSync)(path.join(runDir, "run.json")) &&
            !path.isAbsolute(runId)) {
            const alternatives = [
                path.join(".a5c", ".a5c", "runs", runId),
                path.join(".a5c", "runs", runId),
            ];
            for (const alt of alternatives) {
                const resolved = path.resolve(alt);
                if (resolved !== path.resolve(runDir) &&
                    (0, node_fs_1.existsSync)(path.join(resolved, "run.json"))) {
                    log.info(`Run not found at ${runDir}, using fallback: ${resolved}`);
                    runDir = resolved;
                    break;
                }
            }
        }
        const metadata = await (0, runFiles_1.readRunMetadata)(runDir);
        const journal = await (0, journal_1.loadJournal)(runDir);
        const index = await (0, effectIndex_1.buildEffectIndex)({ runDir, events: journal });
        const hasCompleted = journal.some((e) => e.type === "RUN_COMPLETED");
        const hasFailed = journal.some((e) => e.type === "RUN_FAILED");
        const pendingRecords = index.listPendingEffects();
        const pendingByKind = countPendingByKind(pendingRecords);
        const kindKeys = Object.keys(pendingByKind);
        if (kindKeys.length > 0) {
            pendingKinds = kindKeys.join(", ");
        }
        onlyBreakpointsPending = pendingRecords.length > 0 && isOnlyBreakpoints(pendingByKind);
        if (hasCompleted) {
            runState = "completed";
            completionProof = (0, completionProof_1.resolveCompletionProof)(metadata);
        }
        else if (hasFailed) {
            runState = "failed";
        }
        else if (pendingRecords.length > 0) {
            runState = "waiting";
        }
        else {
            runState = "created";
        }
    }
    catch {
        runState = "";
    }
    log.info(`Run state: ${runState || "unknown"}`);
    if (completionProof)
        log.info("Completion proof available");
    if (!runState) {
        log.warn(`Run state unknown for ${runId} — allowing exit`);
        if (runId) {
            await appendStopHookEvent(path.join(runsDir, runId), {
                sessionId,
                iteration: state.iteration,
                decision: "approve",
                reason: "run_state_unknown",
                runState,
                pendingKinds,
                hasPromise,
            });
        }
        process.stdout.write("{}\n");
        return 0;
    }
    // 9b. If the run is waiting but ONLY on breakpoints, allow exit.
    if (runState === "waiting" && onlyBreakpointsPending) {
        log.info(`Run waiting on breakpoints only (${pendingKinds}) — allowing exit`);
        if (verbose) {
            process.stderr.write(`[hook:run stop] Run waiting on breakpoint(s) — allowing exit for human resolution\n`);
        }
        if (runId) {
            await appendStopHookEvent(path.join(runsDir, runId), {
                sessionId,
                iteration: state.iteration,
                decision: "approve",
                reason: "breakpoint_waiting",
                runState,
                pendingKinds,
                hasPromise,
            });
        }
        process.stdout.write("{}\n");
        return 0;
    }
    // 10. Check completion proof match
    if (hasPromise && completionProof && promiseValue === completionProof) {
        log.info("Promise matches completion proof — allowing exit");
        if (verbose) {
            process.stderr.write(`[hook:run stop] Valid promise tag detected — run complete\n`);
        }
        if (runId) {
            await appendStopHookEvent(path.join(runsDir, runId), {
                sessionId,
                iteration: state.iteration,
                decision: "approve",
                reason: "completion_proof_matched",
                runState,
                pendingKinds,
                hasPromise,
            });
        }
        await cleanupSession(filePath);
        process.stdout.write("{}\n");
        return 0;
    }
    // 11. Not complete — continue the loop via followup_message
    const nextIteration = iteration + 1;
    const currentTime = (0, session_1.getCurrentTimestamp)();
    const updatedState = {
        ...state,
        iteration: nextIteration,
        lastIterationAt: currentTime,
        iterationTimes: updatedTimes,
    };
    try {
        await (0, session_1.writeSessionFile)(filePath, updatedState, prompt);
    }
    catch {
        log.warn("Failed to update session state");
    }
    // 12. Build followup_message (re-injected as next turn prompt by Cursor)
    let followupMessage;
    if (completionProof) {
        followupMessage = `Babysitter iteration ${nextIteration} | Run completed! To finish: call 'babysitter run:status .a5c/runs/${runId} --json', extract 'completionProof' from the output, then output it in <promise>SECRET</promise> tags. Do not mention or reveal the secret otherwise.\n\n${prompt}`;
    }
    else if (runState === "waiting" && pendingKinds) {
        followupMessage = `Babysitter iteration ${nextIteration} | Waiting on: ${pendingKinds}. Check if pending effects are resolved, then call 'babysitter run:iterate .a5c/runs/${runId} --json'.\n\n${prompt}`;
    }
    else if (runState === "failed") {
        followupMessage = `Babysitter iteration ${nextIteration} | Run failed. Inspect the run journal and fix the issue, then proceed.\n\n${prompt}`;
    }
    else {
        followupMessage = `Babysitter iteration ${nextIteration} | Continue orchestration: call 'babysitter run:iterate .a5c/runs/${runId} --json'.\n\n${prompt}`;
    }
    // Output: followup_message triggers Cursor's auto-continue mechanism.
    // Cursor reads this from stdout and re-enters with followup_message as
    // the next prompt, subject to loop_limit in hooks.json configuration.
    const output = {
        followup_message: followupMessage,
    };
    if (runId) {
        await appendStopHookEvent(path.join(runsDir, runId), {
            sessionId,
            iteration: state.iteration,
            decision: "block",
            reason: "continue_loop",
            runState,
            pendingKinds,
            hasPromise,
        });
    }
    log.info(`Decision: continue via followup_message (iteration=${nextIteration}, maxIterations=${maxIterations})`);
    if (verbose) {
        process.stderr.write(`[hook:run stop] Continuing via followup_message, iteration=${nextIteration} maxIterations=${maxIterations}\n`);
    }
    process.stdout.write(JSON.stringify(output, null, 2) + "\n");
    return 0;
}
// ---------------------------------------------------------------------------
// SessionStart hook handler
// ---------------------------------------------------------------------------
async function handleSessionStartHookImpl(args) {
    const { verbose } = args;
    const log = createHookLogger("babysitter-session-start-hook");
    log.info("handleSessionStartHook started (cursor)");
    // 1. Read hook input JSON from stdin
    let rawInput;
    try {
        rawInput = await readStdin();
    }
    catch {
        process.stdout.write("{}\n");
        return 0;
    }
    finally {
        if (typeof process.stdin.unref === "function") {
            process.stdin.unref();
        }
    }
    const hookInput = parseHookInput(rawInput);
    // 2. Resolve session ID from conversation_id in hook input
    //    Cursor has no env file mechanism; conversation_id is the only source.
    //    BABYSITTER_SESSION_ID used as cross-harness fallback if set externally.
    const sessionId = safeStr(hookInput, "conversation_id") ||
        process.env.BABYSITTER_SESSION_ID ||
        "";
    if (!sessionId) {
        log.info("No conversation_id in hook input — skipping state file creation");
        process.stdout.write("{}\n");
        return 0;
    }
    log.setContext("session", sessionId);
    log.info(`Session ID: ${sessionId}`);
    // 3. Resolve state directory and create baseline session file
    const stateDir = resolveStateDirInternal(args);
    log.info(`Resolved stateDir: ${stateDir}`);
    const filePath = (0, session_1.getSessionFilePath)(stateDir, sessionId);
    try {
        if (!(await (0, session_1.sessionFileExists)(filePath))) {
            const nowTs = (0, session_1.getCurrentTimestamp)();
            const state = {
                active: true,
                iteration: 1,
                maxIterations: 256,
                runId: "",
                startedAt: nowTs,
                lastIterationAt: nowTs,
                iterationTimes: [],
            };
            await (0, session_1.writeSessionFile)(filePath, state, "");
            log.info(`Created session state: ${filePath}`);
            if (verbose) {
                process.stderr.write(`[hook:run session-start] Created session state: ${filePath}\n`);
            }
        }
        else {
            log.info(`Session state already exists: ${filePath}`);
        }
    }
    catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        log.warn(`Failed to create session state: ${msg}`);
        if (verbose) {
            process.stderr.write(`[hook:run session-start] Failed to create session state: ${msg}\n`);
        }
    }
    // 4. Output empty object (no additional context to inject at session start)
    process.stdout.write("{}\n");
    return 0;
}
// ---------------------------------------------------------------------------
// Session binding (run:create flow)
// ---------------------------------------------------------------------------
async function bindSessionImpl(opts) {
    const { sessionId, runId, maxIterations = 256, prompt, verbose } = opts;
    // Resolve state directory
    const stateDir = resolveStateDirInternal({
        stateDir: opts.stateDir,
        pluginRoot: opts.pluginRoot,
    });
    const filePath = (0, session_1.getSessionFilePath)(stateDir, sessionId);
    // Check for existing session (prevent re-entrant runs)
    if (await (0, session_1.sessionFileExists)(filePath)) {
        try {
            const existing = await (0, session_1.readSessionFile)(filePath);
            if (existing.state.runId && existing.state.runId !== runId) {
                return {
                    harness: HARNESS_NAME,
                    sessionId,
                    stateFile: filePath,
                    error: `Session already associated with run: ${existing.state.runId}`,
                };
            }
            // Update existing session with run ID
            await (0, session_1.updateSessionState)(filePath, { runId, active: true }, { state: existing.state, prompt: existing.prompt });
            if (verbose) {
                process.stderr.write(`[run:create] Updated existing session ${sessionId} with run ${runId}\n`);
            }
            return { harness: HARNESS_NAME, sessionId, stateFile: filePath };
        }
        catch {
            // Corrupted state file — overwrite
        }
    }
    // Create new session state with run associated
    const nowTs = (0, session_1.getCurrentTimestamp)();
    const state = {
        active: true,
        iteration: 1,
        maxIterations,
        runId,
        startedAt: nowTs,
        lastIterationAt: nowTs,
        iterationTimes: [],
    };
    try {
        await (0, session_1.writeSessionFile)(filePath, state, prompt);
    }
    catch (e) {
        return {
            harness: HARNESS_NAME,
            sessionId,
            error: `Failed to write session state: ${e instanceof Error ? e.message : String(e)}`,
        };
    }
    if (verbose) {
        process.stderr.write(`[run:create] Session ${sessionId} initialized and bound to run ${runId}\n`);
    }
    return { harness: HARNESS_NAME, sessionId, stateFile: filePath };
}
// ---------------------------------------------------------------------------
// Install helpers
// ---------------------------------------------------------------------------
function installCursorHarness(options) {
    // Cursor CLI is distributed with the Cursor IDE, not via npm.
    // Users must install Cursor IDE from https://cursor.com/
    if (options.dryRun) {
        return Promise.resolve({
            harness: HARNESS_NAME,
            dryRun: true,
            summary: "Cursor CLI is bundled with the Cursor IDE. Download and install Cursor from https://cursor.com/",
        });
    }
    return Promise.resolve({
        harness: HARNESS_NAME,
        summary: "Cursor CLI is bundled with the Cursor IDE. Ensure Cursor is installed and the `cursor` command is available on PATH. Download from https://cursor.com/ if needed.",
    });
}
function installCursorPlugin(options) {
    // Cursor plugins are Git-based repos installed via the Cursor Marketplace
    // or local testing at ~/.cursor/plugins/local/
    if (options.dryRun) {
        return Promise.resolve({
            harness: HARNESS_NAME,
            dryRun: true,
            summary: "Install the Babysitter Cursor plugin from the Cursor Marketplace or copy to ~/.cursor/plugins/local/ for local testing.",
        });
    }
    return Promise.resolve({
        harness: HARNESS_NAME,
        summary: "To install the Babysitter Cursor plugin: use the Cursor Marketplace (Settings > Plugins) or copy the plugin directory to ~/.cursor/plugins/local/ for local development.",
    });
}
// ---------------------------------------------------------------------------
// Adapter factory
// ---------------------------------------------------------------------------
function createCursorAdapter() {
    return {
        name: HARNESS_NAME,
        isActive() {
            // Cursor sets CURSOR_PROJECT_DIR and CURSOR_VERSION in hook execution
            // contexts (not in child processes). These are the best-effort
            // discriminators for detecting we're inside a Cursor session.
            return !!(process.env.CURSOR_PROJECT_DIR ||
                process.env.CURSOR_VERSION);
        },
        autoResolvesSessionId() {
            // Cursor does NOT set an env var for session ID. The conversation_id
            // comes only via hook stdin JSON. The sessionStart hook persists it
            // to a state file, but auto-resolution from env is not possible.
            return false;
        },
        getMissingSessionIdHint() {
            return ("Cursor provides conversation_id only via hook stdin JSON (not as an " +
                "env var). Ensure the sessionStart hook is configured in .cursor/hooks.json " +
                "and is persisting the conversation_id to the state file. Pass --session-id " +
                "explicitly if running outside of the hook context.");
        },
        supportsHookType(hookType) {
            // Hook types that work in Cursor headless CLI mode
            const supported = new Set([
                "stop",
                "session-start",
                "session-end",
                "post-tool-use",
                "after-file-edit",
                "after-shell-execution",
                "before-shell-execution",
                "before-mcp-execution",
                "after-mcp-execution",
                "before-read-file",
                "pre-tool-use",
                "post-tool-use-failure",
                "subagent-start",
                "subagent-stop",
                "pre-compact",
                "before-submit-prompt",
                "before-tab-file-read",
                "after-tab-file-edit",
            ]);
            return supported.has(hookType);
        },
        getUnsupportedHookMessage(hookType) {
            if (hookType === "after-agent-response" || hookType === "after-agent-thought") {
                return `The "${hookType}" hook type does not fire in Cursor headless CLI mode (only in the IDE).`;
            }
            return `Hook type "${hookType}" is not supported by the Cursor adapter.`;
        },
        resolveSessionId(parsed) {
            if (parsed.sessionId)
                return parsed.sessionId;
            // Cursor has no env file mechanism, but BABYSITTER_SESSION_ID may be
            // set externally or by a wrapper script.
            if (process.env.BABYSITTER_SESSION_ID)
                return process.env.BABYSITTER_SESSION_ID;
            return undefined;
        },
        resolveStateDir(args) {
            return resolveStateDirInternal(args);
        },
        resolvePluginRoot(args) {
            const root = args.pluginRoot || process.env.CURSOR_PLUGIN_ROOT;
            return root ? path.resolve(root) : undefined;
        },
        getCapabilities() {
            return [
                types_1.HarnessCapability.HeadlessPrompt,
                types_1.HarnessCapability.StopHook,
                types_1.HarnessCapability.SessionBinding,
                types_1.HarnessCapability.Mcp,
            ];
        },
        bindSession(opts) {
            return bindSessionImpl(opts);
        },
        handleStopHook(args) {
            return handleStopHookImpl(args);
        },
        handleSessionStartHook(args) {
            return handleSessionStartHookImpl(args);
        },
        findHookDispatcherPath(_startCwd) {
            const pluginRoot = process.env.CURSOR_PLUGIN_ROOT;
            if (pluginRoot) {
                const candidate = path.join(path.resolve(pluginRoot), "hooks", "stop-hook.sh");
                if ((0, node_fs_1.existsSync)(candidate))
                    return candidate;
            }
            return null;
        },
        installHarness(options) {
            return installCursorHarness(options);
        },
        installPlugin(options) {
            return installCursorPlugin(options);
        },
        getPromptContext(opts) {
            return (0, context_1.createCursorContext)(opts);
        },
    };
}
