// @a5c-ai/babysitter-sdk/dist/harness/customAdapter.js
"use strict";
/**
 * Custom harness adapter.
 *
 * Used when running babysitter outside any known harness environment.
 * Requires explicit `--session-id`, `--state-dir`, and other arguments
 * that known adapters (claude-code, codex, etc.) infer automatically
 * from environment variables.
 *
 * This is the fallback adapter when `detectAdapter()` finds no active
 * harness, replacing the null adapter as the default.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCustomAdapter = createCustomAdapter;
const path = __importStar(require("node:path"));
const config_1 = require("../config");
function createCustomAdapter() {
    return {
        name: "custom",
        isActive() {
            // The custom adapter is never auto-detected — it is the explicit
            // fallback when no known harness environment is found.
            return false;
        },
        autoResolvesSessionId() {
            return false;
        },
        resolveSessionId(parsed) {
            if (parsed.sessionId)
                return parsed.sessionId;
            // Cross-harness standard env var as fallback for custom adapters
            if (process.env.BABYSITTER_SESSION_ID)
                return process.env.BABYSITTER_SESSION_ID;
            return undefined;
        },
        resolveStateDir(args) {
            if (args.stateDir)
                return path.resolve(args.stateDir);
            return (0, config_1.getGlobalStateDir)();
        },
        resolvePluginRoot(args) {
            return args.pluginRoot ? path.resolve(args.pluginRoot) : undefined;
        },
        getMissingSessionIdHint() {
            return "Custom harness requires --session-id to be explicitly provided.";
        },
        supportsHookType(_hookType) {
            // Custom adapter supports all hook types but with minimal/no-op handlers.
            return true;
        },
        bindSession(opts) {
            if (!opts.sessionId) {
                return Promise.resolve({
                    harness: "custom",
                    sessionId: "",
                    error: "Custom harness requires --session-id to be explicitly provided.",
                    fatal: false,
                });
            }
            // Minimal binding — store state in the default location.
            return Promise.resolve({
                harness: "custom",
                sessionId: opts.sessionId,
            });
        },
        handleStopHook(_args) {
            // No harness lifecycle — approve (allow exit).
            process.stdout.write('{"decision":"approve"}\n');
            return Promise.resolve(0);
        },
        handleSessionStartHook(_args) {
            // No harness lifecycle — nothing to do.
            process.stdout.write("{}\n");
            return Promise.resolve(0);
        },
        findHookDispatcherPath() {
            return null;
        },
    };
}
