// @a5c-ai/babysitter-sdk/dist/harness/discovery.js
"use strict";
/**
 * Harness discovery module.
 *
 * Two distinct detection mechanisms:
 *
 * 1. **Installed discovery** (`discoverHarnesses`) — probes the local system
 *    for known AI coding agent CLIs via `which`/`where`, checks for config
 *    directories, and reports capabilities.  Pure PATH + filesystem detection,
 *    no environment variable checks.
 *
 * 2. **Caller detection** (`detectCallerHarness`) — inspects environment
 *    variables to determine which harness (if any) spawned the current
 *    process.  This answers "are we running *inside* a Claude Code session?"
 *    rather than "is the `claude` binary installed?".
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.KNOWN_HARNESSES = void 0;
exports.checkCliAvailable = checkCliAvailable;
exports.discoverHarnesses = discoverHarnesses;
exports.detectCallerHarness = detectCallerHarness;
const node_child_process_1 = require("node:child_process");
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
const os = __importStar(require("node:os"));
const types_1 = require("./types");
// ---------------------------------------------------------------------------
// Known harnesses
// ---------------------------------------------------------------------------
/** Detection specs for all supported harnesses. */
exports.KNOWN_HARNESSES = [
    {
        name: "claude-code",
        cli: "claude",
        callerEnvVars: ["CLAUDE_ENV_FILE"],
        capabilities: [types_1.HarnessCapability.SessionBinding, types_1.HarnessCapability.StopHook, types_1.HarnessCapability.Mcp, types_1.HarnessCapability.HeadlessPrompt],
    },
    {
        name: "codex",
        cli: "codex",
        callerEnvVars: ["CODEX_THREAD_ID", "CODEX_SESSION_ID", "CODEX_PLUGIN_ROOT"],
        capabilities: [types_1.HarnessCapability.SessionBinding, types_1.HarnessCapability.StopHook, types_1.HarnessCapability.HeadlessPrompt],
    },
    {
        name: "cursor",
        cli: "cursor",
        // CURSOR_PROJECT_DIR and CURSOR_VERSION are set by Cursor in hook execution
        // contexts only (not in child processes spawned by hooks).
        callerEnvVars: ["CURSOR_PROJECT_DIR", "CURSOR_VERSION"],
        capabilities: [types_1.HarnessCapability.HeadlessPrompt, types_1.HarnessCapability.StopHook, types_1.HarnessCapability.SessionBinding, types_1.HarnessCapability.Mcp],
    },
    {
        name: "gemini-cli",
        cli: "gemini",
        callerEnvVars: ["GEMINI_SESSION_ID", "GEMINI_PROJECT_DIR", "GEMINI_CWD"],
        capabilities: [types_1.HarnessCapability.SessionBinding, types_1.HarnessCapability.HeadlessPrompt, types_1.HarnessCapability.StopHook],
    },
    {
        name: "github-copilot",
        cli: "copilot",
        // Note: official Copilot docs do not confirm these env vars are injected
        // into hooks. They are used as best-effort discriminators for caller detection.
        callerEnvVars: ["COPILOT_HOME", "COPILOT_GITHUB_TOKEN"],
        // No StopHook — Copilot CLI uses in-turn orchestration model
        capabilities: [types_1.HarnessCapability.HeadlessPrompt, types_1.HarnessCapability.SessionBinding, types_1.HarnessCapability.Mcp],
    },
    {
        name: "opencode",
        cli: "opencode",
        callerEnvVars: ["BABYSITTER_SESSION_ID", "OPENCODE_CONFIG"],
        capabilities: [types_1.HarnessCapability.HeadlessPrompt],
    },
    {
        name: "oh-my-pi",
        cli: "omp",
        callerEnvVars: ["OMP_SESSION_ID", "OMP_PLUGIN_ROOT"],
        capabilities: [types_1.HarnessCapability.Programmatic, types_1.HarnessCapability.SessionBinding, types_1.HarnessCapability.HeadlessPrompt, types_1.HarnessCapability.Mcp],
    },
    {
        name: "pi",
        cli: "pi",
        callerEnvVars: ["PI_SESSION_ID", "PI_PLUGIN_ROOT"],
        capabilities: [types_1.HarnessCapability.Programmatic, types_1.HarnessCapability.SessionBinding, types_1.HarnessCapability.HeadlessPrompt],
    },
    {
        name: "internal",
        cli: "internal",
        callerEnvVars: [],
        capabilities: [types_1.HarnessCapability.Programmatic, types_1.HarnessCapability.SessionBinding, types_1.HarnessCapability.StopHook, types_1.HarnessCapability.HeadlessPrompt],
    },
];
// ---------------------------------------------------------------------------
// Config path detection
// ---------------------------------------------------------------------------
/** Map of harness names to config directory names to look for. */
const CONFIG_PATHS = {
    "claude-code": [".claude"],
    codex: [".codex"],
    pi: [".pi"],
    "oh-my-pi": [".omp"],
    "gemini-cli": [".gemini"],
    "github-copilot": [".copilot", ".github"],
    cursor: [".cursor", ".cursorrules"],
    opencode: [".opencode"],
};
/**
 * Checks whether a harness config directory exists in cwd or home.
 */
async function detectConfig(harnessName) {
    const configDirs = CONFIG_PATHS[harnessName];
    if (!configDirs || configDirs.length === 0)
        return false;
    const cwd = process.cwd();
    const home = os.homedir();
    for (const dir of configDirs) {
        // Check in cwd
        try {
            await node_fs_1.promises.access(path.join(cwd, dir));
            return true;
        }
        catch { /* skip */ }
        // Check in home
        try {
            await node_fs_1.promises.access(path.join(home, dir));
            return true;
        }
        catch { /* skip */ }
    }
    return false;
}
// ---------------------------------------------------------------------------
// CLI availability check
// ---------------------------------------------------------------------------
/** Timeout in milliseconds for each CLI probe. */
const CLI_CHECK_TIMEOUT_MS = 2000;
/**
 * Checks whether a single CLI command is available on the system.
 *
 * Uses `which` on Unix and `where` on Windows to locate the binary, then
 * attempts `<command> --version` to extract a version string.
 *
 * @returns An object indicating availability and, when found, the resolved
 *          path and version string.
 */
async function checkCliAvailable(command) {
    const locateCmd = process.platform === "win32" ? "where" : "which";
    let cliPath;
    try {
        cliPath = await execFilePromise(locateCmd, [command], CLI_CHECK_TIMEOUT_MS);
    }
    catch {
        return { available: false };
    }
    // `which`/`where` may return multiple lines; take the first.
    const resolvedPath = cliPath.split(/\r?\n/)[0]?.trim();
    if (!resolvedPath) {
        return { available: false };
    }
    // Attempt to extract version.
    let version;
    try {
        const raw = await execFilePromise(command, ["--version"], CLI_CHECK_TIMEOUT_MS);
        version = parseVersionString(raw);
    }
    catch {
        // Version extraction is best-effort; the CLI is still considered installed.
    }
    return { available: true, path: resolvedPath, version };
}
// ---------------------------------------------------------------------------
// Installed discovery (mechanism 1)
// ---------------------------------------------------------------------------
/**
 * Probes the local environment for all known harness CLIs in parallel.
 *
 * Checks only whether each CLI binary is installed on PATH, whether config
 * directories exist, and what capabilities the harness advertises.
 *
 * Does **not** inspect environment variables — use `detectCallerHarness()`
 * for that.
 *
 * Each harness is checked independently via `Promise.allSettled` so that a
 * single failing/timing-out probe does not block the rest.
 *
 * @returns An array of discovery results sorted alphabetically by harness name.
 */
async function discoverHarnesses() {
    const settled = await Promise.allSettled(exports.KNOWN_HARNESSES.map((spec) => probeHarness(spec)));
    const results = settled.map((outcome, idx) => {
        if (outcome.status === "fulfilled") {
            return outcome.value;
        }
        // Graceful degradation: if the probe itself threw, return a safe default.
        const spec = exports.KNOWN_HARNESSES[idx];
        return {
            name: spec.name,
            installed: false,
            cliCommand: spec.cli,
            configFound: false,
            capabilities: spec.capabilities,
            platform: process.platform,
        };
    });
    // The 'internal' harness is always available (built into the SDK).
    const internalIdx = results.findIndex((r) => r.name === "internal");
    if (internalIdx >= 0) {
        results[internalIdx].installed = true;
    }
    results.sort((a, b) => a.name.localeCompare(b.name));
    return results;
}
// ---------------------------------------------------------------------------
// Caller detection (mechanism 2)
// ---------------------------------------------------------------------------
/**
 * Detects which harness (if any) spawned the current process by inspecting
 * environment variables.
 *
 * This answers the question "are we running inside a Claude Code / Codex /
 * Gemini CLI / etc. session right now?" — completely independent of whether
 * those CLIs are installed on PATH.
 *
 * @returns The detected caller harness, or `null` if no caller is detected.
 *          When multiple harnesses match (e.g. pi and oh-my-pi share some
 *          env vars), the first match in `KNOWN_HARNESSES` order wins.
 */
function detectCallerHarness() {
    for (const spec of exports.KNOWN_HARNESSES) {
        if (spec.callerEnvVars.length === 0)
            continue;
        const matchedVars = spec.callerEnvVars.filter((v) => !!process.env[v]);
        if (matchedVars.length > 0) {
            return {
                name: spec.name,
                matchedEnvVars: matchedVars,
                capabilities: spec.capabilities,
            };
        }
    }
    return null;
}
// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------
/**
 * Probes a single harness spec and returns an installed-discovery result.
 */
async function probeHarness(spec) {
    const cliCheck = await checkCliAvailable(spec.cli);
    const configFound = await detectConfig(spec.name);
    return {
        name: spec.name,
        installed: cliCheck.available,
        version: cliCheck.version,
        cliPath: cliCheck.path,
        cliCommand: spec.cli,
        configFound,
        capabilities: spec.capabilities,
        platform: process.platform,
    };
}
/**
 * Promise wrapper around `child_process.execFile` (no shell).
 *
 * Resolves with trimmed stdout on exit code 0, rejects otherwise.
 */
function execFilePromise(command, args, timeoutMs) {
    return new Promise((resolve, reject) => {
        const child = (0, node_child_process_1.execFile)(command, args, { timeout: timeoutMs, windowsHide: true }, (error, stdout) => {
            if (error) {
                reject(error);
                return;
            }
            resolve(String(stdout).trim());
        });
        // Safety: if the child process lingers beyond the timeout, kill it.
        // Node's built-in timeout sends SIGTERM; this is a belt-and-suspenders guard.
        child.on("error", (err) => {
            reject(err);
        });
    });
}
/**
 * Extracts a semver-ish version from raw `--version` output.
 *
 * Looks for patterns like "1.2.3", "v2.0.0-beta.1", etc.
 */
function parseVersionString(raw) {
    const match = raw.match(/v?(\d+\.\d+\.\d+(?:[^\s]*))/);
    return match?.[1];
}
