// @a5c-ai/babysitter-sdk/dist/harness/installSupport.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.execFilePromise = execFilePromise;
exports.renderCommand = renderCommand;
exports.getNpmCommand = getNpmCommand;
exports.getNpxCommand = getNpxCommand;
exports.runPackageBinaryViaNpx = runPackageBinaryViaNpx;
exports.installCliViaNpm = installCliViaNpm;
exports.getCodexHome = getCodexHome;
exports.getInstalledCodexSkillDir = getInstalledCodexSkillDir;
exports.isCodexPluginInstalled = isCodexPluginInstalled;
exports.getClaudeInstalledPluginsPath = getClaudeInstalledPluginsPath;
exports.isClaudePluginInstalled = isClaudePluginInstalled;
exports.getGeminiExtensionDir = getGeminiExtensionDir;
exports.isGeminiPluginInstalled = isGeminiPluginInstalled;
const node_child_process_1 = require("node:child_process");
const node_fs_1 = require("node:fs");
const os = __importStar(require("node:os"));
const path = __importStar(require("node:path"));
const discovery_1 = require("./discovery");
const exceptions_1 = require("../runtime/exceptions");
async function execFilePromise(command, args, options = {}) {
    return new Promise((resolve) => {
        (0, node_child_process_1.execFile)(command, args, {
            cwd: options.cwd,
            env: options.env,
            windowsHide: true,
            maxBuffer: 20 * 1024 * 1024,
        }, (error, stdout, stderr) => {
            const execError = error;
            resolve({
                stdout: String(stdout ?? ""),
                stderr: String(stderr ?? ""),
                exitCode: typeof execError?.status === "number" ? execError.status : error ? 1 : 0,
            });
        });
    });
}
function renderCommand(command, args) {
    return [command, ...args].join(" ");
}
function getNpmCommand() {
    return process.platform === "win32" ? "npm.cmd" : "npm";
}
function getNpxCommand() {
    return process.platform === "win32" ? "npx.cmd" : "npx";
}
async function runPackageBinaryViaNpx(args) {
    const displayCommand = "npx";
    const command = getNpxCommand();
    const commandArgs = ["--yes", args.packageName, ...args.packageArgs];
    if (args.options.dryRun) {
        return {
            harness: args.harness,
            dryRun: true,
            summary: args.summary,
            command: renderCommand(displayCommand, commandArgs),
            location: args.location,
        };
    }
    const result = await execFilePromise(command, commandArgs, {
        cwd: args.cwd,
        env: args.env,
    });
    if (result.exitCode !== 0) {
        throw new exceptions_1.BabysitterRuntimeError("HarnessPluginInstallFailed", `${renderCommand(displayCommand, commandArgs)} failed`, {
            category: exceptions_1.ErrorCategory.External,
            details: {
                stdout: result.stdout,
                stderr: result.stderr,
                exitCode: result.exitCode,
            },
        });
    }
    return {
        harness: args.harness,
        summary: args.summary,
        command: renderCommand(displayCommand, commandArgs),
        location: args.location,
        output: [result.stdout.trim(), result.stderr.trim()].filter(Boolean).join("\n"),
    };
}
async function installCliViaNpm(args) {
    const cliInfo = await (0, discovery_1.checkCliAvailable)(args.cliCommand);
    if (cliInfo.available) {
        return {
            harness: args.harness,
            warning: `${args.harness} is already installed; nothing to do.`,
            location: cliInfo.path,
        };
    }
    const command = getNpmCommand();
    const commandArgs = ["install", "-g", args.packageName];
    if (args.options.dryRun) {
        return {
            harness: args.harness,
            dryRun: true,
            summary: args.summary,
            command: renderCommand("npm", commandArgs),
        };
    }
    const result = await execFilePromise(command, commandArgs);
    if (result.exitCode !== 0) {
        throw new exceptions_1.BabysitterRuntimeError("HarnessInstallFailed", `${renderCommand(command, commandArgs)} failed`, {
            category: exceptions_1.ErrorCategory.External,
            details: {
                stdout: result.stdout,
                stderr: result.stderr,
                exitCode: result.exitCode,
            },
        });
    }
    return {
        harness: args.harness,
        summary: args.summary,
        command: renderCommand(command, commandArgs),
        output: [result.stdout.trim(), result.stderr.trim()].filter(Boolean).join("\n"),
    };
}
function getCodexHome() {
    return process.env.CODEX_HOME || path.join(os.homedir(), ".codex");
}
function getInstalledCodexSkillDir() {
    return path.join(getCodexHome(), "skills", "babysit");
}
function isCodexPluginInstalled() {
    return (0, node_fs_1.existsSync)(path.join(getInstalledCodexSkillDir(), "SKILL.md"));
}
function getClaudeInstalledPluginsPath() {
    return path.join(os.homedir(), ".claude", "plugins", "installed_plugins.json");
}
function isClaudePluginInstalled() {
    const installedPath = getClaudeInstalledPluginsPath();
    if (!(0, node_fs_1.existsSync)(installedPath)) {
        return false;
    }
    try {
        const parsed = JSON.parse((0, node_fs_1.readFileSync)(installedPath, "utf8"));
        return Object.keys(parsed.plugins ?? {}).some((key) => key.includes("babysitter@a5c.ai"));
    }
    catch {
        return false;
    }
}
function getGeminiExtensionDir(workspace) {
    const base = path.resolve(workspace ?? os.homedir());
    return path.join(base, ".gemini", "extensions", "babysitter");
}
function isGeminiPluginInstalled(workspace) {
    return (0, node_fs_1.existsSync)(getGeminiExtensionDir(workspace));
}
