// @a5c-ai/babysitter-sdk/dist/harness/internal.js
"use strict";
/**
 * Internal harness adapter.
 *
 * The 'internal' harness represents the SDK's built-in programmatic execution
 * engine (piWrapper). It is always available, requires no external CLI, and is
 * the default harness for harness:create-run.
 *
 * Unlike the 'pi' harness (which invokes the pi CLI as a child process), the
 * internal harness calls piWrapper.prompt() directly in-process.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInternalAdapter = createInternalAdapter;
const path = __importStar(require("node:path"));
const claudeCode_1 = require("./claudeCode");
const types_1 = require("./types");
const context_1 = require("../prompts/context");
function resolveInternalPluginRoot(args = {}) {
    const root = args.pluginRoot || process.env.OMP_PLUGIN_ROOT || process.env.PI_PLUGIN_ROOT;
    return root ? path.resolve(root) : undefined;
}
function resolveInternalStateDir(args) {
    if (args.stateDir)
        return path.resolve(args.stateDir);
    if (process.env.BABYSITTER_STATE_DIR) {
        return path.resolve(process.env.BABYSITTER_STATE_DIR);
    }
    const pluginRoot = resolveInternalPluginRoot(args);
    if (pluginRoot) {
        return path.resolve(pluginRoot, "..", ".a5c");
    }
    return path.resolve(".a5c");
}
function resolveInternalSessionId(parsed) {
    if (parsed.sessionId)
        return parsed.sessionId;
    if (process.env.OMP_SESSION_ID)
        return process.env.OMP_SESSION_ID;
    if (process.env.PI_SESSION_ID)
        return process.env.PI_SESSION_ID;
    return undefined;
}
function createInternalAdapter() {
    const claude = (0, claudeCode_1.createClaudeCodeAdapter)();
    return {
        name: "internal",
        // The internal harness is never auto-detected via env vars.
        // It is explicitly selected as the default or via --harness internal.
        isActive() {
            return false;
        },
        autoResolvesSessionId() {
            return true;
        },
        resolveSessionId(parsed) {
            return resolveInternalSessionId(parsed);
        },
        resolveStateDir(args) {
            return resolveInternalStateDir(args);
        },
        resolvePluginRoot(args) {
            return resolveInternalPluginRoot(args);
        },
        getCapabilities() {
            return [types_1.HarnessCapability.Programmatic, types_1.HarnessCapability.SessionBinding, types_1.HarnessCapability.StopHook, types_1.HarnessCapability.HeadlessPrompt];
        },
        async bindSession(opts) {
            const stateDir = resolveInternalStateDir({
                stateDir: opts.stateDir,
                pluginRoot: opts.pluginRoot,
            });
            const result = await claude.bindSession({
                ...opts,
                stateDir,
            });
            return { ...result, harness: "internal" };
        },
        handleStopHook(args) {
            const pluginRoot = resolveInternalPluginRoot(args);
            const stateDir = resolveInternalStateDir({
                stateDir: args.stateDir,
                pluginRoot,
            });
            return claude.handleStopHook({
                ...args,
                pluginRoot,
                stateDir,
            });
        },
        handleSessionStartHook(args) {
            const pluginRoot = resolveInternalPluginRoot(args);
            const stateDir = resolveInternalStateDir({
                stateDir: args.stateDir,
                pluginRoot,
            });
            return claude.handleSessionStartHook({
                ...args,
                pluginRoot,
                stateDir,
            });
        },
        findHookDispatcherPath(_startCwd) {
            return null;
        },
        // Internal harness has no CLI to install — it IS the SDK.
        // eslint-disable-next-line @typescript-eslint/require-await
        async installHarness(options) {
            return {
                harness: "internal",
                summary: "The internal harness is built into the babysitter SDK and requires no separate installation.",
                dryRun: options.dryRun,
            };
        },
        getPromptContext(opts) {
            return (0, context_1.createPiContext)(opts);
        },
    };
}
