// @a5c-ai/babysitter-sdk/dist/harness/invoker.js
"use strict";
/**
 * Harness invoker module.
 *
 * Provides functions to programmatically invoke harness CLIs (claude, codex,
 * pi, gemini, cursor, opencode) as child processes. Each harness has its own
 * flag mapping; this module abstracts those differences behind a uniform API.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.HARNESS_CLI_MAP = void 0;
exports.buildHarnessArgs = buildHarnessArgs;
exports.invokeHarness = invokeHarness;
const node_child_process_1 = require("node:child_process");
const node_fs_1 = require("node:fs");
const os = __importStar(require("node:os"));
const path = __importStar(require("node:path"));
const discovery_1 = require("./discovery");
const piWrapper_1 = require("./piWrapper");
const exceptions_1 = require("../runtime/exceptions");
/**
 * Mapping from harness identifier to CLI command and flag details.
 */
exports.HARNESS_CLI_MAP = {
    "claude-code": { cli: "claude", supportsModel: true, promptStyle: "flag" },
    codex: {
        cli: "codex",
        workspaceFlag: "-C",
        supportsModel: true,
        promptStyle: "positional",
        baseArgs: ["exec", "--dangerously-bypass-approvals-and-sandbox", "--skip-git-repo-check"],
    },
    pi: { cli: "pi", workspaceFlag: "--workspace", supportsModel: true, promptStyle: "flag" },
    "oh-my-pi": { cli: "omp", workspaceFlag: "--workspace", supportsModel: true, promptStyle: "flag" },
    "gemini-cli": { cli: "gemini", supportsModel: true, promptStyle: "flag" },
    "github-copilot": { cli: "copilot", supportsModel: true, promptStyle: "flag" },
    cursor: { cli: "cursor", supportsModel: true, promptStyle: "positional", baseArgs: ["agent"], workspaceFlag: "--workspace" },
    opencode: { cli: "opencode", supportsModel: true, promptStyle: "positional", baseArgs: ["run"] },
};
const PROGRAMMATIC_ONLY_HARNESSES = ["internal"];
const SUPPORTED_HARNESS_NAMES = [
    ...PROGRAMMATIC_ONLY_HARNESSES,
    ...Object.keys(exports.HARNESS_CLI_MAP),
];
function quotePowerShellArg(value) {
    return `'${value.replace(/'/g, "''")}'`;
}
function buildLaunchSpec(name, spec, cliPath, args, promptFilePath) {
    if (process.platform === "win32" && name === "codex") {
        const commandLine = [
            "Get-Content -Raw",
            quotePowerShellArg(promptFilePath ?? ""),
            "|",
            "&",
            quotePowerShellArg(spec.cli),
            ...args.map(quotePowerShellArg),
        ].join(" ");
        return {
            command: "powershell.exe",
            args: ["-NoProfile", "-Command", commandLine],
            shell: false,
        };
    }
    if (process.platform === "win32") {
        return {
            command: spec.cli,
            args,
            shell: true,
        };
    }
    return {
        command: cliPath ?? spec.cli,
        args,
        shell: false,
    };
}
// ---------------------------------------------------------------------------
// Arg builder (pure function)
// ---------------------------------------------------------------------------
/**
 * Builds CLI argument array for a given harness and invocation options.
 *
 * This is a pure function with no side-effects, suitable for unit testing the
 * flag mapping logic in isolation.
 *
 * @throws {BabysitterRuntimeError} if `name` is not a known harness.
 */
function buildHarnessArgs(name, options) {
    const spec = exports.HARNESS_CLI_MAP[name];
    if (!spec) {
        throw new exceptions_1.BabysitterRuntimeError("UnknownHarnessError", `Unknown harness: "${name}". Supported harnesses: ${SUPPORTED_HARNESS_NAMES.join(", ")}`, {
            category: exceptions_1.ErrorCategory.Validation,
            suggestions: [`Did you mean one of: ${SUPPORTED_HARNESS_NAMES.join(", ")}?`],
            nextSteps: ["Use a supported harness name"],
        });
    }
    const args = [...(spec.baseArgs ?? [])];
    if ((spec.promptStyle ?? "flag") === "positional") {
        args.push(name === "codex" ? "-" : options.prompt);
    }
    else {
        args.push("--prompt", options.prompt);
    }
    if (options.model && spec.supportsModel) {
        args.push("--model", options.model);
    }
    if (options.workspace && spec.workspaceFlag) {
        args.push(spec.workspaceFlag, options.workspace);
    }
    return args;
}
// ---------------------------------------------------------------------------
// Invoker
// ---------------------------------------------------------------------------
/** Default timeout for harness invocations (15 minutes). */
const DEFAULT_TIMEOUT_MS = 900_000;
/**
 * Invokes a harness CLI as a child process and returns the result.
 *
 * Steps:
 *   1. Validate that `name` is a known harness.
 *   2. Check that the CLI binary is installed via `checkCliAvailable`.
 *   3. Build args via `buildHarnessArgs`.
 *   4. Spawn via `child_process.execFile`. On Windows, enable shell mode so
 *      PATH-resolved `.cmd` shims (for example npm-installed CLIs) can launch.
 *   5. Capture stdout + stderr, measure wall-clock duration.
 *   6. Return a `HarnessInvokeResult`.
 *
 * @throws {BabysitterRuntimeError} if the harness is unknown or the CLI is
 *   not installed.
 */
async function invokeHarness(name, options) {
    if (name === "internal") {
        const session = (0, piWrapper_1.createPiSession)({
            workspace: options.workspace,
            model: options.model,
            timeout: options.timeout,
            ephemeral: true,
        });
        try {
            const result = await session.prompt(options.prompt, options.timeout);
            return {
                ...result,
                harness: "internal",
            };
        }
        finally {
            session.dispose();
        }
    }
    const spec = exports.HARNESS_CLI_MAP[name];
    if (!spec) {
        throw new exceptions_1.BabysitterRuntimeError("UnknownHarnessError", `Unknown harness: "${name}". Supported harnesses: ${SUPPORTED_HARNESS_NAMES.join(", ")}`, {
            category: exceptions_1.ErrorCategory.Validation,
            suggestions: [`Did you mean one of: ${SUPPORTED_HARNESS_NAMES.join(", ")}?`],
            nextSteps: ["Use a supported harness name"],
        });
    }
    // Verify CLI availability.
    const cliCheck = await (0, discovery_1.checkCliAvailable)(spec.cli);
    if (!cliCheck.available) {
        throw new exceptions_1.BabysitterRuntimeError("HarnessCliNotInstalledError", `Harness CLI "${spec.cli}" is not installed or not found on PATH`, {
            category: exceptions_1.ErrorCategory.External,
            nextSteps: [
                `Install the "${spec.cli}" CLI and ensure it is on your PATH`,
                `Verify installation by running: ${spec.cli} --version`,
            ],
        });
    }
    const args = buildHarnessArgs(name, options);
    const timeoutMs = options.timeout ?? DEFAULT_TIMEOUT_MS;
    let promptTempDir;
    let promptFilePath;
    if (process.platform === "win32" && name === "codex") {
        promptTempDir = await node_fs_1.promises.mkdtemp(path.join(os.tmpdir(), "babysitter-codex-"));
        promptFilePath = path.join(promptTempDir, "prompt.txt");
        await node_fs_1.promises.writeFile(promptFilePath, options.prompt, "utf8");
    }
    const launch = buildLaunchSpec(name, spec, cliCheck.path, args, promptFilePath);
    const childCwd = name === "codex" ? process.cwd() : options.workspace;
    const startTime = Date.now();
    return new Promise((resolve, reject) => {
        const childEnv = options.env
            ? { ...process.env, ...options.env }
            : process.env;
        const cleanupPromptFile = async () => {
            if (!promptTempDir) {
                return;
            }
            await node_fs_1.promises.rm(promptTempDir, { recursive: true, force: true }).catch(() => { });
        };
        try {
            const child = (0, node_child_process_1.execFile)(launch.command, launch.args, {
                cwd: childCwd,
                timeout: timeoutMs,
                windowsHide: true,
                maxBuffer: 50 * 1024 * 1024, // 50 MiB
                env: childEnv,
                shell: launch.shell,
            }, (error, stdout, stderr) => {
                void cleanupPromptFile();
                const duration = Date.now() - startTime;
                const stderrStr = String(stderr);
                const output = stderrStr.length > 0
                    ? `${String(stdout)}\n${stderrStr}`.trim()
                    : String(stdout).trim();
                if (error) {
                    const execError = error;
                    const killed = execError.killed === true;
                    const exitCode = typeof execError.status === "number" ? execError.status : 1;
                    resolve({
                        success: false,
                        output: killed
                            ? `Process timed out after ${timeoutMs}ms\n${output}`.trim()
                            : output,
                        exitCode,
                        duration,
                        harness: name,
                    });
                    return;
                }
                resolve({
                    success: true,
                    output,
                    exitCode: 0,
                    duration,
                    harness: name,
                });
            });
            if (name === "codex" && process.platform !== "win32" && child.stdin) {
                child.stdin.end(options.prompt);
            }
        }
        catch (err) {
            void cleanupPromptFile();
            reject(new exceptions_1.BabysitterRuntimeError("HarnessSpawnError", `Failed to spawn ${spec.cli}: ${err instanceof Error ? err.message : String(err)}`, { category: exceptions_1.ErrorCategory.External }));
        }
    });
}
