// @a5c-ai/babysitter-sdk/dist/harness/pi.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.installPiPlugin = installPiPlugin;
exports.createPiAdapter = createPiAdapter;
const path = __importStar(require("node:path"));
const types_1 = require("./types");
const context_1 = require("../prompts/context");
const config_1 = require("../config");
const session_1 = require("../session");
const installSupport_1 = require("./installSupport");
function resolvePiPluginRoot(args = {}) {
    const root = args.pluginRoot || process.env.PI_PLUGIN_ROOT;
    return root ? path.resolve(root) : undefined;
}
function resolvePiStateDir(args) {
    if (args.stateDir)
        return path.resolve(args.stateDir);
    return (0, config_1.getGlobalStateDir)();
}
function resolvePiSessionId(parsed) {
    if (parsed.sessionId)
        return parsed.sessionId;
    if (process.env.BABYSITTER_SESSION_ID)
        return process.env.BABYSITTER_SESSION_ID;
    if (process.env.PI_SESSION_ID)
        return process.env.PI_SESSION_ID;
    return undefined;
}
async function bindPiSession(opts) {
    const stateDir = resolvePiStateDir({
        stateDir: opts.stateDir,
        pluginRoot: opts.pluginRoot,
    });
    const stateFile = (0, session_1.getSessionFilePath)(stateDir, opts.sessionId);
    if (await (0, session_1.sessionFileExists)(stateFile)) {
        const existing = await (0, session_1.readSessionFile)(stateFile);
        if (existing.state.runId && existing.state.runId !== opts.runId) {
            return {
                harness: "pi",
                sessionId: opts.sessionId,
                stateFile,
                error: `Session already associated with run ${existing.state.runId}`,
            };
        }
        await (0, session_1.updateSessionState)(stateFile, { active: true, runId: opts.runId }, existing);
        return {
            harness: "pi",
            sessionId: opts.sessionId,
            stateFile,
        };
    }
    const now = (0, session_1.getCurrentTimestamp)();
    const state = {
        active: true,
        iteration: 1,
        maxIterations: opts.maxIterations ?? 256,
        runId: opts.runId,
        startedAt: now,
        lastIterationAt: now,
        iterationTimes: [],
    };
    await (0, session_1.writeSessionFile)(stateFile, state, opts.prompt);
    return {
        harness: "pi",
        sessionId: opts.sessionId,
        stateFile,
    };
}
function writeNoopHookResult() {
    process.stdout.write("{}\n");
}
async function installPiFamilyPlugin(args) {
    const packageArgs = ["install"];
    if (args.options.workspace) {
        packageArgs.push("--workspace", path.resolve(args.options.workspace));
    }
    else {
        packageArgs.push("--global");
    }
    return (0, installSupport_1.runPackageBinaryViaNpx)({
        harness: args.harness,
        packageName: args.packageName,
        packageArgs,
        summary: args.options.workspace
            ? `Install the published Babysitter ${args.harness} package for the target workspace.`
            : `Install the published Babysitter ${args.harness} package into the user profile.`,
        options: args.options,
        env: process.env,
    });
}
async function installPiPlugin(options) {
    return installPiFamilyPlugin({
        harness: "pi",
        packageName: "@a5c-ai/babysitter-pi",
        options,
    });
}
function createPiAdapter() {
    return {
        name: "pi",
        isActive() {
            return !!(process.env.BABYSITTER_SESSION_ID ||
                process.env.PI_SESSION_ID ||
                process.env.PI_PLUGIN_ROOT);
        },
        autoResolvesSessionId() {
            return true;
        },
        getMissingSessionIdHint() {
            return "Pi should provide PI_SESSION_ID when the Babysitter package is active.";
        },
        supportsHookType(_hookType) {
            return false;
        },
        getUnsupportedHookMessage(hookType) {
            return `Pi does not use babysitter hook:run for "${hookType}". Use the Pi package skills and extension bridge instead.`;
        },
        resolveSessionId(parsed) {
            return resolvePiSessionId(parsed);
        },
        resolveStateDir(args) {
            return resolvePiStateDir(args);
        },
        resolvePluginRoot(args) {
            return resolvePiPluginRoot(args);
        },
        bindSession(opts) {
            return bindPiSession(opts);
        },
        handleStopHook(_args) {
            writeNoopHookResult();
            return Promise.resolve(0);
        },
        handleSessionStartHook(_args) {
            writeNoopHookResult();
            return Promise.resolve(0);
        },
        findHookDispatcherPath(_startCwd) {
            return null;
        },
        installHarness(options) {
            return (0, installSupport_1.installCliViaNpm)({
                harness: "pi",
                cliCommand: "pi",
                packageName: "@mariozechner/pi-coding-agent",
                summary: "Install the Pi Coding Agent CLI globally via npm.",
                options,
            });
        },
        installPlugin(options) {
            return installPiPlugin(options);
        },
        getCapabilities() {
            return [types_1.HarnessCapability.Programmatic, types_1.HarnessCapability.SessionBinding, types_1.HarnessCapability.HeadlessPrompt];
        },
        getPromptContext(opts) {
            return (0, context_1.createPiContext)(opts);
        },
    };
}
