// @a5c-ai/babysitter-sdk/dist/harness/piWrapper.js
"use strict";
/**
 * Pi programmatic API wrapper.
 *
 * Wraps `@mariozechner/pi-coding-agent`'s `createAgentSession()` to expose
 * a persistent `AgentSession` through a babysitter-friendly interface.
 *
 * The session is lazily initialized on first use and reused across prompts.
 * Events are forwarded through the `subscribe()` method.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PiSessionHandle = void 0;
exports.discoverRepoInstructionPrompts = discoverRepoInstructionPrompts;
exports.createPiSession = createPiSession;
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
const exceptions_1 = require("../runtime/exceptions");
const config_loader_1 = require("../compression/config-loader");
const piSecureSandbox_1 = require("./piSecureSandbox");
const DEFAULT_TIMEOUT_MS = 900_000;
const DEFAULT_BASH_SANDBOX_MODE = "local";
const AGENT_END_PROMPT_SETTLE_GRACE_MS = 250;
const CLAUDE_MD_FILENAME = "CLAUDE.md";
const AGENTS_MD_FILENAME = "AGENTS.md";
// Use a variable to prevent TypeScript from resolving the module at compile time.
// The package is optional — only needed when pi harness is actually used.
const PI_MODULE_ID = "@mariozechner/pi-coding-agent";
// Use an indirect dynamic import so TypeScript does not downlevel to require()
// in CommonJS builds.  The pi-coding-agent package is ESM-only so a real
// import() is required.  In Vitest, the standard import() works because mocks
// intercept it; in production builds we use the new Function trick to preserve
// the native import().
const dynamicImportPi = (() => {
    if (process.env.VITEST) {
        return (specifier) => Promise.resolve(`${specifier}`).then(s => __importStar(require(s)));
    }
    // eslint-disable-next-line @typescript-eslint/no-implied-eval
    return new Function("id", "return import(id)");
})();
async function loadPiModule() {
    try {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        const mod = await dynamicImportPi(PI_MODULE_ID);
        return mod;
    }
    catch {
        throw new exceptions_1.BabysitterRuntimeError("PiModuleNotFound", "Cannot load @mariozechner/pi-coding-agent — is the package installed?", { category: exceptions_1.ErrorCategory.Configuration });
    }
}
function normalizeAzureOpenAiBaseUrl(value) {
    const trimmed = value.trim().replace(/\/+$/, "");
    if (!trimmed) {
        return trimmed;
    }
    try {
        const parsed = new URL(trimmed);
        const normalizedPath = parsed.pathname.replace(/\/+$/, "");
        const isAzureOpenAiHost = /\.openai\.azure\.com$/i.test(parsed.hostname);
        if (normalizedPath === "/openai/v1") {
            return `${parsed.origin}${normalizedPath}`;
        }
        if (normalizedPath === "/openai") {
            return `${parsed.origin}/openai/v1`;
        }
        if (isAzureOpenAiHost && (normalizedPath === "" || normalizedPath === "/")) {
            return `${parsed.origin}/openai/v1`;
        }
    }
    catch {
        // Fall back to the raw value when the URL cannot be parsed.
    }
    return trimmed;
}
function configureAzureOpenAiEnvDefaults(requestedModel) {
    const resourceName = process.env.AZURE_OPENAI_RESOURCE_NAME || process.env.AZURE_OPENAI_PROJECT_NAME;
    if (!process.env.AZURE_OPENAI_RESOURCE_NAME && process.env.AZURE_OPENAI_PROJECT_NAME) {
        process.env.AZURE_OPENAI_RESOURCE_NAME = process.env.AZURE_OPENAI_PROJECT_NAME;
    }
    if (process.env.AZURE_OPENAI_BASE_URL) {
        process.env.AZURE_OPENAI_BASE_URL = normalizeAzureOpenAiBaseUrl(process.env.AZURE_OPENAI_BASE_URL);
    }
    else if (resourceName) {
        process.env.AZURE_OPENAI_BASE_URL = normalizeAzureOpenAiBaseUrl(`https://${resourceName}.openai.azure.com`);
    }
    if (requestedModel &&
        !requestedModel.includes(":") &&
        process.env.AZURE_OPENAI_DEPLOYMENT &&
        !process.env.AZURE_OPENAI_DEPLOYMENT_NAME_MAP) {
        process.env.AZURE_OPENAI_DEPLOYMENT_NAME_MAP = `${requestedModel}=${process.env.AZURE_OPENAI_DEPLOYMENT}`;
    }
}
function synthesizeAzureModelEntry(modelId) {
    if (!process.env.AZURE_OPENAI_API_KEY) {
        return undefined;
    }
    const resourceName = process.env.AZURE_OPENAI_RESOURCE_NAME || process.env.AZURE_OPENAI_PROJECT_NAME;
    const baseUrl = process.env.AZURE_OPENAI_BASE_URL
        ? normalizeAzureOpenAiBaseUrl(process.env.AZURE_OPENAI_BASE_URL)
        : (resourceName ? normalizeAzureOpenAiBaseUrl(`https://${resourceName}.openai.azure.com`) : undefined);
    if (!baseUrl) {
        return undefined;
    }
    return {
        id: modelId,
        name: modelId,
        provider: "azure-openai-responses",
        api: "openai-responses",
        baseUrl,
        reasoning: true,
        input: ["text"],
        contextWindow: 128000,
        maxTokens: 16384,
        cost: {
            input: 0,
            output: 0,
            cacheRead: 0,
            cacheWrite: 0,
        },
    };
}
function extractAssistantFailure(messages) {
    if (!Array.isArray(messages) || messages.length === 0) {
        return undefined;
    }
    for (let index = messages.length - 1; index >= 0; index -= 1) {
        const message = messages[index];
        if (!message || typeof message !== "object") {
            continue;
        }
        const candidate = message;
        if (candidate.role !== "assistant") {
            continue;
        }
        if (candidate.stopReason === "error" && typeof candidate.errorMessage === "string" && candidate.errorMessage.trim()) {
            return candidate.errorMessage.trim();
        }
    }
    return undefined;
}
async function modelHasUsableAuth(registry, model) {
    if (typeof registry.hasConfiguredAuth === "function") {
        return registry.hasConfiguredAuth(model);
    }
    if (typeof registry.getApiKey === "function") {
        const key = await registry.getApiKey(model);
        return Boolean(key);
    }
    if (typeof registry.getApiKeyAndHeaders === "function") {
        const result = await registry.getApiKeyAndHeaders(model);
        return Boolean(result.ok &&
            ((typeof result.apiKey === "string" && result.apiKey.length > 0) ||
                (result.headers && Object.keys(result.headers).length > 0)));
    }
    if (typeof registry.getApiKeyForProvider === "function") {
        const key = await registry.getApiKeyForProvider(model.provider);
        return Boolean(key);
    }
    return false;
}
function pathExistsAsFile(filePath) {
    try {
        return (0, node_fs_1.statSync)(filePath).isFile();
    }
    catch {
        return false;
    }
}
function findInstructionTraversalRoot(startDir) {
    let current = path.resolve(startDir);
    for (;;) {
        if ((0, node_fs_1.existsSync)(path.join(current, ".git"))) {
            return current;
        }
        const parent = path.dirname(current);
        if (parent === current) {
            return path.resolve(startDir);
        }
        current = parent;
    }
}
function discoverRepoInstructionPrompts(startDir) {
    const resolvedStart = path.resolve(startDir);
    const traversalRoot = findInstructionTraversalRoot(resolvedStart);
    const directories = [];
    let current = resolvedStart;
    for (;;) {
        directories.push(current);
        if (current === traversalRoot) {
            break;
        }
        const parent = path.dirname(current);
        if (parent === current) {
            break;
        }
        current = parent;
    }
    const prompts = [];
    for (const dir of directories.reverse()) {
        const claudePath = path.join(dir, CLAUDE_MD_FILENAME);
        const agentsPath = path.join(dir, AGENTS_MD_FILENAME);
        const selectedPath = pathExistsAsFile(claudePath)
            ? claudePath
            : pathExistsAsFile(agentsPath)
                ? agentsPath
                : undefined;
        if (!selectedPath) {
            continue;
        }
        const relativeLabel = path.relative(traversalRoot, selectedPath) || path.basename(selectedPath);
        prompts.push([
            `Repository instructions from ${relativeLabel}:`,
            (0, node_fs_1.readFileSync)(selectedPath, "utf8"),
        ].join("\n"));
    }
    return prompts;
}
// ---------------------------------------------------------------------------
// PiSessionHandle
// ---------------------------------------------------------------------------
/**
 * Handle for interacting with the Pi coding agent programmatically.
 *
 * Wraps `AgentSession` from `@mariozechner/pi-coding-agent`. The underlying
 * session is created lazily on first `prompt()` call and reused thereafter.
 */
class PiSessionHandle {
    options;
    session = null;
    initPromise = null;
    cleanupTasks = [];
    constructor(options = {}) {
        this.options = options;
    }
    /**
     * Initialize the underlying AgentSession.
     *
     * Called automatically by `prompt()` if the session hasn't been created yet.
     * Safe to call multiple times — subsequent calls are no-ops.
     */
    async initialize() {
        if (this.session)
            return;
        if (this.initPromise) {
            await this.initPromise;
            return;
        }
        this.initPromise = this.doInitialize().catch((err) => {
            this.initPromise = null;
            throw err;
        });
        await this.initPromise;
    }
    /**
     * Send a prompt to the Pi agent and wait for completion.
     *
     * Initializes the session if needed, sends the prompt, waits for the
     * `agent_end` event, and returns collected output.
     */
    async prompt(text, timeout) {
        await this.initialize();
        const session = this.requireSession();
        const effectiveTimeout = timeout ?? this.options.timeout ?? DEFAULT_TIMEOUT_MS;
        const start = Date.now();
        return new Promise((resolve, reject) => {
            let settled = false;
            let timer;
            let agentEndGraceTimer;
            let agentEndResult = null;
            let promptSettled = false;
            const finishWithResult = (result) => {
                if (settled)
                    return;
                settled = true;
                if (timer)
                    clearTimeout(timer);
                if (agentEndGraceTimer)
                    clearTimeout(agentEndGraceTimer);
                resolve(result);
            };
            const finishWithPromptError = (err) => {
                if (settled)
                    return;
                settled = true;
                if (timer)
                    clearTimeout(timer);
                if (agentEndGraceTimer)
                    clearTimeout(agentEndGraceTimer);
                const message = err instanceof Error ? err.message : String(err);
                resolve({
                    output: message,
                    exitCode: 1,
                    duration: Date.now() - start,
                    success: false,
                });
            };
            // Set up timeout
            if (effectiveTimeout > 0) {
                timer = setTimeout(() => {
                    if (settled)
                        return;
                    settled = true;
                    unsubscribe();
                    void session.abort();
                    reject(new exceptions_1.BabysitterRuntimeError("PiTimeoutError", `Pi prompt timed out after ${effectiveTimeout}ms`, { category: exceptions_1.ErrorCategory.External }));
                }, effectiveTimeout);
            }
            // Subscribe to events to detect completion
            const unsubscribe = session.subscribe((event) => {
                if (settled)
                    return;
                if (event.type === "agent_end") {
                    const messages = Array.isArray(event.messages)
                        ? event.messages
                        : undefined;
                    unsubscribe();
                    const assistantFailure = extractAssistantFailure(messages);
                    const assistantText = session.getLastAssistantText();
                    const output = assistantText && assistantText.trim().length > 0
                        ? assistantText
                        : assistantFailure ?? "";
                    agentEndResult = {
                        output,
                        exitCode: assistantFailure ? 1 : 0,
                        duration: Date.now() - start,
                        success: !assistantFailure,
                    };
                    if (promptSettled) {
                        finishWithResult(agentEndResult);
                        return;
                    }
                    agentEndGraceTimer = setTimeout(() => {
                        if (agentEndResult) {
                            finishWithResult(agentEndResult);
                        }
                    }, AGENT_END_PROMPT_SETTLE_GRACE_MS);
                }
            });
            // Fire the prompt — errors are caught and resolved as failures
            session.prompt(text)
                .then(() => {
                promptSettled = true;
                if (agentEndResult) {
                    finishWithResult(agentEndResult);
                }
            })
                .catch((err) => {
                promptSettled = true;
                if (agentEndResult) {
                    finishWithResult(agentEndResult);
                    return;
                }
                unsubscribe();
                finishWithPromptError(err);
            });
        });
    }
    /**
     * Steer the running agent with an instruction.
     *
     * Steering messages are delivered immediately to the agent while it is
     * processing a prompt.
     */
    async steer(text) {
        const session = this.requireSession();
        await session.steer(text);
    }
    /**
     * Queue a follow-up message for after the current turn completes.
     */
    async followUp(text) {
        const session = this.requireSession();
        await session.followUp(text);
    }
    /**
     * Subscribe to session events.
     *
     * Returns an unsubscribe function.
     */
    subscribe(listener) {
        const session = this.requireSession();
        return session.subscribe(listener);
    }
    /**
     * Execute a bash command through the agent's sandbox.
     */
    async executeBash(command, onChunk) {
        await this.initialize();
        const session = this.requireSession();
        const result = await session.executeBash(command, onChunk);
        return {
            output: result.output,
            exitCode: result.exitCode,
            cancelled: result.cancelled,
        };
    }
    /**
     * Abort the current prompt execution.
     */
    async abort() {
        if (this.session) {
            await this.session.abort();
        }
    }
    /**
     * Dispose of the session and release resources.
     */
    dispose() {
        if (this.session) {
            const session = this.session;
            if (session.isStreaming) {
                void session.abort().catch(() => undefined);
            }
            session.dispose();
            this.session = null;
            this.initPromise = null;
        }
        while (this.cleanupTasks.length > 0) {
            const cleanup = this.cleanupTasks.pop();
            if (cleanup) {
                void cleanup();
            }
        }
    }
    /** The underlying pi session ID, if initialized. */
    get sessionId() {
        return this.session?.sessionId;
    }
    /** Whether the session is currently streaming a response. */
    get isStreaming() {
        return this.session?.isStreaming ?? false;
    }
    /** Whether the session has been initialized. */
    get isInitialized() {
        return this.session !== null;
    }
    // ---------------------------------------------------------------------------
    // Internal helpers
    // ---------------------------------------------------------------------------
    async doInitialize() {
        const mod = await loadPiModule();
        // Bridge common Azure env var aliases that pi-coding-agent doesn't know
        // about.  Pi expects AZURE_OPENAI_RESOURCE_NAME; the user's profile may
        // set AZURE_OPENAI_PROJECT_NAME instead.
        configureAzureOpenAiEnvDefaults(typeof this.options.model === "string" ? this.options.model : undefined);
        const createOpts = {};
        const cwd = this.options.workspace ?? process.cwd();
        const compressionConfig = loadCompressionConfigSafe(cwd);
        const compactionEnabled = this.options.enableCompaction ??
            Boolean(compressionConfig?.enabled && compressionConfig.layers.sdkContextHook.enabled);
        const compactionSettings = buildCompactionSettings(compactionEnabled);
        createOpts.cwd = cwd;
        if (this.options.agentDir)
            createOpts.agentDir = this.options.agentDir;
        if (this.options.thinkingLevel)
            createOpts.thinkingLevel = this.options.thinkingLevel;
        if (this.options.customTools)
            createOpts.customTools = this.options.customTools;
        if (this.options.ephemeral) {
            createOpts.sessionManager = mod.SessionManager.inMemory();
        }
        const secureBashBackend = this.options.toolsMode === "coding" || this.options.toolsMode === "readonly"
            ? await (0, piSecureSandbox_1.createSecureBashBackend)({
                workspace: cwd,
                mode: this.options.bashSandbox ?? DEFAULT_BASH_SANDBOX_MODE,
            })
            : null;
        if (secureBashBackend) {
            this.cleanupTasks.push(() => secureBashBackend.dispose());
        }
        const toolOptions = secureBashBackend
            ? {
                bash: {
                    operations: secureBashBackend.operations,
                },
            }
            : undefined;
        if (this.options.toolsMode === "coding") {
            createOpts.tools = mod.createCodingTools
                ? mod.createCodingTools(cwd, toolOptions)
                : mod.codingTools;
        }
        else if (this.options.toolsMode === "readonly") {
            createOpts.tools = mod.createReadOnlyTools
                ? mod.createReadOnlyTools(cwd, toolOptions)
                : mod.readOnlyTools;
        }
        const appendedSystemPrompt = [
            ...discoverRepoInstructionPrompts(cwd),
            ...(this.options.appendSystemPrompt ?? []),
            ...(secureBashBackend ? [secureBashBackend.promptNote] : []),
        ];
        if (this.options.systemPrompt ||
            appendedSystemPrompt.length > 0 ||
            this.options.isolated ||
            compactionEnabled) {
            const settingsManager = mod.SettingsManager.inMemory({
                quietStartup: true,
                compaction: compactionSettings.compaction,
                branchSummary: compactionSettings.branchSummary,
            });
            createOpts.settingsManager = settingsManager;
            if (this.options.systemPrompt ||
                appendedSystemPrompt.length > 0 ||
                this.options.isolated) {
                const resourceLoader = new mod.DefaultResourceLoader({
                    cwd,
                    agentDir: this.options.agentDir,
                    settingsManager,
                    noExtensions: this.options.isolated === true,
                    noSkills: this.options.isolated === true,
                    noPromptTemplates: this.options.isolated === true,
                    noThemes: this.options.isolated === true,
                    agentsFilesOverride: this.options.isolated === true
                        ? () => ({ agentsFiles: [] })
                        : undefined,
                    systemPromptOverride: this.options.systemPrompt
                        ? () => this.options.systemPrompt
                        : undefined,
                    appendSystemPromptOverride: appendedSystemPrompt.length > 0
                        ? (base) => [...base, ...appendedSystemPrompt]
                        : undefined,
                });
                await resourceLoader.reload();
                createOpts.resourceLoader = resourceLoader;
            }
        }
        // Resolve model string to a model object from pi's ModelRegistry.
        // The `createAgentSession` API expects a model object (with provider,
        // api, baseUrl, etc.), not a plain string.  We accept formats:
        //   "provider:modelId"  e.g. "azure-openai-responses:gpt-4.1"
        //   "modelId"           e.g. "gpt-4.1" (searches all providers)
        if (typeof this.options.model === "string") {
            const modelStr = this.options.model;
            const auth = mod.AuthStorage.create();
            const registry = new mod.ModelRegistry(auth);
            let resolved;
            if (modelStr.includes(":")) {
                const [provider, modelId] = modelStr.split(":", 2);
                resolved = registry.find(provider, modelId);
            }
            if (!resolved) {
                // Search all models for a matching id
                const all = registry.getAll();
                // Prefer models with a working API key
                for (const m of all) {
                    if (m.id === modelStr || m.id === modelStr.split(":").pop()) {
                        if (await modelHasUsableAuth(registry, m)) {
                            resolved = m;
                            break;
                        }
                    }
                }
            }
            if (!resolved) {
                resolved = synthesizeAzureModelEntry(modelStr);
            }
            if (resolved) {
                createOpts.model = resolved;
            }
            // If not resolved, let createAgentSession handle default model selection
        }
        try {
            const { session } = await mod.createAgentSession(createOpts);
            if (this.options.uiContext && typeof session.bindExtensions === "function") {
                await session.bindExtensions({ uiContext: this.options.uiContext });
            }
            this.session = session;
        }
        catch (error) {
            while (this.cleanupTasks.length > 0) {
                const cleanup = this.cleanupTasks.pop();
                if (cleanup) {
                    await cleanup();
                }
            }
            throw error;
        }
    }
    requireSession() {
        if (!this.session) {
            throw new exceptions_1.BabysitterRuntimeError("PiSessionNotInitialized", "Pi session has not been initialized — call initialize() or prompt() first", { category: exceptions_1.ErrorCategory.Runtime });
        }
        return this.session;
    }
}
exports.PiSessionHandle = PiSessionHandle;
/**
 * Create a new Pi session handle.
 *
 * The underlying `AgentSession` is created lazily on first use.
 */
function createPiSession(options) {
    return new PiSessionHandle(options);
}
function loadCompressionConfigSafe(cwd) {
    try {
        return (0, config_loader_1.loadCompressionConfig)(cwd);
    }
    catch {
        return null;
    }
}
function readPositiveIntegerEnv(name, fallback) {
    const raw = process.env[name];
    if (!raw)
        return fallback;
    const parsed = Number.parseInt(raw, 10);
    return Number.isFinite(parsed) && parsed > 0 ? parsed : fallback;
}
function buildCompactionSettings(enabled) {
    if (!enabled) {
        return {
            compaction: { enabled: false },
        };
    }
    return {
        compaction: {
            enabled: true,
            reserveTokens: readPositiveIntegerEnv("BABYSITTER_PI_COMPACTION_RESERVE_TOKENS", 8_192),
            keepRecentTokens: readPositiveIntegerEnv("BABYSITTER_PI_COMPACTION_KEEP_RECENT_TOKENS", 12_288),
        },
        branchSummary: {
            reserveTokens: readPositiveIntegerEnv("BABYSITTER_PI_BRANCH_SUMMARY_RESERVE_TOKENS", 4_096),
            skipPrompt: false,
        },
    };
}
