// @a5c-ai/babysitter-sdk/dist/harness/registry.js
"use strict";
/**
 * Harness adapter registry with auto-detection.
 *
 * Maintains a lazy singleton of the active adapter. On first access,
 * probes registered adapters via `isActive()` and returns the first match
 * (or the null adapter if none match).
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.detectAdapter = detectAdapter;
exports.getAdapterByName = getAdapterByName;
exports.listSupportedHarnesses = listSupportedHarnesses;
exports.getAdapter = getAdapter;
exports.setAdapter = setAdapter;
exports.resetAdapter = resetAdapter;
const claudeCode_1 = require("./claudeCode");
const codex_1 = require("./codex");
const geminiCli_1 = require("./geminiCli");
const pi_1 = require("./pi");
const ohMyPi_1 = require("./ohMyPi");
const internal_1 = require("./internal");
const cursor_1 = require("./cursor");
const githubCopilot_1 = require("./githubCopilot");
const opencode_1 = require("./opencode");
const customAdapter_1 = require("./customAdapter");
// ---------------------------------------------------------------------------
// Registry of known adapters (ordered by priority)
// ---------------------------------------------------------------------------
const knownAdapters = [
    (0, codex_1.createCodexAdapter)(),
    (0, ohMyPi_1.createOhMyPiAdapter)(),
    (0, internal_1.createInternalAdapter)(),
    (0, pi_1.createPiAdapter)(),
    (0, opencode_1.createOpenCodeAdapter)(),
    (0, claudeCode_1.createClaudeCodeAdapter)(),
    (0, geminiCli_1.createGeminiCliAdapter)(),
    (0, cursor_1.createCursorAdapter)(),
    (0, githubCopilot_1.createGithubCopilotAdapter)(),
    (0, customAdapter_1.createCustomAdapter)(),
];
// ---------------------------------------------------------------------------
// Auto-detection
// ---------------------------------------------------------------------------
/**
 * Probe each registered adapter and return the first that reports active.
 * Falls back to the custom adapter (which requires explicit args).
 */
function detectAdapter() {
    for (const adapter of knownAdapters) {
        if (adapter.isActive())
            return adapter;
    }
    return (0, customAdapter_1.createCustomAdapter)();
}
/**
 * Look up an adapter by harness name (e.g. "claude-code").
 * Returns null if the name is not recognized.
 */
function getAdapterByName(name) {
    for (const adapter of knownAdapters) {
        if (adapter.name === name)
            return adapter;
    }
    return null;
}
/**
 * List the names of all supported harnesses.
 */
function listSupportedHarnesses() {
    return knownAdapters.map((a) => a.name);
}
// ---------------------------------------------------------------------------
// Lazy singleton
// ---------------------------------------------------------------------------
let current = null;
/**
 * Get the active harness adapter (auto-detected on first call).
 */
function getAdapter() {
    if (!current) {
        current = detectAdapter();
    }
    return current;
}
/**
 * Override the active adapter (useful for testing).
 */
function setAdapter(adapter) {
    current = adapter;
}
/**
 * Reset the singleton so the next `getAdapter()` call re-detects.
 */
function resetAdapter() {
    current = null;
}
