// @a5c-ai/babysitter-sdk/dist/harness/types.js
"use strict";
/**
 * Harness adapter interface.
 *
 * A "harness" is the host tool that invokes the babysitter SDK (e.g. Claude Code,
 * Cursor, Windsurf). Each harness has its own session lifecycle, env vars, and
 * hook input/output formats. The adapter interface abstracts these differences
 * so the SDK core remains harness-agnostic.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HarnessCapability = void 0;
// ---------------------------------------------------------------------------
// Harness capability enum
// ---------------------------------------------------------------------------
/** Capabilities that a harness adapter may support. */
var HarnessCapability;
(function (HarnessCapability) {
    /** Harness supports programmatic (non-interactive) invocation. */
    HarnessCapability["Programmatic"] = "programmatic";
    /** Harness can bind a babysitter run to a host session. */
    HarnessCapability["SessionBinding"] = "session-binding";
    /** Harness implements the stop-hook lifecycle event. */
    HarnessCapability["StopHook"] = "stop-hook";
    /** Harness exposes an MCP (Model Context Protocol) server. */
    HarnessCapability["Mcp"] = "mcp";
    /** Harness can accept a prompt without a TTY (headless mode). */
    HarnessCapability["HeadlessPrompt"] = "headless-prompt";
})(HarnessCapability || (exports.HarnessCapability = HarnessCapability = {}));
