// @a5c-ai/babysitter-sdk/dist/cli/commands/compressionStatus.js
"use strict";
/**
 * compression:status — print the resolved compression config as a table or JSON.
 *
 * Usage:
 *   babysitter compression:status [--json]
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleCompressionStatus = handleCompressionStatus;
const path = __importStar(require("node:path"));
const config_loader_1 = require("../../compression/config-loader");
function resolveProjectDir(cwd) {
    return cwd ?? process.cwd();
}
function buildRows(config) {
    const { layers } = config;
    return [
        {
            layer: 'userPromptHook',
            enabled: layers.userPromptHook.enabled,
            engine: layers.userPromptHook.engine,
            keySettings: `threshold=${layers.userPromptHook.threshold}, keepRatio=${layers.userPromptHook.keepRatio}`,
        },
        {
            layer: 'commandOutputHook',
            enabled: layers.commandOutputHook.enabled,
            engine: layers.commandOutputHook.engine,
            keySettings: `excludeCommands=[${layers.commandOutputHook.excludeCommands.join(', ')}]`,
        },
        {
            layer: 'sdkContextHook',
            enabled: layers.sdkContextHook.enabled,
            engine: layers.sdkContextHook.engine,
            keySettings: `targetReduction=${layers.sdkContextHook.targetReduction}, minTokens=${layers.sdkContextHook.minCompressionTokens}`,
        },
        {
            layer: 'processLibraryCache',
            enabled: layers.processLibraryCache.enabled,
            engine: layers.processLibraryCache.engine,
            keySettings: `targetReduction=${layers.processLibraryCache.targetReduction}, ttlHours=${layers.processLibraryCache.ttlHours}`,
        },
    ];
}
function printTable(masterEnabled, rows) {
    console.log(`\nCompression master switch: ${masterEnabled ? 'ENABLED' : 'DISABLED'}\n`);
    const col = { layer: 22, enabled: 9, engine: 22, keySettings: 60 };
    const header = [
        'Layer'.padEnd(col.layer),
        'Enabled'.padEnd(col.enabled),
        'Engine'.padEnd(col.engine),
        'Key settings',
    ].join('  ');
    const separator = [
        '-'.repeat(col.layer),
        '-'.repeat(col.enabled),
        '-'.repeat(col.engine),
        '-'.repeat(col.keySettings),
    ].join('  ');
    console.log(header);
    console.log(separator);
    for (const row of rows) {
        const line = [
            row.layer.padEnd(col.layer),
            String(row.enabled).padEnd(col.enabled),
            row.engine.padEnd(col.engine),
            row.keySettings,
        ].join('  ');
        console.log(line);
    }
    console.log('');
}
function handleCompressionStatus(opts) {
    const projectDir = resolveProjectDir(opts.cwd);
    let config;
    try {
        config = (0, config_loader_1.loadCompressionConfig)(projectDir);
    }
    catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        if (opts.json) {
            console.error(JSON.stringify({ error: `Failed to load compression config: ${message}` }));
        }
        else {
            console.error(`Error: Failed to load compression config: ${message}`);
        }
        return 1;
    }
    const rows = buildRows(config);
    if (opts.json) {
        console.log(JSON.stringify({
            masterEnabled: config.enabled,
            configFile: path.join(projectDir, '.a5c', 'compression.config.json'),
            layers: rows,
        }, null, 2));
    }
    else {
        printTable(config.enabled, rows);
    }
    return 0;
}
