// @a5c-ai/babysitter-sdk/dist/cli/commands/compressionSet.js
"use strict";
/**
 * compression:set — update a specific compression config value in the project config.
 *
 * Usage:
 *   babysitter compression:set <layer.key> <value> [--json]
 *
 * Examples:
 *   babysitter compression:set userPromptHook.threshold 300
 *   babysitter compression:set userPromptHook.keepRatio 0.85
 *   babysitter compression:set sdkContextHook.targetReduction 0.2
 *   babysitter compression:set sdkContextHook.minCompressionTokens 100
 *   babysitter compression:set commandOutputHook.excludeCommands node,python,ruby
 *   babysitter compression:set processLibraryCache.targetReduction 0.3
 *   babysitter compression:set processLibraryCache.ttlHours 48
 *   babysitter compression:set enabled false
 *
 * Writes the change to .a5c/compression.config.json in the current project.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleCompressionSet = handleCompressionSet;
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
// ---------------------------------------------------------------------------
// Known settable fields
// ---------------------------------------------------------------------------
const KNOWN_FIELDS = [
    { layer: null, field: 'enabled', type: 'boolean', description: 'Master compression switch' },
    { layer: 'userPromptHook', field: 'threshold', type: 'number', description: 'Min tokens before compression triggers' },
    { layer: 'userPromptHook', field: 'keepRatio', type: 'number', description: 'Fraction of tokens to keep (0–1)' },
    { layer: 'commandOutputHook', field: 'excludeCommands', type: 'string[]', description: 'Comma-separated command names to skip compression' },
    { layer: 'sdkContextHook', field: 'targetReduction', type: 'number', description: 'Fraction of SDK context to remove (0–1)' },
    { layer: 'sdkContextHook', field: 'minCompressionTokens', type: 'number', description: 'Min tokens before SDK context compression triggers' },
    { layer: 'processLibraryCache', field: 'targetReduction', type: 'number', description: 'Fraction of library cache to remove (0–1)' },
    { layer: 'processLibraryCache', field: 'ttlHours', type: 'number', description: 'How many hours a compressed cache entry is valid' },
];
function findField(dotKey) {
    const parts = dotKey.split('.');
    if (parts.length === 1) {
        return KNOWN_FIELDS.find(f => f.layer === null && f.field === parts[0]);
    }
    if (parts.length === 2) {
        return KNOWN_FIELDS.find(f => f.layer === parts[0] && f.field === parts[1]);
    }
    return undefined;
}
function validKeysText() {
    return KNOWN_FIELDS.map(f => (f.layer ? `${f.layer}.${f.field}` : f.field)).join(', ');
}
// ---------------------------------------------------------------------------
// Value parsing
// ---------------------------------------------------------------------------
function parseValue(raw, type, key) {
    switch (type) {
        case 'number': {
            const n = Number(raw);
            if (isNaN(n))
                throw new Error(`"${key}" expects a number, got: ${raw}`);
            return n;
        }
        case 'boolean': {
            const v = raw.trim().toLowerCase();
            if (v === 'true' || v === '1' || v === 'yes' || v === 'on')
                return true;
            if (v === 'false' || v === '0' || v === 'no' || v === 'off')
                return false;
            throw new Error(`"${key}" expects true/false, got: ${raw}`);
        }
        case 'string[]':
            return raw.split(',').map(s => s.trim()).filter(Boolean);
    }
}
// ---------------------------------------------------------------------------
// Config file helpers
// ---------------------------------------------------------------------------
function configFilePath(projectDir) {
    return path.join(projectDir, '.a5c', 'compression.config.json');
}
async function readExistingConfig(filePath) {
    try {
        const raw = await node_fs_1.promises.readFile(filePath, 'utf8');
        const parsed = JSON.parse(raw);
        if (typeof parsed === 'object' && parsed !== null && !Array.isArray(parsed)) {
            return parsed;
        }
    }
    catch {
        // File doesn't exist or is malformed — start fresh
    }
    return {};
}
// ---------------------------------------------------------------------------
// Handler
// ---------------------------------------------------------------------------
async function handleCompressionSet(opts) {
    const projectDir = opts.cwd ?? process.cwd();
    const { key, value } = opts;
    const field = findField(key);
    if (!field) {
        const msg = `Unknown key: "${key}". Valid keys: ${validKeysText()}`;
        if (opts.json) {
            process.stderr.write(JSON.stringify({ error: msg }) + "\n");
        }
        else {
            process.stderr.write(`Error: ${msg}\n`);
        }
        return 1;
    }
    let parsed;
    try {
        parsed = parseValue(value, field.type, key);
    }
    catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        if (opts.json) {
            process.stderr.write(JSON.stringify({ error: msg }) + "\n");
        }
        else {
            process.stderr.write(`Error: ${msg}\n`);
        }
        return 1;
    }
    const filePath = configFilePath(projectDir);
    await node_fs_1.promises.mkdir(path.dirname(filePath), { recursive: true });
    const existing = await readExistingConfig(filePath);
    // Apply the value at the right depth
    if (field.layer === null) {
        existing[field.field] = parsed;
    }
    else {
        const layers = (existing.layers ?? {});
        const layerObj = (layers[field.layer] ?? {});
        layerObj[field.field] = parsed;
        layers[field.layer] = layerObj;
        existing.layers = layers;
    }
    await node_fs_1.promises.writeFile(filePath, JSON.stringify(existing, null, 2) + '\n', 'utf8');
    if (opts.json) {
        process.stdout.write(JSON.stringify({ key, value: parsed, configFile: filePath }) + "\n");
    }
    else {
        process.stdout.write(`Set ${key} = ${JSON.stringify(parsed)}\n`);
        process.stdout.write(`Config written to: ${filePath}\n`);
    }
    return 0;
}
