// @a5c-ai/babysitter-sdk/dist/cli/commands/harnessInstall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleHarnessInstall = handleHarnessInstall;
exports.handleHarnessInstallPlugin = handleHarnessInstallPlugin;
exports.formatHarnessInstallError = formatHarnessInstallError;
const harness_1 = require("../../harness");
const exceptions_1 = require("../../runtime/exceptions");
function supportsColors() {
    if (process.env.NO_COLOR !== undefined)
        return false;
    if (process.env.FORCE_COLOR !== undefined)
        return true;
    return Boolean(process.stderr?.isTTY);
}
function formatInstallResult(payload, json) {
    if (json) {
        console.log(JSON.stringify(payload, null, 2));
        return;
    }
    if (payload.warning) {
        console.log(`Warning: ${payload.warning}`);
    }
    if (payload.summary) {
        console.log(payload.summary);
    }
    if (payload.command) {
        console.log(`Command: ${payload.command}`);
    }
    if (payload.output) {
        console.log(payload.output);
    }
    if (payload.location) {
        console.log(`Location: ${payload.location}`);
    }
}
function requireHarnessName(harnessName, commandName) {
    if (!harnessName) {
        throw new exceptions_1.BabysitterRuntimeError("MissingArgument", `${commandName} requires a harness name as the first argument`, {
            category: exceptions_1.ErrorCategory.Validation,
            suggestions: [`babysitter ${commandName} codex`],
        });
    }
    return harnessName;
}
function requireInstallMethod(harnessName, commandName, kind) {
    const adapter = (0, harness_1.getAdapterByName)(harnessName);
    if (!adapter) {
        throw new exceptions_1.BabysitterRuntimeError("UnsupportedHarnessInstall", `${commandName} does not support "${harnessName}" yet.`, {
            category: exceptions_1.ErrorCategory.Validation,
            suggestions: [
                "Supported harnesses: claude-code, codex, gemini-cli, pi, oh-my-pi",
            ],
        });
    }
    const run = adapter[kind];
    if (!run) {
        throw new exceptions_1.BabysitterRuntimeError("UnsupportedHarnessInstall", `${commandName} does not support "${harnessName}" yet.`, {
            category: exceptions_1.ErrorCategory.Validation,
            suggestions: [
                "Supported harnesses: claude-code, codex, gemini-cli, pi, oh-my-pi",
            ],
        });
    }
    return {
        adapter,
        run: run.bind(adapter),
    };
}
async function handleHarnessInstall(args) {
    const harnessName = requireHarnessName(args.harnessName, "harness:install");
    const { run } = requireInstallMethod(harnessName, "harness:install", "installHarness");
    const result = await run(args);
    formatInstallResult(result, args.json);
    return 0;
}
async function handleHarnessInstallPlugin(args) {
    const harnessName = requireHarnessName(args.harnessName, "harness:install-plugin");
    const { run } = requireInstallMethod(harnessName, "harness:install-plugin", "installPlugin");
    const result = await run(args);
    formatInstallResult(result, args.json);
    return 0;
}
function formatHarnessInstallError(error, json) {
    const err = error instanceof Error
        ? error
        : new Error(String(error));
    if (json) {
        console.error(JSON.stringify((0, exceptions_1.toStructuredError)(err), null, 2));
    }
    else {
        console.error((0, exceptions_1.formatErrorWithContext)(err, { colors: supportsColors() }));
    }
    return 1;
}
