// @a5c-ai/babysitter-sdk/dist/cli/commands/harnessResumeRun.js
"use strict";
/**
 * harness:resume-run command handler.
 *
 * Uses an agentic Pi session to discover existing runs, present them to the
 * user, assess state, and resume orchestration by delegating to
 * handleHarnessCreateRun.
 *
 * The Pi agent has access to three domain-specific tools
 * (babysitter_list_runs, babysitter_assess_run, babysitter_resume_run) plus
 * the full agentic tool set (read/write/edit/grep/bash/etc.) for ad-hoc
 * inspection of run directories.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleHarnessResumeRun = handleHarnessResumeRun;
const node_fs_1 = require("node:fs");
const path = __importStar(require("node:path"));
const typebox_1 = require("@sinclair/typebox");
const runFiles_1 = require("../../storage/runFiles");
const journal_1 = require("../../storage/journal");
const stateCache_1 = require("../../runtime/replay/stateCache");
const piWrapper_1 = require("../../harness/piWrapper");
const agenticTools_1 = require("../../harness/agenticTools");
const harnessCreateRun_1 = require("./harnessCreateRun");
const DIM = "\x1b[2m";
const RESET = "\x1b[0m";
const BOLD = "\x1b[1m";
const RED = "\x1b[31m";
const MAGENTA = "\x1b[35m";
const VERBOSE_LOG_LIMIT = 4_000;
// ---------------------------------------------------------------------------
// Utility functions (retained — used as backing implementations for tools)
// ---------------------------------------------------------------------------
/**
 * Determine high-level run status from journal events and state cache.
 */
function deriveRunStatus(journal, pendingEffects) {
    const types = new Set(journal.map((e) => e.type));
    if (types.has("RUN_COMPLETED"))
        return "completed";
    if (types.has("RUN_FAILED"))
        return "failed";
    const totalPending = Object.values(pendingEffects).reduce((a, b) => a + b, 0);
    if (totalPending > 0)
        return "waiting";
    if (types.has("EFFECT_REQUESTED"))
        return "in-progress";
    if (types.has("RUN_CREATED"))
        return "created";
    return "unknown";
}
/**
 * Discover runs from the runs directory.
 */
async function discoverRuns(runsDir, verbose) {
    let entries;
    try {
        entries = await node_fs_1.promises.readdir(runsDir);
    }
    catch (err) {
        const e = err;
        if (e.code === "ENOENT")
            return [];
        throw err;
    }
    const runs = [];
    for (const entry of entries) {
        const runDir = path.join(runsDir, entry);
        try {
            const stat = await node_fs_1.promises.stat(runDir);
            if (!stat.isDirectory())
                continue;
            const metadata = await (0, runFiles_1.readRunMetadata)(runDir);
            const journal = await (0, journal_1.loadJournal)(runDir);
            const stateCache = await (0, stateCache_1.readStateCache)(runDir);
            const pendingEffects = stateCache?.pendingEffectsByKind ?? {};
            const totalEffects = stateCache
                ? Object.keys(stateCache.effectsByInvocation).length
                : 0;
            const totalPending = Object.values(pendingEffects).reduce((a, b) => a + b, 0);
            const resolvedEffects = totalEffects - totalPending;
            const status = deriveRunStatus(journal, pendingEffects);
            runs.push({
                runId: metadata.runId,
                runDir,
                processId: metadata.processId,
                prompt: metadata.prompt,
                createdAt: metadata.createdAt,
                status,
                pendingEffects,
                totalEffects,
                resolvedEffects,
                entrypoint: metadata.entrypoint,
            });
        }
        catch {
            if (verbose) {
                process.stderr.write(`${DIM}  skipping ${entry}: unable to read metadata${RESET}\n`);
            }
        }
    }
    // Sort by createdAt descending
    runs.sort((a, b) => b.createdAt.localeCompare(a.createdAt));
    return runs.slice(0, 20);
}
/**
 * Build a detailed state assessment for a single run.
 */
async function assessRun(runDir) {
    const metadata = await (0, runFiles_1.readRunMetadata)(runDir);
    const journal = await (0, journal_1.loadJournal)(runDir);
    const stateCache = await (0, stateCache_1.readStateCache)(runDir);
    const pendingEffects = stateCache?.pendingEffectsByKind ?? {};
    const totalEffects = stateCache
        ? Object.keys(stateCache.effectsByInvocation).length
        : 0;
    const totalPending = Object.values(pendingEffects).reduce((a, b) => a + b, 0);
    const resolvedEffects = totalEffects - totalPending;
    const status = deriveRunStatus(journal, pendingEffects);
    const lastEvent = journal.length > 0
        ? { type: journal[journal.length - 1].type, recordedAt: journal[journal.length - 1].recordedAt }
        : null;
    return {
        run: {
            runId: metadata.runId,
            runDir,
            processId: metadata.processId,
            prompt: metadata.prompt,
            createdAt: metadata.createdAt,
            status,
            pendingEffects,
            totalEffects,
            resolvedEffects,
            entrypoint: metadata.entrypoint,
        },
        journalLength: journal.length,
        lastEvent,
    };
}
// ---------------------------------------------------------------------------
// Verbose / progress helpers
// ---------------------------------------------------------------------------
function writeVerboseLine(enabled, json, message) {
    if (!json && enabled) {
        process.stderr.write(`${DIM}${message}${RESET}\n`);
    }
}
function writeVerboseBlock(enabled, json, label, value, maxChars = VERBOSE_LOG_LIMIT) {
    if (json || !enabled) {
        return;
    }
    let raw;
    try {
        raw = JSON.stringify(value, null, 2);
    }
    catch {
        raw = String(value);
    }
    if (raw.length > maxChars) {
        raw = raw.slice(0, maxChars) + `... (truncated, ${raw.length} chars total)`;
    }
    process.stderr.write(`${DIM}[${label}]${RESET}\n${DIM}${raw}${RESET}\n`);
}
function formatToolResult(data, message) {
    if (typeof data === "string") {
        return {
            content: [{
                    type: "text",
                    text: message ? `${message}\n${data}` : data,
                }],
            details: data,
        };
    }
    const content = message
        ? `${message}\n${JSON.stringify(data, null, 2)}`
        : JSON.stringify(data, null, 2);
    return {
        content: [{
                type: "text",
                text: content,
            }],
        details: data,
    };
}
function errorResult(message) {
    return {
        content: [{ type: "text", text: `Error: ${message}` }],
    };
}
// ---------------------------------------------------------------------------
// System prompt
// ---------------------------------------------------------------------------
function buildResumeSystemPrompt(runsDir, runIdHint) {
    const parts = [
        "You are an assistant that helps users find and resume babysitter orchestration runs.",
        "",
        "## Your workflow",
        "",
        "1. **Discover runs** -- call `babysitter_list_runs` to list recent runs.",
        runIdHint
            ? `2. The user has requested run ID "${runIdHint}". Call \`babysitter_assess_run\` with that run ID to inspect it.`
            : "2. Present the runs to the user. If there are multiple, use `AskUserQuestion` to let them pick one.",
        "3. **Assess** the selected run by calling `babysitter_assess_run` with the run ID.",
        "4. Show the user the run's state (status, pending effects, journal summary).",
        "5. If the run is resumable (not completed), call `babysitter_resume_run` to resume it.",
        "6. If the run is already completed, tell the user and stop.",
        "",
        "## Important rules",
        "",
        "- Always call `babysitter_list_runs` first to see what's available.",
        "- Use `AskUserQuestion` for interactive selection when there are multiple runs.",
        "- You may also use read, grep, bash and other tools to inspect run files in " + runsDir + " if more detail is needed.",
        "- Call `babysitter_resume_run` exactly once when ready to resume.",
        "- Do NOT attempt to modify run files yourself -- only use the provided resume tool.",
    ];
    return parts.join("\n");
}
function buildResumeUserPrompt(runIdHint) {
    if (runIdHint) {
        return `Resume the run with ID "${runIdHint}". Assess it and proceed.`;
    }
    return "Help me find and resume a babysitter run. Start by listing the available runs.";
}
// ---------------------------------------------------------------------------
// Main handler
// ---------------------------------------------------------------------------
async function handleHarnessResumeRun(args) {
    const { runsDir, json, verbose, interactive = true, } = args;
    const writeVerbose = (message) => {
        writeVerboseLine(verbose, json, message);
    };
    const writeVerboseData = (label, value, maxChars) => {
        writeVerboseBlock(verbose, json, label, value, maxChars);
    };
    // Track whether the agent has triggered a resume (set by the resume tool).
    let resumeTriggered = false;
    let resumeExitCode = 0;
    // -----------------------------------------------------------------
    // Define domain-specific tools
    // -----------------------------------------------------------------
    const customTools = [
        {
            name: "babysitter_list_runs",
            label: "List Runs",
            description: "List recent runs in the runs directory with status and metadata. " +
                "Returns up to 20 runs sorted by most recent first.",
            parameters: typebox_1.Type.Object({
                statusFilter: typebox_1.Type.Optional(typebox_1.Type.String({ description: "Optional status filter: created, in-progress, waiting, failed, completed" })),
            }),
            execute: async (_toolCallId, params) => {
                writeVerboseData("resume tool babysitter_list_runs", params);
                try {
                    let runs = await discoverRuns(runsDir, verbose);
                    if (params.statusFilter) {
                        runs = runs.filter((r) => r.status === params.statusFilter);
                    }
                    if (runs.length === 0) {
                        return formatToolResult({ runs: [], runsDir }, "No runs found.");
                    }
                    const summary = runs.map((r) => ({
                        runId: r.runId,
                        processId: r.processId,
                        status: r.status,
                        createdAt: r.createdAt,
                        prompt: r.prompt
                            ? r.prompt.length > 80 ? r.prompt.slice(0, 77) + "..." : r.prompt
                            : undefined,
                        effects: `${r.resolvedEffects}/${r.totalEffects} resolved`,
                        pendingEffects: r.pendingEffects,
                    }));
                    return formatToolResult({ count: runs.length, runs: summary }, `Found ${runs.length} run(s).`);
                }
                catch (err) {
                    const msg = err instanceof Error ? err.message : String(err);
                    return errorResult(`Failed to list runs: ${msg}`);
                }
            },
        },
        {
            name: "babysitter_assess_run",
            label: "Assess Run",
            description: "Get detailed state assessment for a specific run by ID. " +
                "Returns status, effects, journal summary, and entrypoint info.",
            parameters: typebox_1.Type.Object({
                runId: typebox_1.Type.String({ description: "The run ID (or prefix) to assess" }),
            }),
            execute: async (_toolCallId, params) => {
                writeVerboseData("resume tool babysitter_assess_run", params);
                const runDir = path.join(runsDir, params.runId);
                try {
                    const assessment = await assessRun(runDir);
                    return formatToolResult({
                        runId: assessment.run.runId,
                        processId: assessment.run.processId,
                        status: assessment.run.status,
                        createdAt: assessment.run.createdAt,
                        prompt: assessment.run.prompt,
                        totalEffects: assessment.run.totalEffects,
                        resolvedEffects: assessment.run.resolvedEffects,
                        pendingEffects: assessment.run.pendingEffects,
                        journalEvents: assessment.journalLength,
                        lastEvent: assessment.lastEvent,
                        entrypoint: assessment.run.entrypoint,
                        runDir: assessment.run.runDir,
                        resumable: assessment.run.status !== "completed",
                    }, `Run ${assessment.run.runId} assessed.`);
                }
                catch (err) {
                    const msg = err instanceof Error ? err.message : String(err);
                    return errorResult(`Failed to assess run "${params.runId}": ${msg}`);
                }
            },
        },
        {
            name: "babysitter_resume_run",
            label: "Resume Run",
            description: "Resume orchestration for a specific run. Call this exactly once " +
                "after assessing the run and confirming it should be resumed.",
            parameters: typebox_1.Type.Object({
                runId: typebox_1.Type.String({ description: "The run ID to resume" }),
            }),
            execute: async (_toolCallId, params) => {
                writeVerboseData("resume tool babysitter_resume_run", params);
                if (resumeTriggered) {
                    return errorResult("Resume has already been triggered for this session.");
                }
                const runDir = path.join(runsDir, params.runId);
                try {
                    const assessment = await assessRun(runDir);
                    const selectedRun = assessment.run;
                    if (selectedRun.status === "completed") {
                        return formatToolResult({ runId: selectedRun.runId, status: "completed" }, "Run is already completed. Nothing to resume.");
                    }
                    resumeTriggered = true;
                    if (!json) {
                        process.stderr.write(`${MAGENTA}Resuming run ${BOLD}${selectedRun.runId}${RESET}${MAGENTA}...${RESET}\n\n`);
                    }
                    // Resolve the process entry path
                    const entryImportPath = selectedRun.entrypoint.importPath;
                    const processPath = path.isAbsolute(entryImportPath)
                        ? entryImportPath
                        : path.resolve(entryImportPath);
                    resumeExitCode = await (0, harnessCreateRun_1.handleHarnessCreateRun)({
                        processPath,
                        prompt: selectedRun.prompt,
                        harness: args.harness,
                        workspace: args.workspace,
                        model: args.model,
                        maxIterations: args.maxIterations,
                        runsDir: args.runsDir,
                        json: args.json,
                        verbose: args.verbose,
                        interactive: args.interactive,
                        existingRunId: selectedRun.runId,
                        existingRunDir: selectedRun.runDir,
                    });
                    return formatToolResult({ runId: selectedRun.runId, exitCode: resumeExitCode }, resumeExitCode === 0
                        ? "Run resumed successfully."
                        : `Run resume completed with exit code ${resumeExitCode}.`);
                }
                catch (err) {
                    const msg = err instanceof Error ? err.message : String(err);
                    return errorResult(`Failed to resume run "${params.runId}": ${msg}`);
                }
            },
        },
    ];
    // -----------------------------------------------------------------
    // Build agentic tools (read/write/edit/grep/bash/etc.)
    // -----------------------------------------------------------------
    const agenticTools = (0, agenticTools_1.createAgenticToolDefinitions)({
        workspace: args.workspace ?? process.cwd(),
        interactive: interactive,
        askUserQuestionHandler: interactive
            ? undefined // Let the default AskUserQuestion handler work
            : undefined,
    });
    const mergedCustomTools = [...customTools, ...agenticTools];
    writeVerbose(`[resume setup] runsDir=${runsDir} workspace=${path.resolve(args.workspace ?? process.cwd())} model=${args.model ?? "(default)"}`);
    writeVerboseData("resume tools", mergedCustomTools.map((tool) => ({
        name: tool.name,
        label: tool.label,
    })));
    // -----------------------------------------------------------------
    // Create Pi session
    // -----------------------------------------------------------------
    const systemPrompt = buildResumeSystemPrompt(runsDir, args.runId);
    const userPrompt = buildResumeUserPrompt(args.runId);
    writeVerboseData("resume system prompt", systemPrompt);
    writeVerboseData("resume user prompt", userPrompt);
    let session = null;
    try {
        session = (0, piWrapper_1.createPiSession)({
            workspace: args.workspace,
            model: args.model,
            thinkingLevel: "low",
            toolsMode: "coding",
            customTools: mergedCustomTools,
            systemPrompt,
            isolated: true,
            ephemeral: true,
        });
        await session.initialize();
        // Subscribe to stream text output to stderr in non-JSON mode
        let unsubscribe = null;
        if (!json) {
            process.stderr.write(`\n${BOLD}${MAGENTA}Run Discovery${RESET} ${DIM}Agent is searching for runs...${RESET}\n\n`);
            unsubscribe = session.subscribe((event) => {
                if (event.type === "text_delta") {
                    const text = event.text;
                    if (text)
                        process.stderr.write(text);
                }
            });
        }
        // Prompt the agent
        const result = await session.prompt(userPrompt, 300_000);
        if (unsubscribe)
            unsubscribe();
        if (!json)
            process.stderr.write("\n");
        writeVerboseData("resume agent result", {
            success: result.success,
            outputPreview: result.output.length > 500
                ? result.output.slice(0, 497) + "..."
                : result.output,
            resumeTriggered,
        });
        if (!result.success && !resumeTriggered) {
            if (json) {
                process.stdout.write(JSON.stringify({ ok: false, error: "Resume agent failed", details: result.output }) + "\n");
            }
            else {
                process.stderr.write(`${RED}Resume agent failed:${RESET} ${result.output}\n`);
            }
            return 1;
        }
        // If the agent never triggered resume, it decided not to (e.g., no runs or user quit)
        if (!resumeTriggered) {
            if (json) {
                process.stdout.write(JSON.stringify({ ok: true, resumed: false, message: "No run was resumed" }) + "\n");
            }
            return 0;
        }
        return resumeExitCode;
    }
    catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        if (json) {
            process.stdout.write(JSON.stringify({ ok: false, error: "Resume session failed", details: message }) + "\n");
        }
        else {
            process.stderr.write(`${RED}Error:${RESET} ${message}\n`);
        }
        return 1;
    }
    finally {
        if (session) {
            try {
                session.dispose();
            }
            catch {
                // Ignore cleanup errors
            }
        }
    }
}
