// @a5c-ai/babysitter-sdk/dist/cli/commands/harnessUtils.js
"use strict";
/**
 * Shared utilities, types, constants, and logging helpers for the
 * harness:create-run command handler phases.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Type = exports.ErrorCategory = exports.BabysitterRuntimeError = exports.discoverHarnesses = exports.PiSessionHandle = exports.createPiSession = exports.createAskUserQuestionResponse = exports.createApprovalAskUserQuestion = exports.createReadlineAskUserQuestionUiContext = exports.ASK_USER_QUESTION_SCHEMA = exports.ASK_QUESTION_SCHEMA = exports.ASK_OPTION_SCHEMA = exports.PI_WORKER_TIMEOUT_MS = exports.PI_DEFAULT_PROMPT_TIMEOUT_MS = exports.PI_PARENT_PROMPT_TIMEOUT_MS = exports.DEFAULT_INTERACTIVE_ASK_TIMEOUT_MS = exports.TRANSIENT_PI_PROMPT_RETRY_DELAYS_MS = exports.DEFAULT_EFFECT_RETRY_CONFIG = exports.PROCESS_LIBRARY_SEARCH_DEFAULT_LIMIT = exports.PROCESS_LIBRARY_READ_MAX_CHARS = exports.VERBOSE_LOG_LIMIT = exports.MAGENTA = exports.RED = exports.YELLOW = exports.GREEN = exports.CYAN = exports.BOLD = exports.RESET = exports.DIM = void 0;
exports.truncateForVerboseLog = truncateForVerboseLog;
exports.stringifyForVerboseLog = stringifyForVerboseLog;
exports.writeVerboseLine = writeVerboseLine;
exports.writeVerboseBlock = writeVerboseBlock;
exports.emitProgress = emitProgress;
exports.loadSessionCompressionConfig = loadSessionCompressionConfig;
exports.compressInternalHarnessPrompt = compressInternalHarnessPrompt;
exports.buildPromptContext = buildPromptContext;
exports.selectHarness = selectHarness;
exports.createReadlineInterface = createReadlineInterface;
exports.askLine = askLine;
exports.readInteractivePrompt = readInteractivePrompt;
exports.formatToolResult = formatToolResult;
exports.isApprovalAskRequest = isApprovalAskRequest;
exports.askUserQuestionViaTool = askUserQuestionViaTool;
exports.resolveTaskHarness = resolveTaskHarness;
exports.isPiHarness = isPiHarness;
exports.isInternalHarness = isInternalHarness;
exports.usesExternalHarness = usesExternalHarness;
exports.shouldUseExternalHarness = shouldUseExternalHarness;
exports.shellQuoteArg = shellQuoteArg;
exports.readBooleanMetadata = readBooleanMetadata;
exports.readBashSandboxMetadata = readBashSandboxMetadata;
exports.readStringMetadata = readStringMetadata;
exports.isProcessModuleLoadFailure = isProcessModuleLoadFailure;
exports.isRetryablePiPromptFailure = isRetryablePiPromptFailure;
exports.isIgnorablePiPromptFailure = isIgnorablePiPromptFailure;
exports.promptPiWithRetry = promptPiWithRetry;
exports.buildPiWorkerSessionOptions = buildPiWorkerSessionOptions;
const path = __importStar(require("node:path"));
const readline = __importStar(require("node:readline"));
const node_fs_1 = require("node:fs");
const typebox_1 = require("@sinclair/typebox");
const discovery_1 = require("../../harness/discovery");
const config_loader_1 = require("../../compression/config-loader");
const density_filter_1 = require("../../compression/density-filter");
const interaction_1 = require("../../interaction");
// ── Constants ────────────────────────────────────────────────────────
exports.DIM = "\x1b[2m";
exports.RESET = "\x1b[0m";
exports.BOLD = "\x1b[1m";
exports.CYAN = "\x1b[36m";
exports.GREEN = "\x1b[32m";
exports.YELLOW = "\x1b[33m";
exports.RED = "\x1b[31m";
exports.MAGENTA = "\x1b[35m";
exports.VERBOSE_LOG_LIMIT = 4_000;
exports.PROCESS_LIBRARY_READ_MAX_CHARS = 24_000;
exports.PROCESS_LIBRARY_SEARCH_DEFAULT_LIMIT = 12;
exports.DEFAULT_EFFECT_RETRY_CONFIG = {
    maxRetries: 3,
    baseDelayMs: 1_000,
    backoffMultiplier: 4,
    maxDelayMs: 60_000,
    nonRetryableKinds: ["breakpoint", "sleep"],
};
const HARNESS_PRIORITY = [
    "internal",
    "oh-my-pi",
    "pi",
    "claude-code",
    "codex",
    "gemini-cli",
    "opencode",
];
exports.TRANSIENT_PI_PROMPT_RETRY_DELAYS_MS = process.env.VITEST
    ? [0, 0]
    : [1_000, 3_000];
exports.DEFAULT_INTERACTIVE_ASK_TIMEOUT_MS = 600_000;
// Parent orchestration turns may legitimately wait on user input or external
// work for an unbounded duration. A timeout of 0 disables the wrapper timer.
exports.PI_PARENT_PROMPT_TIMEOUT_MS = 0;
exports.PI_DEFAULT_PROMPT_TIMEOUT_MS = 900_000;
exports.PI_WORKER_TIMEOUT_MS = 1_800_000;
exports.ASK_OPTION_SCHEMA = typebox_1.Type.Object({
    label: typebox_1.Type.String(),
    description: typebox_1.Type.Optional(typebox_1.Type.String()),
    preview: typebox_1.Type.Optional(typebox_1.Type.String()),
});
exports.ASK_QUESTION_SCHEMA = typebox_1.Type.Object({
    question: typebox_1.Type.String(),
    header: typebox_1.Type.Optional(typebox_1.Type.String()),
    options: typebox_1.Type.Optional(typebox_1.Type.Array(exports.ASK_OPTION_SCHEMA)),
    multiSelect: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    allowOther: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    required: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    recommended: typebox_1.Type.Optional(typebox_1.Type.Number({ description: "Index of the recommended option (0-based). Auto-selected in non-interactive mode or on timeout." })),
});
exports.ASK_USER_QUESTION_SCHEMA = typebox_1.Type.Object({
    questions: typebox_1.Type.Array(exports.ASK_QUESTION_SCHEMA, { minItems: 1, maxItems: 4 }),
    timeout: typebox_1.Type.Optional(typebox_1.Type.Number({ description: "Timeout in ms. On expiry, the recommended option (or first) is auto-selected." })),
});
// ── Logging Utilities ────────────────────────────────────────────────
function truncateForVerboseLog(text, maxChars = exports.VERBOSE_LOG_LIMIT) {
    const normalized = text.replace(/\r\n/g, "\n");
    if (normalized.length <= maxChars) {
        return normalized;
    }
    return `${normalized.slice(0, maxChars)}\n... [truncated ${normalized.length - maxChars} chars]`;
}
function stringifyForVerboseLog(value, maxChars = exports.VERBOSE_LOG_LIMIT) {
    try {
        const raw = typeof value === "string" ? value : JSON.stringify(value, null, 2);
        return truncateForVerboseLog(raw, maxChars);
    }
    catch {
        return truncateForVerboseLog(String(value), maxChars);
    }
}
function writeVerboseLine(enabled, json, message) {
    if (!json && enabled) {
        process.stderr.write(`${exports.DIM}${message}${exports.RESET}\n`);
    }
}
function writeVerboseBlock(enabled, json, label, value, maxChars = exports.VERBOSE_LOG_LIMIT) {
    if (json || !enabled) {
        return;
    }
    process.stderr.write(`${exports.DIM}[${label}]${exports.RESET}\n${exports.DIM}${stringifyForVerboseLog(value, maxChars)}${exports.RESET}\n`);
}
function emitProgress(payload, json, verbose) {
    if (json) {
        console.log(JSON.stringify(payload, null, 2));
        return;
    }
    switch (payload.phase) {
        case "1":
            if (payload.status === "skipped") {
                process.stderr.write(`${exports.DIM}Phase 1 skipped (--process provided)${exports.RESET}\n`);
            }
            else if (payload.status === "started") {
                process.stderr.write(`\n${exports.BOLD}Phase 1${exports.RESET} ${exports.DIM}Interview / process definition via ${payload.harness}...${exports.RESET}\n`);
            }
            else if (payload.status === "completed") {
                process.stderr.write(`${exports.GREEN}Phase 1 complete${exports.RESET} ${exports.DIM}${payload.processPath}${exports.RESET}\n`);
            }
            else if (payload.status === "failed") {
                process.stderr.write(`${exports.RED}Phase 1 failed:${exports.RESET} ${payload.error}\n`);
            }
            else if (payload.status === "interview") {
                process.stderr.write(`${exports.DIM}Interview answers: ${payload.answer}${exports.RESET}\n`);
            }
            break;
        case "2":
            if (payload.status === "started") {
                process.stderr.write(`\n${exports.BOLD}Phase 2${exports.RESET} ${exports.DIM}Bound orchestration loop${exports.RESET}\n`);
            }
            else if (payload.status === "run-created") {
                process.stderr.write(`${exports.GREEN}Run created${exports.RESET} runId=${exports.CYAN}${payload.runId}${exports.RESET}\n`);
                if (verbose)
                    process.stderr.write(`  ${exports.DIM}runDir: ${payload.runDir}${exports.RESET}\n`);
            }
            else if (payload.status === "bound") {
                if (payload.error) {
                    process.stderr.write(`${exports.YELLOW}Session binding warning:${exports.RESET} ${payload.error}\n`);
                }
                else {
                    process.stderr.write(`${exports.DIM}session bound via ${payload.harness}: ${payload.sessionId}${exports.RESET}\n`);
                }
            }
            else if (payload.status === "iteration") {
                process.stderr.write(`\n${exports.DIM}-- iteration ${payload.iteration} --${exports.RESET} status=${payload.runStatus} pending=${payload.pendingEffects}\n`);
            }
            else if (payload.status === "effect") {
                const icon = payload.effectStatus === "ok" ? `${exports.GREEN}✓${exports.RESET}` : `${exports.RED}✗${exports.RESET}`;
                process.stderr.write(`  ${icon} ${exports.MAGENTA}${payload.effectKind}${exports.RESET} ${payload.effectTitle ?? payload.effectId}${payload.effectStatus === "error" ? ` ${exports.RED}${payload.error}${exports.RESET}` : ""}\n`);
            }
            else if (payload.status === "completed") {
                process.stderr.write(`\n${exports.GREEN}${exports.BOLD}Run completed${exports.RESET} ${exports.DIM}(${payload.iteration} iterations)${exports.RESET}\n`);
            }
            else if (payload.status === "failed") {
                process.stderr.write(`\n${exports.RED}${exports.BOLD}Run failed:${exports.RESET} ${payload.error}\n`);
            }
            break;
    }
}
// ── Compression ──────────────────────────────────────────────────────
function loadSessionCompressionConfig(workspace) {
    try {
        return (0, config_loader_1.loadCompressionConfig)(workspace ?? process.cwd());
    }
    catch {
        return null;
    }
}
function compressInternalHarnessPrompt(text, compressionConfig, taskKind = "agent") {
    if (!compressionConfig?.enabled || !compressionConfig.layers.sdkContextHook.enabled) {
        return text;
    }
    const layer = compressionConfig.layers.sdkContextHook;
    if ((0, density_filter_1.estimateTokens)(text) <= layer.minCompressionTokens) {
        return text;
    }
    const targetReduction = layer.perTaskKind?.[taskKind] ?? layer.targetReduction;
    return (0, density_filter_1.densityFilterText)(text, targetReduction);
}
// ── Harness Selection ────────────────────────────────────────────────
function buildPromptContext(args) {
    const envNames = [
        "CI",
        "BABYSITTER_COMPRESSION_ENABLED",
        "BABYSITTER_PI_SANDBOX_IMAGE",
        "AZURE_OPENAI_API_KEY",
        "AZURE_OPENAI_PROJECT_NAME",
        "AZURE_OPENAI_DEPLOYMENT",
        "OPENAI_API_KEY",
        "ANTHROPIC_API_KEY",
    ];
    return {
        platform: process.platform,
        arch: process.arch,
        nodeVersion: process.version,
        cwd: process.cwd(),
        workspace: path.resolve(args.workspace ?? process.cwd()),
        selectedHarnessName: args.selectedHarnessName,
        discoveredHarnesses: args.discovered,
        compressionEnabled: Boolean(args.compressionConfig?.enabled &&
            args.compressionConfig.layers.sdkContextHook.enabled),
        secureSandboxImage: process.env.BABYSITTER_PI_SANDBOX_IMAGE || "node:22-bookworm",
        piDefaultBashSandbox: "local",
        piIsolationDefault: false,
        envFlags: envNames.map((name) => ({
            name,
            value: process.env[name]
                ? (name.endsWith("_API_KEY") ? "set" : process.env[name])
                : "unset",
        })),
    };
}
function selectHarness(discovered, preferred) {
    if (preferred) {
        const match = discovered.find((h) => h.name === preferred && h.installed);
        if (match)
            return match;
    }
    const caller = (0, discovery_1.detectCallerHarness)();
    if (caller) {
        const callerMatch = discovered.find((h) => h.name === caller.name && h.installed);
        if (callerMatch)
            return callerMatch;
    }
    for (const name of HARNESS_PRIORITY) {
        const match = discovered.find((h) => h.name === name && h.installed);
        if (match)
            return match;
    }
    return undefined;
}
// ── Readline Utilities ───────────────────────────────────────────────
function createReadlineInterface() {
    return readline.createInterface({
        input: process.stdin,
        output: process.stderr,
        terminal: true,
    });
}
function askLine(rl, promptText) {
    return new Promise((resolve) => {
        rl.question(`${promptText} `, (answer) => {
            resolve(answer);
        });
        rl.once("close", () => resolve(null));
    });
}
async function readInteractivePrompt(rl) {
    process.stderr.write("\n");
    process.stderr.write(`${exports.BOLD}${exports.CYAN}babysitter harness:create-run${exports.RESET}\n`);
    process.stderr.write(`${exports.DIM}Enter your request below. Press Enter to submit.${exports.RESET}\n`);
    process.stderr.write(`${exports.DIM}(Use \\ at end of line for multi-line input, Ctrl+C to cancel)${exports.RESET}\n\n`);
    const lines = [];
    for (;;) {
        const line = await askLine(rl, lines.length === 0 ? ">" : "...");
        if (line === null)
            return null; // EOF / Ctrl+C / Ctrl+D
        if (line.endsWith("\\")) {
            lines.push(line.slice(0, -1));
            continue;
        }
        lines.push(line);
        const combined = lines.join("\n").trim();
        if (combined)
            return combined;
        lines.length = 0; // Reset if empty, re-prompt
    }
}
// ── Tool Result Formatting ───────────────────────────────────────────
function formatToolResult(data, message) {
    if (typeof data === "string") {
        return {
            content: [{
                    type: "text",
                    text: message ? `${message}\n${data}` : data,
                }],
            details: data,
        };
    }
    const content = message
        ? `${message}\n${JSON.stringify(data, null, 2)}`
        : JSON.stringify(data, null, 2);
    return {
        content: [{
                type: "text",
                text: content,
            }],
        details: data,
    };
}
// ── AskUserQuestion Utilities ────────────────────────────────────────
function isApprovalAskRequest(request) {
    const question = request.questions[0];
    if (!question || request.questions.length !== 1 || !question.options) {
        return false;
    }
    const labels = question.options.map((option) => option.label);
    return labels.length === 2 &&
        labels[0] === "Approve" &&
        labels[1] === "Reject" &&
        question.allowOther === false;
}
async function askUserQuestionViaTool(request, interactive, rl, toolContext) {
    // Apply a default timeout only when interactive AND no explicit timeout was
    // provided.  An explicit timeout of 0 means "wait indefinitely" and must be
    // respected — do NOT override it with the default.
    const effectiveRequest = interactive && request.timeout == null
        ? {
            ...request,
            timeout: exports.DEFAULT_INTERACTIVE_ASK_TIMEOUT_MS,
        }
        : request;
    (0, interaction_1.validateAskUserQuestionRequest)(effectiveRequest);
    if (interactive) {
        try {
            if (toolContext?.hasUI && toolContext.ui) {
                return await (0, interaction_1.promptAskUserQuestionWithUiContext)(toolContext.ui, effectiveRequest);
            }
            if (rl) {
                return await (0, interaction_1.promptAskUserQuestionWithReadline)(rl, effectiveRequest);
            }
        }
        catch {
            return (0, interaction_1.createDefaultAskUserQuestionResponse)(effectiveRequest);
        }
    }
    const answers = {};
    for (const [index, question] of effectiveRequest.questions.entries()) {
        const key = question.header?.trim() || `Question ${index + 1}`;
        if (question.recommended != null &&
            question.options &&
            question.options[question.recommended]) {
            answers[key] = question.options[question.recommended].label;
        }
        else {
            answers[key] = "";
        }
    }
    if (isApprovalAskRequest(effectiveRequest)) {
        const key = effectiveRequest.questions[0]?.header?.trim() || "Decision";
        const rec = effectiveRequest.questions[0]?.recommended;
        const opts = effectiveRequest.questions[0]?.options;
        if (rec != null && opts && opts[rec]) {
            answers[key] = opts[rec].label;
        }
        else {
            answers[key] = "Approve";
        }
    }
    return (0, interaction_1.createAskUserQuestionResponse)(effectiveRequest, answers);
}
// ── Harness Resolution Utilities ─────────────────────────────────────
function resolveTaskHarness(action, defaultHarness, discovered) {
    const meta = action.taskDef?.metadata;
    if (typeof meta?.harness === "string") {
        const match = discovered.find((h) => h.name === meta.harness && h.installed);
        if (match)
            return match.name;
    }
    return defaultHarness;
}
/**
 * Returns true if the harness uses the internal programmatic engine (piWrapper).
 * 'internal' is the primary programmatic harness; 'oh-my-pi' also uses piWrapper.
 * 'pi' is CLI-only and does NOT use the programmatic engine.
 *
 * @deprecated Prefer isInternalHarness — isPiHarness is kept for backward compat.
 */
function isPiHarness(harnessName) {
    return isInternalHarness(harnessName);
}
function isInternalHarness(harnessName) {
    return harnessName === "internal" || harnessName === "oh-my-pi";
}
function usesExternalHarness(harnessName) {
    return !isPiHarness(harnessName);
}
function shouldUseExternalHarness(harnessName) {
    if (!usesExternalHarness(harnessName)) {
        return false;
    }
    return (0, discovery_1.detectCallerHarness)()?.name === harnessName;
}
function shellQuoteArg(value) {
    return `'${value.replace(/'/g, `'"'"'`)}'`;
}
function readBooleanMetadata(metadata, key) {
    const value = metadata?.[key];
    if (typeof value === "boolean") {
        return value;
    }
    if (value === "true") {
        return true;
    }
    if (value === "false") {
        return false;
    }
    return undefined;
}
function readBashSandboxMetadata(metadata) {
    const value = metadata?.bashSandbox;
    return value === "auto" || value === "secure" || value === "local"
        ? value
        : undefined;
}
function readStringMetadata(metadata, key) {
    const value = metadata?.[key];
    return typeof value === "string" ? value : undefined;
}
// ── PI Prompt Retry ──────────────────────────────────────────────────
function isProcessModuleLoadFailure(error) {
    return error instanceof Error && /^Failed to load process module at /.test(error.message);
}
function isRetryablePiPromptFailure(error) {
    const message = typeof error === "string"
        ? error
        : error instanceof Error
            ? error.message
            : typeof error === "object" && error !== null && "output" in error
                ? String(error.output ?? "")
                : "";
    if (!message) {
        return false;
    }
    const normalized = message.toLowerCase();
    return normalized.includes("the server had an error processing your request") ||
        normalized.includes("please retry your request") ||
        normalized.includes("temporarily unavailable") ||
        normalized.includes("rate limit") ||
        normalized.includes("too many requests") ||
        normalized.includes("pitimeouterror") ||
        normalized.includes("pi prompt timed out") ||
        normalized.includes("deadline exceeded");
}
function isIgnorablePiPromptFailure(output) {
    return output.includes("msg.content.filter is not a function");
}
async function promptPiWithRetry(args) {
    let attempt = 0;
    for (;;) {
        try {
            const result = await args.session.prompt(args.message, args.timeout);
            if (result.success ||
                !isRetryablePiPromptFailure(result.output) ||
                attempt >= exports.TRANSIENT_PI_PROMPT_RETRY_DELAYS_MS.length) {
                return result;
            }
            const delayMs = exports.TRANSIENT_PI_PROMPT_RETRY_DELAYS_MS[attempt] ?? 0;
            attempt += 1;
            args.writeVerbose?.(`[${args.label} retry] transient PI failure; retrying prompt attempt ${attempt}/${exports.TRANSIENT_PI_PROMPT_RETRY_DELAYS_MS.length} after ${delayMs}ms`);
            args.writeVerboseData?.(`${args.label} retry failure output`, result.output);
            if (delayMs > 0) {
                await new Promise((resolve) => setTimeout(resolve, delayMs));
            }
        }
        catch (error) {
            if (!isRetryablePiPromptFailure(error) || attempt >= exports.TRANSIENT_PI_PROMPT_RETRY_DELAYS_MS.length) {
                throw error;
            }
            const delayMs = exports.TRANSIENT_PI_PROMPT_RETRY_DELAYS_MS[attempt] ?? 0;
            attempt += 1;
            args.writeVerbose?.(`[${args.label} retry] transient PI exception; retrying prompt attempt ${attempt}/${exports.TRANSIENT_PI_PROMPT_RETRY_DELAYS_MS.length} after ${delayMs}ms`);
            args.writeVerboseData?.(`${args.label} retry error`, error);
            if (delayMs > 0) {
                await new Promise((resolve) => setTimeout(resolve, delayMs));
            }
        }
    }
}
// ── PI Worker Session Options ────────────────────────────────────────
function buildPiWorkerSessionOptions(args) {
    const metadata = args.action.taskDef?.metadata;
    const isolated = readBooleanMetadata(metadata, "isolated");
    const enableCompaction = readBooleanMetadata(metadata, "enableCompaction");
    // Priority: delegationConfig > task metadata > defaults
    const effectiveModel = args.delegationConfig?.model ?? args.model;
    const effectiveTimeout = args.delegationConfig?.timeout ?? exports.PI_WORKER_TIMEOUT_MS;
    const effectiveToolsMode = args.delegationConfig?.toolsMode
        ?? readStringMetadata(metadata, "toolsMode")
        ?? "coding";
    const rawThinkingLevel = args.delegationConfig?.thinkingLevel
        ?? readStringMetadata(metadata, "thinkingLevel");
    // Map delegation "none" to undefined (no thinking header sent to PI).
    const effectiveThinkingLevel = rawThinkingLevel === "none" ? undefined : rawThinkingLevel;
    const effectiveBashSandbox = args.delegationConfig?.bashSandbox
        ?? readBashSandboxMetadata(metadata);
    // Resolve skills to appendSystemPrompt content
    let appendSystemPrompt;
    const skillNames = args.delegationConfig?.skills ?? metadata?.skills;
    if (skillNames?.length) {
        const skillContents = [];
        for (const skillName of skillNames) {
            try {
                const candidates = [
                    path.join(args.workspace ?? process.cwd(), ".a5c", "skills", skillName, "SKILL.md"),
                    path.join(args.workspace ?? process.cwd(), ".claude", "plugins", skillName, "SKILL.md"),
                ];
                for (const candidate of candidates) {
                    if ((0, node_fs_1.existsSync)(candidate)) {
                        skillContents.push((0, node_fs_1.readFileSync)(candidate, "utf8"));
                        break;
                    }
                }
            }
            catch { /* skip missing skills */ }
        }
        if (skillContents.length) {
            appendSystemPrompt = [skillContents.join("\n\n---\n\n")];
        }
    }
    // Resolve subagentName to agentDir
    const subagentName = args.delegationConfig?.subagentName ?? readStringMetadata(metadata, "subagentName");
    let agentDir;
    if (subagentName) {
        const candidates = [
            path.join(args.workspace ?? process.cwd(), ".claude", "agents", subagentName),
        ];
        for (const candidate of candidates) {
            if ((0, node_fs_1.existsSync)(candidate)) {
                agentDir = candidate;
                break;
            }
        }
    }
    return {
        workspace: args.workspace,
        model: effectiveModel,
        timeout: effectiveTimeout,
        toolsMode: effectiveToolsMode,
        ephemeral: true,
        ...(args.customTools?.length ? { customTools: args.customTools } : {}),
        ...(isolated !== undefined ? { isolated } : {}),
        ...(enableCompaction !== undefined ? { enableCompaction } : {}),
        ...(effectiveBashSandbox ? { bashSandbox: effectiveBashSandbox } : {}),
        ...(effectiveThinkingLevel ? { thinkingLevel: effectiveThinkingLevel } : {}),
        ...(appendSystemPrompt ? { appendSystemPrompt } : {}),
        ...(agentDir ? { agentDir } : {}),
    };
}
var interaction_2 = require("../../interaction");
Object.defineProperty(exports, "createReadlineAskUserQuestionUiContext", { enumerable: true, get: function () { return interaction_2.createReadlineAskUserQuestionUiContext; } });
Object.defineProperty(exports, "createApprovalAskUserQuestion", { enumerable: true, get: function () { return interaction_2.createApprovalAskUserQuestion; } });
Object.defineProperty(exports, "createAskUserQuestionResponse", { enumerable: true, get: function () { return interaction_2.createAskUserQuestionResponse; } });
var piWrapper_1 = require("../../harness/piWrapper");
Object.defineProperty(exports, "createPiSession", { enumerable: true, get: function () { return piWrapper_1.createPiSession; } });
Object.defineProperty(exports, "PiSessionHandle", { enumerable: true, get: function () { return piWrapper_1.PiSessionHandle; } });
var discovery_2 = require("../../harness/discovery");
Object.defineProperty(exports, "discoverHarnesses", { enumerable: true, get: function () { return discovery_2.discoverHarnesses; } });
var exceptions_1 = require("../../runtime/exceptions");
Object.defineProperty(exports, "BabysitterRuntimeError", { enumerable: true, get: function () { return exceptions_1.BabysitterRuntimeError; } });
Object.defineProperty(exports, "ErrorCategory", { enumerable: true, get: function () { return exceptions_1.ErrorCategory; } });
var typebox_2 = require("@sinclair/typebox");
Object.defineProperty(exports, "Type", { enumerable: true, get: function () { return typebox_2.Type; } });
