// @a5c-ai/babysitter-sdk/dist/cli/commands/harnessCreateRun.js
"use strict";
/**
 * harness:create-run command handler.
 *
 * Drives a full babysitter session lifecycle through two agentic phases:
 *   Phase 1 - Unbound interview / intent / process-definition
 *   Phase 2 - Bound orchestration loop with hook-style continuation semantics
 *
 * Both phases are driven through a Pi agent session and LLM-callable tools.
 * Interactive user input is exposed through an AskUserQuestion tool instead of
 * direct imperative prompts inside the host loop.
 *
 * This module is a thin coordinator that imports the phase implementations
 * from separate modules and exports the public API.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleSessionCreate = exports.runOrchestrationPhase = exports.selectHarness = void 0;
exports.handleHarnessCreateRun = handleHarnessCreateRun;
const harnessUtils_1 = require("./harnessUtils");
const harnessPhase1_1 = require("./harnessPhase1");
const harnessPhase2_1 = require("./harnessPhase2");
var harnessUtils_2 = require("./harnessUtils");
Object.defineProperty(exports, "selectHarness", { enumerable: true, get: function () { return harnessUtils_2.selectHarness; } });
var harnessPhase2_2 = require("./harnessPhase2");
Object.defineProperty(exports, "runOrchestrationPhase", { enumerable: true, get: function () { return harnessPhase2_2.runOrchestrationPhase; } });
// ── Main Entry Point ─────────────────────────────────────────────────
async function handleHarnessCreateRun(parsed) {
    const { prompt: initialPrompt, harness: preferredHarness, processPath: providedProcessPath, workspace, model, maxIterations = 256, runsDir, json, verbose, } = parsed;
    const interactive = parsed.interactive ?? (process.stdin.isTTY === true && !json);
    const rl = interactive ? (0, harnessUtils_1.createReadlineInterface)() : null;
    try {
        let prompt = initialPrompt;
        if (!prompt && !providedProcessPath) {
            if (interactive && rl) {
                const userPrompt = await (0, harnessUtils_1.readInteractivePrompt)(rl);
                if (!userPrompt) {
                    return 0; // User cancelled
                }
                prompt = userPrompt;
            }
            else {
                const error = "Either --prompt or --process must be provided";
                if (json) {
                    console.error(JSON.stringify({ error: "MISSING_PROMPT", message: error }, null, 2));
                }
                else {
                    process.stderr.write(`${harnessUtils_1.RED}Error:${harnessUtils_1.RESET} ${error}\n`);
                }
                return 1;
            }
        }
        const discovered = await (0, harnessUtils_1.discoverHarnesses)();
        const selectedHarnessName = preferredHarness ?? "internal";
        const compressionConfig = (0, harnessUtils_1.loadSessionCompressionConfig)(workspace);
        const promptContext = (0, harnessUtils_1.buildPromptContext)({
            workspace,
            selectedHarnessName,
            discovered,
            compressionConfig,
        });
        let processPath = providedProcessPath;
        if (processPath) {
            (0, harnessUtils_1.emitProgress)({ phase: "1", status: "skipped", processPath }, json, verbose);
        }
        else {
            const workDir = workspace ?? process.cwd();
            processPath = await (0, harnessPhase1_1.runProcessDefinitionPhase)({
                prompt: prompt,
                outputDir: (0, harnessPhase1_1.getProcessOutputDir)(workDir),
                workspace: workDir,
                model,
                interactive,
                rl,
                json,
                verbose,
                compressionConfig,
                promptContext,
                selectedHarnessName,
            });
        }
        if (parsed.planOnly) {
            (0, harnessUtils_1.emitProgress)({ phase: "2", status: "skipped-plan-only", processPath }, json, verbose);
            if (json) {
                process.stdout.write(JSON.stringify({ ok: true, planOnly: true, processPath }) + "\n");
            }
            else {
                process.stderr.write(`${harnessUtils_1.GREEN}Process definition created: ${harnessUtils_1.BOLD}${processPath}${harnessUtils_1.RESET}\n`);
                process.stderr.write(`${harnessUtils_1.DIM}Run /babysitter:call or harness:create-run --process ${processPath} to execute.${harnessUtils_1.RESET}\n`);
            }
            return 0;
        }
        return await (0, harnessPhase2_1.runOrchestrationPhase)({
            processPath,
            prompt,
            workspace,
            model,
            runsDir,
            maxIterations,
            json,
            verbose,
            interactive,
            rl,
            selectedHarnessName,
            discovered,
            compressionConfig,
            promptContext,
            existingRunId: parsed.existingRunId,
            existingRunDir: parsed.existingRunDir,
        });
    }
    finally {
        rl?.close();
    }
}
/** @deprecated Use handleHarnessCreateRun instead */
exports.handleSessionCreate = handleHarnessCreateRun;
