// @a5c-ai/babysitter-sdk/dist/cli/commands/hookRun.js
"use strict";
/**
 * hook:run CLI command.
 *
 * Dispatches hook handling to the appropriate harness adapter.
 * Each harness (e.g. "claude-code") implements its own stop and
 * session-start handlers via the HarnessAdapter interface.
 *
 * The "user-prompt-submit" hook type is harness-agnostic: it reads the
 * Claude Code UserPromptSubmit JSON payload from stdin, applies density-filter
 * compression if the prompt exceeds the configured token threshold, and writes
 * the (possibly compressed) payload to stdout.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleHookRun = handleHookRun;
const harness_1 = require("../../harness");
const config_loader_1 = require("../../compression/config-loader");
const density_filter_1 = require("../../compression/density-filter");
const runLogger_1 = require("../../logging/runLogger");
// ---------------------------------------------------------------------------
// Stdin reader
// ---------------------------------------------------------------------------
async function readStdin() {
    const chunks = [];
    return new Promise((resolve, reject) => {
        process.stdin.on("data", (chunk) => chunks.push(chunk));
        process.stdin.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
        process.stdin.on("error", reject);
    });
}
async function handleUserPromptSubmit() {
    let raw;
    try {
        raw = await readStdin();
    }
    catch {
        return 0;
    }
    let payload;
    try {
        payload = JSON.parse(raw);
    }
    catch {
        // Not valid JSON — pass through unchanged
        process.stdout.write(raw);
        return 0;
    }
    // Apply compression only if enabled; if disabled, falls through and outputs original payload
    const config = (0, config_loader_1.loadCompressionConfig)(process.cwd());
    const layer = config.layers.userPromptHook;
    if (config.enabled && layer.enabled) {
        const prompt = payload.prompt;
        if (typeof prompt === "string") {
            const tokenCount = (0, density_filter_1.estimateTokens)(prompt);
            if (tokenCount > layer.threshold) {
                payload.prompt = (0, density_filter_1.densityFilterText)(prompt, 1 - layer.keepRatio);
            }
        }
    }
    process.stdout.write(JSON.stringify(payload));
    return 0;
}
// ---------------------------------------------------------------------------
// pre-tool-use handler (harness-agnostic, self-contained)
//
// Reads the Claude Code PreToolUse JSON payload from stdin, checks if the
// Bash command is a simple compressible one (git, ls, grep, diff, cat, …),
// and rewrites it to `babysitter compress-output <cmd>` so command output is
// compressed before it enters Claude's context window.
//
// Replaces the external rtk-rewrite.sh hook with a self-contained babysitter
// implementation.
// ---------------------------------------------------------------------------
const COMPRESSIBLE_BINS = new Set([
    "git", "ls", "dir", "grep", "rg", "ag", "diff", "delta",
    "cat", "head", "tail", "less", "more",
]);
/** Returns true if the command is a plain invocation with no shell operators. */
function isSimpleCommand(cmd) {
    // Skip rewrites for commands that contain pipes, redirects, subshells, etc.
    return !/[|;&<>`]|\$\(/.test(cmd);
}
async function handlePreToolUse() {
    let raw;
    try {
        raw = await readStdin();
    }
    catch {
        return 0;
    }
    // Check feature flag before doing anything
    const config = (0, config_loader_1.loadCompressionConfig)(process.cwd());
    if (!config.enabled || !config.layers.commandOutputHook.enabled) {
        return 0;
    }
    let payload;
    try {
        payload = JSON.parse(raw);
    }
    catch {
        // Not JSON — pass through
        return 0;
    }
    const toolInput = payload.tool_input;
    const command = typeof toolInput?.command === "string" ? toolInput.command.trim() : "";
    if (!command)
        return 0;
    const firstToken = command.split(/\s+/)[0]?.toLowerCase() ?? "";
    if (!COMPRESSIBLE_BINS.has(firstToken))
        return 0;
    if (!isSimpleCommand(command))
        return 0;
    // Respect excludeCommands from config
    const excluded = config.layers.commandOutputHook.excludeCommands;
    if (excluded.some(exc => firstToken === exc.toLowerCase()))
        return 0;
    // Rewrite: prepend `babysitter compress-output`
    const rewritten = `babysitter compress-output ${command}`;
    const updatedInput = { ...toolInput, command: rewritten };
    const response = {
        hookSpecificOutput: {
            hookEventName: "PreToolUse",
            permissionDecision: "allow",
            permissionDecisionReason: "babysitter output compression",
            updatedInput,
        },
    };
    process.stdout.write(JSON.stringify(response) + "\n");
    return 0;
}
// ---------------------------------------------------------------------------
// Main dispatcher
// ---------------------------------------------------------------------------
/** Fire-and-forget log to the structured run log facility. */
function logHookEvent(hookType, message, context) {
    void (0, runLogger_1.appendRunLog)({
        timestamp: new Date().toISOString(),
        level: "info",
        type: "hook",
        label: `hook:${hookType}`,
        message,
        source: "hook:run",
        context,
    }).catch(() => {
        // Never let logging break hook execution.
    });
}
async function handleHookRun(args) {
    const { hookType, harness, json } = args;
    if (!hookType) {
        const error = {
            error: "MISSING_HOOK_TYPE",
            message: "--hook-type is required for hook:run",
        };
        if (json) {
            process.stderr.write(JSON.stringify(error, null, 2) + "\n");
        }
        else {
            process.stderr.write("Error: --hook-type is required for hook:run\n");
        }
        return 1;
    }
    logHookEvent(hookType, "Hook invoked", { harness });
    // harness-agnostic hook types — handle before adapter lookup
    if (hookType === "user-prompt-submit") {
        const code = await handleUserPromptSubmit();
        logHookEvent(hookType, `Hook completed with code=${code}`);
        return code;
    }
    if (hookType === "pre-tool-use") {
        const code = await handlePreToolUse();
        logHookEvent(hookType, `Hook completed with code=${code}`);
        return code;
    }
    const adapter = (0, harness_1.getAdapterByName)(harness);
    if (!adapter) {
        const supported = (0, harness_1.listSupportedHarnesses)();
        const error = {
            error: "UNSUPPORTED_HARNESS",
            message: `Unsupported harness: "${harness}". Supported: ${supported.join(", ")}`,
        };
        logHookEvent(hookType, `Unsupported harness: ${harness}`, { error: error.error });
        if (json) {
            process.stderr.write(JSON.stringify(error, null, 2) + "\n");
        }
        else {
            process.stderr.write(`Error: ${error.message}\n`);
        }
        return 1;
    }
    if (adapter.supportsHookType && !adapter.supportsHookType(hookType)) {
        const message = adapter.getUnsupportedHookMessage
            ? adapter.getUnsupportedHookMessage(hookType)
            : `Harness "${harness}" does not support hook type "${hookType}".`;
        const error = {
            error: "UNSUPPORTED_HOOK_TYPE",
            message,
        };
        logHookEvent(hookType, message, { error: error.error });
        if (json) {
            process.stderr.write(JSON.stringify(error, null, 2) + "\n");
        }
        else {
            process.stderr.write(`Error: ${message}\n`);
        }
        return 1;
    }
    let exitCode;
    switch (hookType) {
        case "stop":
            exitCode = await adapter.handleStopHook(args);
            break;
        case "session-start":
            exitCode = await adapter.handleSessionStartHook(args);
            break;
        default: {
            const error = {
                error: "UNKNOWN_HOOK_TYPE",
                message: `Unknown hook type: ${hookType}. Supported: stop, session-start, user-prompt-submit`,
            };
            logHookEvent(hookType, error.message, { error: error.error });
            if (json) {
                process.stderr.write(JSON.stringify(error, null, 2) + "\n");
            }
            else {
                process.stderr.write(`Error: Unknown hook type: ${hookType}. Supported: stop, session-start, user-prompt-submit\n`);
            }
            return 1;
        }
    }
    logHookEvent(hookType, `Hook completed with code=${exitCode}`, { harness });
    return exitCode;
}
