// @a5c-ai/babysitter-sdk/dist/cli/commands/processLibrary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleProcessLibraryClone = handleProcessLibraryClone;
exports.handleProcessLibraryUpdate = handleProcessLibraryUpdate;
exports.handleProcessLibraryUse = handleProcessLibraryUse;
exports.handleProcessLibraryActive = handleProcessLibraryActive;
exports.handleProcessLibraryCommand = handleProcessLibraryCommand;
const active_1 = require("../../processLibrary/active");
async function handleProcessLibraryClone(args) {
    const defaults = (0, active_1.getDefaultProcessLibrarySpec)({
        stateDir: args.stateDir,
        repo: args.repo,
        cloneDir: args.dir,
        ref: args.ref,
    });
    const repo = args.repo ?? defaults.repo;
    const dir = args.dir ?? defaults.cloneDir;
    try {
        const result = await (0, active_1.cloneProcessLibrary)({ repo, dir, ref: args.ref });
        if (args.json) {
            console.log(JSON.stringify({ success: true, ...result }, null, 2));
        }
        else {
            console.log(`Process library cloned.\n  Repo: ${result.repo}\n  Dir: ${result.dir}\n  Revision: ${result.revision}`);
        }
        return 0;
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        if (args.json) {
            console.log(JSON.stringify({ error: "clone_failed", message }));
        }
        else {
            console.error(`[process-library:clone] ${message}`);
        }
        return 1;
    }
}
async function handleProcessLibraryUpdate(args) {
    const defaults = (0, active_1.getDefaultProcessLibrarySpec)({
        stateDir: args.stateDir,
        cloneDir: args.dir,
        ref: args.ref,
    });
    const dir = args.dir ?? defaults.cloneDir;
    try {
        const result = await (0, active_1.updateProcessLibrary)({ dir, ref: args.ref });
        if (args.json) {
            console.log(JSON.stringify({ success: true, ...result }, null, 2));
        }
        else {
            console.log(`Process library updated.\n  Dir: ${result.dir}\n  Revision: ${result.revision}`);
        }
        return 0;
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        if (args.json) {
            console.log(JSON.stringify({ error: "update_failed", message }));
        }
        else {
            console.error(`[process-library:update] ${message}`);
        }
        return 1;
    }
}
async function handleProcessLibraryUse(args) {
    try {
        const result = args.dir
            ? await (0, active_1.bindActiveProcessLibrary)({
                dir: args.dir,
                stateDir: args.stateDir,
                runId: args.runId,
                sessionId: args.sessionId,
                ref: args.ref,
            })
            : await (0, active_1.ensureActiveProcessLibrary)({
                stateDir: args.stateDir,
                runId: args.runId,
                sessionId: args.sessionId,
                ref: args.ref,
            });
        if (args.json) {
            console.log(JSON.stringify({ success: true, ...result }, null, 2));
        }
        else {
            const boundDir = result.binding?.dir ?? "(missing)";
            console.log(`Active process library updated.\n  Scope: ${result.bindingScope}\n  Dir: ${boundDir}\n  State: ${result.stateFile}`);
        }
        return 0;
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        if (args.json) {
            console.log(JSON.stringify({ error: "bind_failed", message }));
        }
        else {
            console.error(`[process-library:use] ${message}`);
        }
        return 1;
    }
}
async function handleProcessLibraryActive(args) {
    try {
        const result = await (0, active_1.ensureActiveProcessLibrary)({
            stateDir: args.stateDir,
            runId: args.runId,
            sessionId: args.sessionId,
        });
        if (args.json) {
            console.log(JSON.stringify(result, null, 2));
        }
        else {
            const lines = [
                "Active process library.",
                `  Scope: ${result.bindingScope}`,
                `  Dir: ${result.binding?.dir ?? "(missing)"}`,
                `  Revision: ${result.binding?.revision ?? "unknown"}`,
                `  State: ${result.stateFile}`,
            ];
            if (result.bootstrapped) {
                lines.push(`  Bootstrapped: yes (${result.defaultSpec.cloneDir})`);
            }
            console.log(lines.join("\n"));
        }
        return 0;
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        if (args.json) {
            console.log(JSON.stringify({ error: "active_failed", message }));
        }
        else {
            console.error(`[process-library:active] ${message}`);
        }
        return 1;
    }
}
async function handleProcessLibraryCommand(args) {
    switch (args.subcommand) {
        case "clone":
            return handleProcessLibraryClone(args);
        case "update":
            return handleProcessLibraryUpdate(args);
        case "use":
            return handleProcessLibraryUse(args);
        case "active":
            return handleProcessLibraryActive(args);
        default: {
            const _exhaustive = args.subcommand;
            void _exhaustive;
            console.error("[process-library] Unknown subcommand");
            return 1;
        }
    }
}
