// @a5c-ai/babysitter-sdk/dist/mcp/tools/discovery.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerDiscoveryTools = registerDiscoveryTools;
const path = __importStar(require("path"));
const zod_1 = require("zod");
const health_1 = require("../../cli/commands/health");
const configure_1 = require("../../cli/commands/configure");
const skill_1 = require("../../cli/commands/skill");
const errors_1 = require("../util/errors");
function registerDiscoveryTools(server) {
    // ── skill_discover ──────────────────────────────────────────────────
    server.tool("skill_discover", "Discover available skills, agents, and process definitions", {
        pluginRoot: zod_1.z
            .string()
            .describe("Plugin root directory to scan for skills"),
        runId: zod_1.z
            .string()
            .optional()
            .describe("Run ID for domain-scoped discovery"),
        runsDir: zod_1.z.string().optional().describe("Override runs directory path"),
        includeRemote: zod_1.z
            .boolean()
            .optional()
            .describe("Include remote skill sources"),
        processPath: zod_1.z
            .string()
            .optional()
            .describe("Process file path for specialization scoping"),
    }, async (args) => {
        try {
            const pluginRoot = path.resolve(args.pluginRoot);
            const result = await (0, skill_1.discoverSkillsInternal)({
                pluginRoot,
                runId: args.runId,
                runsDir: args.runsDir,
                includeRemote: args.includeRemote,
                processPath: args.processPath,
                includeProcesses: true,
            });
            return (0, errors_1.toolResult)({
                skills: result.skills,
                agents: result.agents,
                processes: result.processes,
                summary: result.summary,
                cached: result.cached,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
    // ── health ──────────────────────────────────────────────────────────
    server.tool("health", "Check the health and configuration status of the babysitter installation", {}, async () => {
        try {
            const result = await (0, health_1.runHealthCheck)({ json: true });
            return (0, errors_1.toolResult)({
                status: result.status,
                summary: result.summary,
                checks: result.checks,
                nextSteps: result.nextSteps,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
    // ── configure_show ──────────────────────────────────────────────────
    server.tool("configure_show", "Show current babysitter configuration and environment", {}, () => {
        try {
            const result = (0, configure_1.configureShow)({});
            return (0, errors_1.toolResult)({
                values: result.values,
                timestamp: result.timestamp,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
}
