// @a5c-ai/babysitter-sdk/dist/mcp/tools/runs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerRunTools = registerRunTools;
const path = __importStar(require("path"));
const zod_1 = require("zod");
const runtime_1 = require("../../runtime");
const storage_1 = require("../../storage");
const stateCache_1 = require("../../runtime/replay/stateCache");
const errors_1 = require("../util/errors");
const resolve_run_dir_1 = require("../util/resolve-run-dir");
/**
 * Parse an entrypoint specifier like "path/to/file.js#exportName" into its parts.
 */
function parseEntrypoint(entrypoint) {
    const hashIndex = entrypoint.lastIndexOf("#");
    if (hashIndex <= 0) {
        return { importPath: entrypoint };
    }
    const importPath = entrypoint.slice(0, hashIndex);
    const exportName = entrypoint.slice(hashIndex + 1) || undefined;
    return { importPath, exportName };
}
const RUN_LIFECYCLE_TYPES = new Set(["RUN_CREATED", "RUN_COMPLETED", "RUN_FAILED"]);
function findLastLifecycleEvent(events) {
    for (let i = events.length - 1; i >= 0; i--) {
        if (RUN_LIFECYCLE_TYPES.has(events[i].type)) {
            return events[i];
        }
    }
    return undefined;
}
function deriveRunState(lastLifecycleEventType, pendingCount) {
    if (lastLifecycleEventType === "RUN_COMPLETED")
        return "completed";
    if (lastLifecycleEventType === "RUN_FAILED")
        return "failed";
    if (pendingCount > 0)
        return "waiting";
    return "created";
}
function countPendingEffects(events) {
    const requested = new Map();
    const resolved = new Set();
    for (const event of events) {
        if (event.type === "EFFECT_REQUESTED") {
            const data = event.data;
            if (data.effectId) {
                requested.set(data.effectId, {
                    effectId: data.effectId,
                    kind: data.kind ?? "unknown",
                    label: data.label,
                });
            }
        }
        else if (event.type === "EFFECT_RESOLVED") {
            const data = event.data;
            if (data.effectId) {
                resolved.add(data.effectId);
            }
        }
    }
    const pendingEffects = [];
    const pendingByKind = {};
    for (const [effectId, info] of requested) {
        if (!resolved.has(effectId)) {
            pendingEffects.push(info);
            pendingByKind[info.kind] = (pendingByKind[info.kind] ?? 0) + 1;
        }
    }
    return { pendingEffects, pendingByKind };
}
function registerRunTools(server) {
    // ── run_create ──────────────────────────────────────────────────────
    server.tool("run_create", "Create a new babysitter run for a given process definition", {
        processId: zod_1.z.string().describe("The process identifier to run"),
        entrypoint: zod_1.z.string().describe("Path to the process JS entrypoint file (optionally path#exportName)"),
        inputs: zod_1.z.string().optional().describe("JSON-encoded inputs for the process"),
        runsDir: zod_1.z.string().optional().describe("Override runs directory path"),
        prompt: zod_1.z.string().optional().describe("Prompt or description for the run"),
        nonInteractive: zod_1.z.boolean().optional().describe("When true, breakpoints are auto-approved without human interaction"),
    }, async (args) => {
        try {
            const { importPath, exportName } = parseEntrypoint(args.entrypoint);
            const absoluteImportPath = path.resolve(importPath);
            const runsDir = (0, resolve_run_dir_1.resolveRunDir)(args.runsDir);
            let inputs;
            if (args.inputs) {
                try {
                    inputs = JSON.parse(args.inputs);
                }
                catch {
                    return (0, errors_1.toolError)("Invalid JSON in inputs parameter");
                }
            }
            const result = await (0, runtime_1.createRun)({
                runsDir,
                process: {
                    processId: args.processId,
                    importPath: absoluteImportPath,
                    exportName,
                },
                prompt: args.prompt,
                inputs,
                ...(args.nonInteractive ? { metadata: { nonInteractive: true } } : {}),
            });
            return (0, errors_1.toolResult)({
                runId: result.runId,
                runDir: result.runDir,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
    // ── run_status ──────────────────────────────────────────────────────
    server.tool("run_status", "Get the current status and metadata of a run", {
        runId: zod_1.z.string().describe("The run ID to query"),
        runsDir: zod_1.z.string().optional().describe("Override runs directory path"),
    }, async (args) => {
        try {
            const runsDir = (0, resolve_run_dir_1.resolveRunDir)(args.runsDir);
            const runDir = path.join(runsDir, args.runId);
            const metadata = await (0, storage_1.readRunMetadata)(runDir);
            const journal = await (0, storage_1.loadJournal)(runDir);
            const lastLifecycleEvent = findLastLifecycleEvent(journal);
            const { pendingEffects, pendingByKind } = countPendingEffects(journal);
            const state = deriveRunState(lastLifecycleEvent?.type, pendingEffects.length);
            const completionProof = state === "completed" ? metadata.completionProof : undefined;
            return (0, errors_1.toolResult)({
                runId: metadata.runId,
                processId: metadata.processId,
                state,
                pendingEffects,
                pendingByKind,
                completionProof: completionProof ?? null,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
    // ── run_iterate ─────────────────────────────────────────────────────
    server.tool("run_iterate", "Execute one orchestration iteration for a run", {
        runId: zod_1.z.string().describe("The run ID to iterate"),
        runsDir: zod_1.z.string().optional().describe("Override runs directory path"),
    }, async (args) => {
        try {
            const runsDir = (0, resolve_run_dir_1.resolveRunDir)(args.runsDir);
            const runDir = path.join(runsDir, args.runId);
            const result = await (0, runtime_1.orchestrateIteration)({ runDir });
            if (result.status === "completed") {
                const metadata = await (0, storage_1.readRunMetadata)(runDir);
                const completionProof = metadata.completionProof ?? null;
                return (0, errors_1.toolResult)({
                    status: "completed",
                    output: result.output,
                    completionProof,
                    metadata: result.metadata,
                });
            }
            if (result.status === "failed") {
                return (0, errors_1.toolResult)({
                    status: "failed",
                    error: result.error instanceof Error ? result.error.message : result.error,
                    metadata: result.metadata,
                });
            }
            // status === "waiting"
            return (0, errors_1.toolResult)({
                status: "waiting",
                nextActions: result.nextActions.map((action) => ({
                    effectId: action.effectId,
                    kind: action.kind,
                    label: action.label,
                    labels: action.labels,
                    taskId: action.taskId,
                    stepId: action.stepId,
                })),
                metadata: result.metadata,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
    // ── run_events ──────────────────────────────────────────────────────
    server.tool("run_events", "List journal events for a run", {
        runId: zod_1.z.string().describe("The run ID to query"),
        runsDir: zod_1.z.string().optional().describe("Override runs directory path"),
        limit: zod_1.z.number().optional().describe("Maximum number of events to return"),
        filterType: zod_1.z.string().optional().describe("Filter events by type (e.g. EFFECT_REQUESTED)"),
        reverse: zod_1.z.boolean().optional().describe("Return events in reverse chronological order"),
    }, async (args) => {
        try {
            const runsDir = (0, resolve_run_dir_1.resolveRunDir)(args.runsDir);
            const runDir = path.join(runsDir, args.runId);
            const journal = await (0, storage_1.loadJournal)(runDir);
            // Apply type filter
            const filterType = args.filterType?.toUpperCase();
            const filtered = filterType
                ? journal.filter((event) => event.type.toUpperCase() === filterType)
                : journal;
            // Apply ordering
            const ordered = args.reverse ? filtered.slice().reverse() : filtered;
            // Apply limit
            const limited = args.limit !== undefined ? ordered.slice(0, args.limit) : ordered;
            return (0, errors_1.toolResult)({
                total: journal.length,
                matching: filtered.length,
                showing: limited.length,
                events: limited.map((event) => ({
                    seq: event.seq,
                    type: event.type,
                    recordedAt: event.recordedAt,
                    data: event.data,
                })),
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
    // ── run_rebuild_state ───────────────────────────────────────────────
    server.tool("run_rebuild_state", "Rebuild the state cache for a run from its journal", {
        runId: zod_1.z.string().describe("The run ID to rebuild state for"),
        runsDir: zod_1.z.string().optional().describe("Override runs directory path"),
    }, async (args) => {
        try {
            const runsDir = (0, resolve_run_dir_1.resolveRunDir)(args.runsDir);
            const runDir = path.join(runsDir, args.runId);
            const snapshot = await (0, stateCache_1.rebuildStateCache)(runDir);
            return (0, errors_1.toolResult)({
                success: true,
                stateVersion: snapshot.stateVersion,
                journalHead: snapshot.journalHead,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
}
