// @a5c-ai/babysitter-sdk/dist/mcp/tools/sessions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerSessionTools = registerSessionTools;
const path = __importStar(require("path"));
const zod_1 = require("zod");
const session_1 = require("../../session");
const storage_1 = require("../../storage");
const errors_1 = require("../util/errors");
const resolve_run_dir_1 = require("../util/resolve-run-dir");
function registerSessionTools(server) {
    // ── session_init ────────────────────────────────────────────────────
    server.tool("session_init", "Initialize a new session state for orchestration", {
        sessionId: zod_1.z.string().describe("The session ID to initialize"),
        stateDir: zod_1.z.string().describe("Directory to store session state files"),
        maxIterations: zod_1.z
            .number()
            .optional()
            .describe("Maximum number of iterations (default 256)"),
        runId: zod_1.z
            .string()
            .optional()
            .describe("Optional run ID to associate immediately"),
        prompt: zod_1.z
            .string()
            .optional()
            .describe("Optional prompt text for the session"),
    }, async (args) => {
        try {
            const stateDir = path.resolve(args.stateDir);
            const maxIterations = args.maxIterations ?? 256;
            const runId = args.runId ?? "";
            const prompt = args.prompt ?? "";
            const filePath = (0, session_1.getSessionFilePath)(stateDir, args.sessionId);
            // Check for existing session
            if (await (0, session_1.sessionFileExists)(filePath)) {
                try {
                    const existing = await (0, session_1.readSessionFile)(filePath);
                    if (existing.state.runId) {
                        return (0, errors_1.toolError)(`Session already associated with run: ${existing.state.runId}`);
                    }
                    return (0, errors_1.toolError)("A babysitter run is already active for this session");
                }
                catch {
                    return (0, errors_1.toolError)("Session state file exists but could not be read");
                }
            }
            const now = (0, session_1.getCurrentTimestamp)();
            const state = {
                active: true,
                iteration: 1,
                maxIterations,
                runId,
                startedAt: now,
                lastIterationAt: now,
                iterationTimes: [],
            };
            await (0, session_1.writeSessionFile)(filePath, state, prompt);
            return (0, errors_1.toolResult)({
                stateFile: filePath,
                iteration: state.iteration,
                maxIterations: state.maxIterations,
                runId: state.runId,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
    // ── session_associate ───────────────────────────────────────────────
    server.tool("session_associate", "Associate a session with a run ID", {
        sessionId: zod_1.z.string().describe("The session ID to associate"),
        stateDir: zod_1.z.string().describe("Directory containing session state files"),
        runId: zod_1.z.string().describe("The run ID to associate with"),
    }, async (args) => {
        try {
            const stateDir = path.resolve(args.stateDir);
            const filePath = (0, session_1.getSessionFilePath)(stateDir, args.sessionId);
            let existing;
            try {
                existing = await (0, session_1.readSessionFile)(filePath);
            }
            catch {
                return (0, errors_1.toolError)(`No active session found. Expected state file: ${filePath}`);
            }
            if (existing.state.runId) {
                return (0, errors_1.toolError)(`Session already associated with run: ${existing.state.runId}`);
            }
            const updatedState = {
                ...existing.state,
                runId: args.runId,
            };
            await (0, session_1.writeSessionFile)(filePath, updatedState, existing.prompt);
            return (0, errors_1.toolResult)({
                stateFile: filePath,
                runId: args.runId,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
    // ── session_resume ──────────────────────────────────────────────────
    server.tool("session_resume", "Resume an existing run in a new session", {
        sessionId: zod_1.z.string().describe("The session ID for the new session"),
        stateDir: zod_1.z.string().describe("Directory to store session state files"),
        runId: zod_1.z.string().describe("The run ID to resume"),
        maxIterations: zod_1.z
            .number()
            .optional()
            .describe("Maximum number of iterations (default 256)"),
        runsDir: zod_1.z.string().optional().describe("Override runs directory path"),
    }, async (args) => {
        try {
            const stateDir = path.resolve(args.stateDir);
            const runsDir = (0, resolve_run_dir_1.resolveRunDir)(args.runsDir);
            const runDir = path.join(runsDir, args.runId);
            const maxIterations = args.maxIterations ?? 256;
            // Check run state from journal
            let runState = "unknown";
            try {
                const journal = await (0, storage_1.loadJournal)(runDir);
                const hasCompleted = journal.some((e) => e.type === "RUN_COMPLETED");
                const hasFailed = journal.some((e) => e.type === "RUN_FAILED");
                if (hasCompleted)
                    runState = "completed";
                else if (hasFailed)
                    runState = "failed";
                else
                    runState = "waiting";
            }
            catch {
                return (0, errors_1.toolError)(`Run not found: ${args.runId}`);
            }
            if (runState === "completed") {
                return (0, errors_1.toolError)(`Run is already completed. Cannot resume a completed run.`);
            }
            const filePath = (0, session_1.getSessionFilePath)(stateDir, args.sessionId);
            const prompt = `Resume Babysitter run: ${args.runId}\nCurrent state: ${runState}\nContinue orchestration using run:iterate, task:post, etc.`;
            const now = (0, session_1.getCurrentTimestamp)();
            const state = {
                active: true,
                iteration: 1,
                maxIterations,
                runId: args.runId,
                startedAt: now,
                lastIterationAt: now,
                iterationTimes: [],
            };
            await (0, session_1.writeSessionFile)(filePath, state, prompt);
            return (0, errors_1.toolResult)({
                stateFile: filePath,
                runId: args.runId,
                runState,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
    // ── session_state ───────────────────────────────────────────────────
    server.tool("session_state", "Get the current state of a session", {
        sessionId: zod_1.z.string().describe("The session ID to query"),
        stateDir: zod_1.z.string().describe("Directory containing session state files"),
    }, async (args) => {
        try {
            const stateDir = path.resolve(args.stateDir);
            const filePath = (0, session_1.getSessionFilePath)(stateDir, args.sessionId);
            if (!(await (0, session_1.sessionFileExists)(filePath))) {
                return (0, errors_1.toolResult)({
                    found: false,
                    stateFile: filePath,
                });
            }
            const file = await (0, session_1.readSessionFile)(filePath);
            return (0, errors_1.toolResult)({
                found: true,
                state: file.state,
                prompt: file.prompt,
                stateFile: filePath,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
}
