// @a5c-ai/babysitter-sdk/dist/mcp/tools/tasks.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerTaskTools = registerTaskTools;
const path = __importStar(require("path"));
const zod_1 = require("zod");
const commitEffectResult_1 = require("../../runtime/commitEffectResult");
const storage_1 = require("../../storage");
const tasks_1 = require("../../storage/tasks");
const errors_1 = require("../util/errors");
const resolve_run_dir_1 = require("../util/resolve-run-dir");
/**
 * Build a task list from journal events by tracking EFFECT_REQUESTED and
 * EFFECT_RESOLVED events.
 */
function buildTaskList(events) {
    const requested = new Map();
    const resolved = new Map();
    for (const event of events) {
        if (event.type === "EFFECT_REQUESTED") {
            const data = event.data;
            if (data.effectId) {
                requested.set(data.effectId, {
                    effectId: data.effectId,
                    kind: data.kind ?? "unknown",
                    label: data.label,
                    taskId: data.taskId,
                    requestedAt: event.recordedAt,
                });
            }
        }
        else if (event.type === "EFFECT_RESOLVED") {
            const data = event.data;
            if (data.effectId) {
                resolved.set(data.effectId, { resolvedAt: event.recordedAt });
            }
        }
    }
    const tasks = [];
    for (const [effectId, info] of requested) {
        const resolvedInfo = resolved.get(effectId);
        tasks.push({
            effectId,
            kind: info.kind,
            status: resolvedInfo ? "resolved" : "pending",
            label: info.label,
            taskId: info.taskId,
            requestedAt: info.requestedAt,
            resolvedAt: resolvedInfo?.resolvedAt,
        });
    }
    return tasks.sort((a, b) => a.effectId.localeCompare(b.effectId));
}
function registerTaskTools(server) {
    // ── task_post ───────────────────────────────────────────────────────
    server.tool("task_post", "Post a result for a pending task effect", {
        runId: zod_1.z.string().describe("The run ID the task belongs to"),
        effectId: zod_1.z.string().describe("The effect ID of the task to resolve"),
        status: zod_1.z
            .enum(["ok", "error"])
            .describe("Result status: ok for success, error for failure"),
        value: zod_1.z
            .string()
            .optional()
            .describe("JSON-encoded result value (when status=ok)"),
        error: zod_1.z
            .string()
            .optional()
            .describe("JSON-encoded error payload (when status=error)"),
        runsDir: zod_1.z.string().optional().describe("Override runs directory path"),
    }, async (args) => {
        try {
            const runsDir = (0, resolve_run_dir_1.resolveRunDir)(args.runsDir);
            const runDir = path.join(runsDir, args.runId);
            let value;
            let errorPayload;
            if (args.status === "ok" && args.value) {
                try {
                    value = JSON.parse(args.value);
                }
                catch {
                    return (0, errors_1.toolError)("Invalid JSON in value parameter");
                }
            }
            if (args.status === "error") {
                if (args.error) {
                    try {
                        errorPayload = JSON.parse(args.error);
                    }
                    catch {
                        return (0, errors_1.toolError)("Invalid JSON in error parameter");
                    }
                }
                else {
                    errorPayload = { name: "Error", message: "Task reported failure" };
                }
            }
            const nowIso = new Date().toISOString();
            const committed = await (0, commitEffectResult_1.commitEffectResult)({
                runDir,
                effectId: args.effectId,
                result: args.status === "ok"
                    ? {
                        status: "ok",
                        value,
                        startedAt: nowIso,
                        finishedAt: nowIso,
                    }
                    : {
                        status: "error",
                        error: errorPayload,
                        startedAt: nowIso,
                        finishedAt: nowIso,
                    },
            });
            return (0, errors_1.toolResult)({
                status: args.status,
                effectId: args.effectId,
                resultRef: committed.resultRef ?? null,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
    // ── task_list ───────────────────────────────────────────────────────
    server.tool("task_list", "List all tasks for a run, optionally showing only pending tasks", {
        runId: zod_1.z.string().describe("The run ID to list tasks for"),
        pendingOnly: zod_1.z
            .boolean()
            .optional()
            .describe("If true, only show pending (unresolved) tasks"),
        runsDir: zod_1.z.string().optional().describe("Override runs directory path"),
    }, async (args) => {
        try {
            const runsDir = (0, resolve_run_dir_1.resolveRunDir)(args.runsDir);
            const runDir = path.join(runsDir, args.runId);
            const journal = await (0, storage_1.loadJournal)(runDir);
            const allTasks = buildTaskList(journal);
            const tasks = args.pendingOnly
                ? allTasks.filter((t) => t.status === "pending")
                : allTasks;
            return (0, errors_1.toolResult)({
                total: allTasks.length,
                showing: tasks.length,
                pendingCount: allTasks.filter((t) => t.status === "pending").length,
                tasks,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
    // ── task_show ───────────────────────────────────────────────────────
    server.tool("task_show", "Show details of a specific task including its definition and result", {
        runId: zod_1.z.string().describe("The run ID the task belongs to"),
        effectId: zod_1.z.string().describe("The effect ID of the task"),
        runsDir: zod_1.z.string().optional().describe("Override runs directory path"),
    }, async (args) => {
        try {
            const runsDir = (0, resolve_run_dir_1.resolveRunDir)(args.runsDir);
            const runDir = path.join(runsDir, args.runId);
            // Get task status from journal
            const journal = await (0, storage_1.loadJournal)(runDir);
            const allTasks = buildTaskList(journal);
            const taskEntry = allTasks.find((t) => t.effectId === args.effectId);
            if (!taskEntry) {
                return (0, errors_1.toolError)(`Effect ${args.effectId} not found in run ${args.runId}`);
            }
            // Read task definition
            let taskDef = null;
            try {
                taskDef = await (0, tasks_1.readTaskDefinition)(runDir, args.effectId);
            }
            catch {
                // Task definition might not exist yet
            }
            // Read task result if resolved
            let taskResultData = null;
            if (taskEntry.status === "resolved") {
                try {
                    taskResultData = await (0, tasks_1.readTaskResult)(runDir, args.effectId);
                }
                catch {
                    // Result file might not be readable
                }
            }
            return (0, errors_1.toolResult)({
                effect: taskEntry,
                task: taskDef,
                result: taskResultData,
            });
        }
        catch (err) {
            return (0, errors_1.toolError)(err instanceof Error ? err.message : String(err));
        }
    });
}
