// @a5c-ai/babysitter-sdk/dist/plugins/migrations.js
"use strict";
/**
 * Plugin Migration Chain Resolution
 *
 * Parses migration filenames, builds a version graph, and finds
 * the shortest migration path between two versions using BFS.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseMigrationFilename = parseMigrationFilename;
exports.listMigrations = listMigrations;
exports.buildMigrationGraph = buildMigrationGraph;
exports.findMigrationPath = findMigrationPath;
exports.resolveMigrationChain = resolveMigrationChain;
const node_fs_1 = require("node:fs");
const node_path_1 = __importDefault(require("node:path"));
const types_1 = require("./types");
/**
 * Regex for migration filenames: `<from>_to_<to>.<ext>`
 * Versions may contain digits, dots, dashes, and pre-release identifiers.
 *
 * Examples: "1.0.0_to_1.1.0.md", "2.0.0-beta_to_2.0.0.js"
 */
const MIGRATION_FILENAME_REGEX = /^([a-zA-Z0-9._-]+)_to_([a-zA-Z0-9._-]+)\.(md|js)$/;
/**
 * Parses a migration filename into its constituent version parts.
 * Returns undefined if the filename does not match the expected pattern.
 *
 * @param filename - The migration filename to parse
 */
function parseMigrationFilename(filename) {
    const match = MIGRATION_FILENAME_REGEX.exec(filename);
    if (!match) {
        return undefined;
    }
    const [, from, to, ext] = match;
    return {
        from,
        to,
        file: filename,
        type: ext,
    };
}
/**
 * Lists all migration descriptors found in a migrations directory.
 * Ignores files that do not match the expected naming pattern.
 *
 * @param migrationsDir - Absolute path to the migrations directory
 */
async function listMigrations(migrationsDir) {
    let entries;
    try {
        entries = await node_fs_1.promises.readdir(migrationsDir);
    }
    catch (error) {
        if ((0, types_1.isNodeError)(error) && error.code === "ENOENT") {
            return [];
        }
        throw error;
    }
    const descriptors = [];
    for (const entry of entries) {
        const descriptor = parseMigrationFilename(entry);
        if (descriptor) {
            descriptors.push(descriptor);
        }
    }
    return descriptors.sort((a, b) => a.from.localeCompare(b.from));
}
/**
 * Builds a directed adjacency list from version to version,
 * keyed by "from" version, with edges pointing to "to" versions.
 *
 * @param migrations - Array of migration descriptors
 */
function buildMigrationGraph(migrations) {
    const graph = new Map();
    for (const migration of migrations) {
        const existing = graph.get(migration.from);
        if (existing) {
            existing.push(migration);
        }
        else {
            graph.set(migration.from, [migration]);
        }
    }
    return graph;
}
/**
 * Finds the shortest migration path from one version to another using BFS.
 * Returns the ordered list of migration descriptors, or undefined if no path exists.
 *
 * @param migrations - All available migration descriptors
 * @param fromVersion - Starting version
 * @param toVersion - Target version
 */
function findMigrationPath(migrations, fromVersion, toVersion) {
    if (fromVersion === toVersion) {
        return [];
    }
    const graph = buildMigrationGraph(migrations);
    // BFS
    const queue = [
        { version: fromVersion, path: [] },
    ];
    const visited = new Set([fromVersion]);
    while (queue.length > 0) {
        const current = queue.shift();
        const neighbors = graph.get(current.version) ?? [];
        for (const edge of neighbors) {
            if (edge.to === toVersion) {
                return [...current.path, edge];
            }
            if (!visited.has(edge.to)) {
                visited.add(edge.to);
                queue.push({
                    version: edge.to,
                    path: [...current.path, edge],
                });
            }
        }
    }
    return undefined;
}
/**
 * Full migration resolution flow: lists migrations from a package directory,
 * finds the shortest path, and returns the migration descriptors with their
 * file contents loaded.
 *
 * @param packageDir - Absolute path to the plugin package directory (containing migrations/ subdir)
 * @param fromVersion - Current installed version
 * @param toVersion - Target version to migrate to
 * @returns Array of objects with descriptor and content, or undefined if no path exists
 */
async function resolveMigrationChain(packageDir, fromVersion, toVersion) {
    const migrationsDir = node_path_1.default.join(packageDir, "migrations");
    const allMigrations = await listMigrations(migrationsDir);
    const migrationPath = findMigrationPath(allMigrations, fromVersion, toVersion);
    if (!migrationPath) {
        return undefined;
    }
    const results = [];
    for (const descriptor of migrationPath) {
        const filePath = node_path_1.default.join(migrationsDir, descriptor.file);
        const content = await node_fs_1.promises.readFile(filePath, "utf8");
        results.push({ descriptor, content });
    }
    return results;
}
