// @a5c-ai/babysitter-sdk/dist/plugins/packageReader.js
"use strict";
/**
 * Plugin Package Reader
 *
 * Reads plugin package contents including instruction files,
 * migration descriptors, and process definition files.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readPluginPackage = readPluginPackage;
exports.readInstallInstructions = readInstallInstructions;
exports.readUninstallInstructions = readUninstallInstructions;
exports.readConfigureInstructions = readConfigureInstructions;
exports.listMigrations = listMigrations;
exports.readMigration = readMigration;
const node_fs_1 = require("node:fs");
const node_path_1 = __importDefault(require("node:path"));
const types_1 = require("./types");
const migrations_1 = require("./migrations");
/**
 * Standard instruction file names within a plugin package.
 */
const INSTALL_FILE = "install.md";
const UNINSTALL_FILE = "uninstall.md";
const CONFIGURE_FILE = "configure.md";
const MIGRATIONS_DIR = "migrations";
const PROCESS_DIR = "process";
/**
 * Reads a text file, returning undefined if the file does not exist.
 */
async function readOptionalFile(filePath) {
    try {
        return await node_fs_1.promises.readFile(filePath, "utf8");
    }
    catch (error) {
        if ((0, types_1.isNodeError)(error) && error.code === "ENOENT") {
            return undefined;
        }
        throw error;
    }
}
/**
 * Recursively collects all files under a directory, returning paths
 * relative to the given base directory.
 */
async function collectFiles(dir, base) {
    const results = [];
    try {
        const entries = await node_fs_1.promises.readdir(dir, { withFileTypes: true });
        for (const entry of entries) {
            const fullPath = node_path_1.default.join(dir, entry.name);
            if (entry.isDirectory()) {
                const nested = await collectFiles(fullPath, base);
                results.push(...nested);
            }
            else {
                results.push(node_path_1.default.relative(base, fullPath).replace(/\\/g, "/"));
            }
        }
    }
    catch (error) {
        if ((0, types_1.isNodeError)(error) && error.code === "ENOENT") {
            return [];
        }
        throw error;
    }
    return results.sort();
}
/**
 * Reads all metadata and instruction files from a plugin package directory.
 *
 * @param packageDir - Absolute path to the plugin package directory
 */
async function readPluginPackage(packageDir) {
    // Read package.json or plugin.json for the name
    let name = node_path_1.default.basename(packageDir);
    try {
        const pkgRaw = await node_fs_1.promises.readFile(node_path_1.default.join(packageDir, "package.json"), "utf8");
        const pkg = JSON.parse(pkgRaw);
        if (typeof pkg.name === "string") {
            name = pkg.name;
        }
    }
    catch {
        // No package.json is acceptable; use directory name
    }
    const [installInstructions, uninstallInstructions, configureInstructions, migrations, processFiles] = await Promise.all([
        readInstallInstructions(packageDir),
        readUninstallInstructions(packageDir),
        readConfigureInstructions(packageDir),
        listMigrations(packageDir),
        collectFiles(node_path_1.default.join(packageDir, PROCESS_DIR), packageDir),
    ]);
    return {
        name,
        installInstructions,
        uninstallInstructions,
        configureInstructions,
        migrations,
        processFiles,
    };
}
/**
 * Reads the install instructions markdown file from a plugin package.
 *
 * @param packageDir - Absolute path to the plugin package directory
 */
async function readInstallInstructions(packageDir) {
    return readOptionalFile(node_path_1.default.join(packageDir, INSTALL_FILE));
}
/**
 * Reads the uninstall instructions markdown file from a plugin package.
 *
 * @param packageDir - Absolute path to the plugin package directory
 */
async function readUninstallInstructions(packageDir) {
    return readOptionalFile(node_path_1.default.join(packageDir, UNINSTALL_FILE));
}
/**
 * Reads the configure instructions markdown file from a plugin package.
 *
 * @param packageDir - Absolute path to the plugin package directory
 */
async function readConfigureInstructions(packageDir) {
    return readOptionalFile(node_path_1.default.join(packageDir, CONFIGURE_FILE));
}
/**
 * Lists all migration descriptors found in the migrations/ subdirectory.
 * Filenames must follow the pattern: `<from>_to_<to>.<ext>`
 *
 * @param packageDir - Absolute path to the plugin package directory
 */
async function listMigrations(packageDir) {
    return (0, migrations_1.listMigrations)(node_path_1.default.join(packageDir, MIGRATIONS_DIR));
}
/**
 * Reads the content of a specific migration file.
 *
 * @param packageDir - Absolute path to the plugin package directory
 * @param from - Source version
 * @param to - Target version
 */
async function readMigration(packageDir, from, to) {
    const migrationsDir = node_path_1.default.join(packageDir, MIGRATIONS_DIR);
    const descriptors = await (0, migrations_1.listMigrations)(migrationsDir);
    const match = descriptors.find((d) => d.from === from && d.to === to);
    if (!match) {
        throw new Error(`Migration from "${from}" to "${to}" not found in ${migrationsDir}`);
    }
    return node_fs_1.promises.readFile(node_path_1.default.join(migrationsDir, match.file), "utf8");
}
