// @a5c-ai/babysitter-sdk/dist/plugins/registry.js
"use strict";
/**
 * Plugin Registry CRUD
 *
 * Manages the plugin registry file using atomic writes for crash safety.
 * The registry tracks all installed plugins, their versions, and metadata.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.readPluginRegistry = readPluginRegistry;
exports.writePluginRegistry = writePluginRegistry;
exports.getPluginEntry = getPluginEntry;
exports.upsertPluginEntry = upsertPluginEntry;
exports.removePluginEntry = removePluginEntry;
exports.listPluginEntries = listPluginEntries;
const node_fs_1 = require("node:fs");
const atomic_1 = require("../storage/atomic");
const paths_1 = require("./paths");
const types_1 = require("./types");
/**
 * Creates an empty plugin registry with the current schema version.
 */
function createEmptyRegistry() {
    return {
        schemaVersion: types_1.PLUGIN_REGISTRY_SCHEMA_VERSION,
        updatedAt: new Date().toISOString(),
        plugins: {},
    };
}
/**
 * Reads the plugin registry from disk.
 * Returns an empty registry if the file does not exist.
 *
 * @param scope - Whether to read the global or project registry
 * @param projectDir - Required when scope is 'project'
 */
async function readPluginRegistry(scope, projectDir) {
    const registryPath = (0, paths_1.getRegistryPath)(scope, projectDir);
    try {
        const raw = await node_fs_1.promises.readFile(registryPath, "utf8");
        return JSON.parse(raw);
    }
    catch (error) {
        if ((0, types_1.isNodeError)(error) && error.code === "ENOENT") {
            return createEmptyRegistry();
        }
        throw error;
    }
}
/**
 * Writes the plugin registry to disk using an atomic write.
 *
 * @param registry - The full registry object to persist
 * @param scope - Whether to write the global or project registry
 * @param projectDir - Required when scope is 'project'
 */
async function writePluginRegistry(registry, scope, projectDir) {
    const registryPath = (0, paths_1.getRegistryPath)(scope, projectDir);
    await (0, atomic_1.writeFileAtomic)(registryPath, JSON.stringify(registry, null, 2) + "\n");
}
/**
 * Looks up a plugin entry by name in the registry.
 * Returns undefined if the plugin is not registered.
 *
 * @param registry - The registry to search
 * @param pluginName - The plugin identifier to look up
 */
function getPluginEntry(registry, pluginName) {
    return registry.plugins[pluginName];
}
/**
 * Returns a new registry with the given entry inserted or updated.
 * Sets the registry's updatedAt timestamp. Does not mutate the input.
 *
 * @param registry - The current registry
 * @param entry - The plugin entry to insert or update
 */
function upsertPluginEntry(registry, entry) {
    const now = new Date().toISOString();
    return {
        ...registry,
        updatedAt: now,
        plugins: {
            ...registry.plugins,
            [entry.name]: {
                ...entry,
                updatedAt: now,
            },
        },
    };
}
/**
 * Returns a new registry with the named plugin removed.
 * Does not mutate the input. Throws if the plugin is not found.
 *
 * @param registry - The current registry
 * @param pluginName - The plugin identifier to remove
 */
function removePluginEntry(registry, pluginName) {
    if (!(pluginName in registry.plugins)) {
        throw new Error(`Plugin "${pluginName}" is not present in the registry`);
    }
    const { [pluginName]: _removed, ...remaining } = registry.plugins;
    return {
        ...registry,
        updatedAt: new Date().toISOString(),
        plugins: remaining,
    };
}
/**
 * Returns all plugin entries as a sorted array (sorted by name).
 *
 * @param registry - The registry to list
 */
function listPluginEntries(registry) {
    return Object.values(registry.plugins).sort((a, b) => a.name.localeCompare(b.name));
}
